{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing ../get_seqs.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile ../get_seqs.py\n",
    "\n",
    "\"\"\"\n",
    "The goal for this program is to provide an iterable function that can be run\n",
    "across a reference genome to isolate and return sequences that lie between two\n",
    "primer sequences provided, given certain variables.\n",
    "The function takes the following files:\n",
    "\n",
    "    The reference genome path string \n",
    "        Either as a fasta or fastq file ending in that file name\n",
    "\n",
    "    The primers file that contains both the forward and reverse primer sequences\n",
    "\n",
    "    Parameters for determining whether the region selected is ITS or TEF\n",
    "        Implemented as a maximum and minimum length of the extracted sequence\n",
    "            This allows for extraction of other regions\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "import subprocess\n",
    "import os\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "The goal for this program is to provide an iterable function that can be run\n",
    "across a reference genome to isolate and return sequences that lie between two\n",
    "primer sequences provided, given certain variables.\n",
    "The function takes the following files:\n",
    "\n",
    "    The reference genome path string \n",
    "        Either as a fasta or fastq file ending in that file name\n",
    "\n",
    "    The primers file that contains both the forward and reverse primer sequences\n",
    "\n",
    "    Parameters for determining whether the region selected is ITS or TEF\n",
    "        Implemented as a maximum and minimum length of the extracted sequence\n",
    "            This allows for extraction of other regions\n",
    "\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"-v\", \"--verbose\", action=\"store_true\")\n",
    "group.add_argument(\"-q\", \"--quiet\", action=\"store_true\")\n",
    "parser.add_argument(\"reference_genome\", help=\"The reference genome to scan for the particular sequence\")\n",
    "parser.add_argument(\"primer\", help=\"The primer sequences on either side of the desired sequence\")\n",
    "parser.add_argument(\"minimum\", type=int, help=\"Minimum length of the sequence\")\n",
    "parser.add_argument(\"maximum\", type=int, help=\"Maximum length of the sequence\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "\n",
    "if args.reference_genome[-5:] == \"fastq\":\n",
    "    reference = SeqIO.convert(os.path.abspath(args.reference_genome), \"fastq\", args.reference_genome[:-5]+'fasta', \"fasta\")\n",
    "elif args.reference_genome[-5:] == \"fasta\":\n",
    "    reference = os.path.abspath(args.reference_genome)\n",
    "else:\n",
    "    print(\"ERROR: File not supported. Please use a fasta or fastq file with a file ending of either fasta or fastq respectively\")\n",
    "    exit()\n",
    "\n",
    "if args.primer[-5:] == \"fasta\":\n",
    "    pass\n",
    "else:\n",
    "    print(\"ERROR: File not supported. Please use a fasta file with a .fasta ending\")\n",
    "    exit()\n",
    "\n",
    "\n",
    "database = reference[:-6]+'db'\n",
    "outfmt6 = reference[:-6]+'.outfmt6'\n",
    "\n",
    "cmd = 'makeblastdb -in %s -dbtype nucl -out %s' % (reference, database)\n",
    "subprocess.getoutput(cmd)\n",
    "\n",
    "cmd2 = 'blastn -query %s -db %s -evalue=100000 -task \"blastn-short\" -outfmt 6 > %s' % (args.primer, database, outfmt6)\n",
    "subprocess.getoutput(cmd2)\n",
    "\n",
    "df = pd.read_csv(outfmt6, sep=\"\\t\", header=None, names=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\", \"qend\", \"sstart\", \"send\", \"evalue\", \"bitscore\"])\n",
    "\n",
    "forward_df = df.loc[(df['qseqid']=='forward')]\n",
    "forward_df = forward_df.loc[df['evalue']==min(forward_df['evalue'])]\n",
    "forward_df = forward_df.reset_index()\n",
    "\n",
    "reverse_df = df.loc[(df['qseqid']=='reverse')]\n",
    "reverse_df = reverse_df.loc[df['evalue']==min(reverse_df['evalue'])]\n",
    "reverse_df = reverse_df.reset_index()\n",
    "\n",
    "forward_bed = pd.DataFrame(columns=['chrom', 'chromStart', 'chromEnd'])\n",
    "for i in range(0, len(forward_df)):\n",
    "    if forward_df['sstart'][i] <= forward_df['send'][i]:\n",
    "        forward_bed = forward_bed.append({'chrom': forward_df['sseqid'][i],'chromStart': (forward_df['sstart'][i])-1, 'chromEnd': forward_df['send'][i]}, ignore_index=True)\n",
    "    else:\n",
    "        forward_bed = forward_bed.append({'chrom': forward_df['sseqid'][i],'chromStart': (forward_df['send'][i])-1, 'chromEnd': forward_df['sstart'][i]}, ignore_index=True)\n",
    "forward_bed.sort_values(['chromStart'])\n",
    "\n",
    "reverse_bed = pd.DataFrame(columns=['chrom', 'chromStart', 'chromEnd'])\n",
    "for i in reverse_df.index:\n",
    "    if reverse_df['sstart'][i] < reverse_df['send'][i]:\n",
    "        reverse_bed = reverse_bed.append({'chrom': reverse_df['sseqid'][i],'chromStart': (reverse_df['sstart'][i])-1, 'chromEnd': reverse_df['send'][i]}, ignore_index=True)\n",
    "    else:\n",
    "        reverse_bed = reverse_bed.append({'chrom': reverse_df['sseqid'][i],'chromStart': (reverse_df['send'][i])-1, 'chromEnd': reverse_df['sstart'][i]}, ignore_index=True)\n",
    "reverse_bed.sort_values(['chromStart'])\n",
    "\n",
    "intervals_list = []\n",
    "for row in forward_bed.itertuples(index=True, name='Pandas'):\n",
    "    counter = 0\n",
    "    for rows in reverse_bed.itertuples(index=True, name='Pandas'):\n",
    "        # Here lies the difference between sequences other than primers used - expected sequence length\n",
    "        if getattr(row, 'chrom') == getattr(rows, 'chrom') and args.minimum < np.absolute(getattr(row, 'chromEnd') - getattr(rows, 'chromStart')) < args.maximum and getattr(row, 'chromEnd') > getattr(rows, 'chromStart'):\n",
    "            counter += 1\n",
    "            intervals_list.append([forward_bed['chrom'][getattr(row, 'Index')], reverse_bed['chromStart'][getattr(rows, 'Index')], forward_bed['chromEnd'][getattr(row, 'Index')]])\n",
    "        elif getattr(row, 'chrom') == getattr(rows, 'chrom') and args.minimum < np.absolute(getattr(row, 'chromEnd') - getattr(rows, 'chromStart')) < args.maximum and getattr(row, 'chromEnd') < getattr(rows, 'chromStart'):\n",
    "            counter += 1\n",
    "            intervals_list.append([forward_bed['chrom'][getattr(row, 'Index')], forward_bed['chromEnd'][getattr(row, 'Index')], reverse_bed['chromStart'][getattr(rows, 'Index')]])\n",
    "    if counter > 1:\n",
    "        print(\"WARNING: two reverse primers identified for one forward primer\")\n",
    "        print(row)\n",
    "\n",
    "intervals_frame = pd.DataFrame(data=intervals_list, columns=['chrom', 'chromStart', 'chromEnd'])\n",
    "\n",
    "bedfile = reference[:-6]+'.ITS.bedfile'\n",
    "\n",
    "intervals_frame.to_csv(bedfile, sep='\\t', header=False, index=False)\n",
    "\n",
    "bedoutput = reference[:-6]+'.ITS.bedoutput.fasta'\n",
    "cmd3 = 'bedtools getfasta -fo %s -fi %s -bed %s' % (bedoutput, reference, bedfile)\n",
    "subprocess.getoutput(cmd3)\n",
    "\n",
    "\n",
    "\n",
    "visual = SeqIO.parse(os.path.abspath(bedoutput), \"fasta\")\n",
    "\n",
    "    \n",
    "    \n",
    "if args.quiet:\n",
    "    print(\"\\n%i sequences found\\n\" % (len(intervals_frame)))\n",
    "    counter = 1\n",
    "    for record in visual:\n",
    "        assert \"N\" not in record.seq, \"ERROR: sequence contains 'N' \"\n",
    "        counter += 1\n",
    "elif args.verbose:\n",
    "    print(\"\\n%i sequences found, regions shown below\\n\" % (len(intervals_frame)))\n",
    "    print(intervals_frame)\n",
    "    print(\"\\nSequences shown below\")\n",
    "    counter = 1\n",
    "    for record in visual:\n",
    "        assert \"N\" not in record.seq, \"ERROR: sequence contains 'N' \"\n",
    "        print(\"\\nSequence %i\\n\" % counter)\n",
    "        print(record.seq)\n",
    "        counter += 1\n",
    "        print(\"\\n\")\n",
    "else:\n",
    "    print(\"\\n%i sequences found, sequences shown below\" % (len(intervals_frame)))\n",
    "    counter = 1\n",
    "    for record in visual:\n",
    "        assert \"N\" not in record.seq, \"ERROR: sequence contains 'N' \"\n",
    "        print(\"\\nSequence %i\\n\" % counter)\n",
    "        print(record.seq)\n",
    "        counter += 1\n",
    "        print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
