{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from mothur_py import Mothur\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "import re\n",
    "\n",
    "_nsre = re.compile('([0-9]+)')\n",
    "def natural_sort_key(s):\n",
    "    return [int(text) if text.isdigit() else text.lower()\n",
    "            for text in re.split(_nsre, s)] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_dict = SeqIO.to_dict(SeqIO.parse(\"../../database/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta\", \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "205\n",
      "3526\n",
      "553.7004794347716\n",
      "540.0\n"
     ]
    }
   ],
   "source": [
    "lengthe = []\n",
    "for key in fasta_dict:\n",
    "    lengthe.append(len(fasta_dict[key].seq))\n",
    "print(min(lengthe))\n",
    "print(max(lengthe))\n",
    "print(np.mean(lengthe))\n",
    "print(np.median(lengthe))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_dict = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Length_Filtered/clinical/barcode04/length_restricted_reads.fasta\", \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2837\n"
     ]
    }
   ],
   "source": [
    "print(len(fasta_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1707\n",
      "3546\n",
      "2854.5565738456116\n"
     ]
    }
   ],
   "source": [
    "lengthe = []\n",
    "for key in fasta_dict:\n",
    "    lengthe.append(len(fasta_dict[key].seq))\n",
    "print(min(lengthe))\n",
    "print(max(lengthe))\n",
    "print(np.mean(lengthe))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined = {**fasta_dict1,**fasta_dict2}\n",
    "print(len(combined))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SeqIO.write(combined.values(), \"../../analysis/Length_Filtered/20171207_FAH18654/barcode03/length_restricted_for_use.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_dict = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Length_Filtered/20171207_FAH18654/barcode03/length_restricted_for_use.fasta\", \"fasta\"))\n",
    "print(len(combined_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fastq_dict = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Concatenated/20171103_FAH15473/barcode02/merged.fastq\", \"fastq\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lengths = []\n",
    "for key in fastq_dict:\n",
    "    lengths.append(len(fastq_dict[key].seq))\n",
    "print(\"The number of reads in this file is\", len(fastq_dict))\n",
    "ax = sns.distplot(lengths, color=\"k\", kde=False, bins=5000)\n",
    "\n",
    "\n",
    "ax.set(xlim=(250, 3500))\n",
    "ax.set_title(\"Reads spread 20171103_FAH15473/barcode02\", fontsize=15)\n",
    "ax.set_xlabel(\"Length of read\", fontsize=13)\n",
    "ax.set_ylabel(\"Number of reads\", fontsize=13)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "frDNA = fastq_dict.copy()\n",
    "for key in fastq_dict:\n",
    "    if len(fastq_dict[key].seq) not in range(2700, 3200):\n",
    "        del frDNA[key]\n",
    "print(\"The number of reads between 2700 and 3200 bp in length is\", len(frDNA))\n",
    "EF1a = fastq_dict.copy()\n",
    "for key in fastq_dict:\n",
    "    if len(fastq_dict[key].seq) not in range(900, 1400):\n",
    "        del EF1a[key]\n",
    "print(\"The number of reads between 900 and 1400 bp in length is\", len(EF1a))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the reads defined between the frDNA cutoff into a new fasta file\n",
    "SeqIO.write(frDNA.values(), \"../../analysis/Python_Processing/20171103_FAH15473/barcode02/barcode02_frDNA_clipped.fastq\", \"fastq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the reads defined between the EF1a cutoff into a new fasta file\n",
    "SeqIO.write(EF1a.values(), \"../../analysis/Python_Processing/20171103_FAH15473/barcode02/barcode02_EF1a_clipped.fastq\", \"fastq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "frDNA_extract={k: frDNA[k] for k in list(frDNA.keys())[:500]}\n",
    "EF1a_extract={k: EF1a[k] for k in list(EF1a.keys())[:500]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in frDNA_extract:\n",
    "    frDNA_extract[key].annotations = 'frDNA'\n",
    "for key in EF1a_extract:\n",
    "    EF1a_extract[key].annotations = 'EF1a'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_extract = {}\n",
    "combined_extract.update(frDNA_extract)\n",
    "combined_extract.update(EF1a_extract)\n",
    "print(len(frDNA_extract))\n",
    "print(len(EF1a_extract))\n",
    "print(len(combined_extract))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SeqIO.write(frDNA_extract.values(), \"../../analysis/Python_Processing/20171103_FAH15473/barcode02/barcode02_frDNA_extract_test.fastq\", \"fastq\")\n",
    "SeqIO.write(EF1a_extract.values(), \"../../analysis/Python_Processing/20171103_FAH15473/barcode02/barcode02_EF1a_extract_test.fastq\", \"fastq\")\n",
    "SeqIO.write(combined_extract.values(), \"../../analysis/Python_Processing/20171103_FAH15473/barcode02/barcode02_combined_extract_test.fastq\", \"fastq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "frDNA_paf = pd.read_csv(\"../../analysis/Python_Processing/20171103_FAH15473/barcode02/frDNA_clipped_test.paf\", sep='\\t', header=None, engine='python')\n",
    "EF1a_paf = pd.read_csv(\"../../analysis/Python_Processing/20171103_FAH15473/barcode02/EF1a_clipped_test.paf\", sep='\\t', header=None, engine='python')\n",
    "combined_paf = pd.read_csv(\"../../analysis/Python_Processing/20171103_FAH15473/barcode02/combined_test.paf\", sep='\\t', header=None, engine='python')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "frDNA_paf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_paf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"min len of match for frDNA is\", frDNA_paf[1].min())\n",
    "print(\"min len of match for EF1a is\", EF1a_paf[1].min())\n",
    "print(\"min len of match for combined is\", combined_paf[1].min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('num matches with unique ids for frDNA is', len(frDNA_paf[0].unique()))\n",
    "print('num matches with unique ids for EF1a is', len(EF1a_paf[0].unique()))\n",
    "print('num matches with unique ids for combined is', len(combined_paf[0].unique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare above (using minimap2) with BLAST approach\n",
    "    - BLAST may be too slow on a larger dataset\n",
    "\n",
    "\n",
    "Check with other alignment programs eg. lastz, BLAT (check for others)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing of full size-clipped files for alignment via minimap2\n",
    " - For each of the frDNA_clipped and EF1a_clipped files as created above \n",
    "     - Look for the number of unique ids in the resultant file\n",
    "     - Determine the percentage of reads in this range that match homology given total number of reads in the clipped.fastq file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "frDNA_clipped_paf = pd.read_csv(\"../../analysis/Python_Processing/20171103_FAH15473/barcode02/frDNA_clipped_test.paf\", sep='\\t', header=None, engine='python')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "EF1a_clipped_paf = pd.read_csv(\"../../analysis/Python_Processing/20171103_FAH15473/barcode02/EF1a_clipped_test.paf\", sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Total number of matches found for frDNA is', len(frDNA_clipped_paf[0].unique()))\n",
    "print('Percentage of matches in region =', \"{:.3%}\".format((len(frDNA_clipped_paf[0].unique())/167054)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Total number of matches found for EF1a is', len(EF1a_clipped_paf[0].unique()))\n",
    "print('False positive percentage =', \"{:.3%}\".format((len(EF1a_clipped_paf[0].unique())/192712)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Total number of matches found for the total of reads is', len(combined_paf[0].unique()))\n",
    "print('Percentage of matches overall =', \"{:.3%}\".format((len(combined_paf[0].unique())/413127)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "~~Plot distribution of matching reads as above~~\n",
    "\n",
    "~~Repeat minimap2 for above to see if fluctuations~~\n",
    "\n",
    "Explore non-mapping 25%\n",
    "\n",
    "Scaling - plots saved out, file for statistics (loop over later)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combined_ids = []\n",
    "# for key in combined_paf[0].unique():\n",
    "#     combined_ids.append(key)\n",
    "combined_dict = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Concatenated/20171103_FAH15473/barcode02/merged.fastq\", \"fastq\"))\n",
    "comb_dict = {}\n",
    "for key in combined_paf[0].unique():\n",
    "    comb_dict[key] = combined_dict[key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lengths = []\n",
    "comb_keys = []\n",
    "for key in comb_dict:\n",
    "    lengths.append(len(comb_dict[key].seq))\n",
    "    comb_keys.append(key)\n",
    "\n",
    "mean = np.mean(lengths)\n",
    "std = np.std(lengths)\n",
    "print(mean)\n",
    "print(std)\n",
    "    \n",
    "# stats_dict = {'number of frDNA reads':len(lengths),'minimum read length':min(lengths),'maximum read length':max(lengths),'mean read length':\"{:.0f}\".format(np.mean(lengths)),'median read length':\"{:.0f}\".format(np.median(lengths))}\n",
    "# stats = pd.DataFrame(stats_dict, index=['20171103_FAH15473/barcode02'])\n",
    "        \n",
    "              \n",
    "# ax = sns.distplot(lengths, color=\"k\", kde=False, bins=5000)\n",
    "# ax.set(xlim=(250, 3500))\n",
    "# ax.set_title(\"frDNA reads for 20171103_FAH15473/barcode02\", fontsize=15)\n",
    "# ax.set_xlabel(\"Length of read\", fontsize=13)\n",
    "# ax.set_ylabel(\"Number of reads\", fontsize=13)\n",
    "# plt.show()\n",
    "\n",
    "\n",
    "# ax = sns.distplot(lengths, color=\"k\", kde=False, bins=5000)\n",
    "# ax.set(xlim=(2400, 3500))\n",
    "# ax.set_title(\"frDNA reads for 20171103_FAH15473/barcode02\", fontsize=15)\n",
    "# ax.set_xlabel(\"Length of read\", fontsize=13)\n",
    "# ax.set_ylabel(\"Number of reads\", fontsize=13)\n",
    "# plt.show()\n",
    "\n",
    "\n",
    "# display(stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fr_dict = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Python_Processing/20171103_FAH15473/barcode02/barcode02_frDNA_clipped.fastq\", 'fastq'))\n",
    "print(len(fr_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# non_dict = {}\n",
    "# for key in fr_dict:\n",
    "#     if key not in frDNA_clipped_paf[0].unique():\n",
    "#         non_dict[key] = fr_dict[key]\n",
    "# print(len(non_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# lengths = []\n",
    "# non_keys = []\n",
    "# for key in non_dict:\n",
    "#     lengths.append(len(non_dict[key].seq))\n",
    "#     non_keys.append(key)\n",
    "\n",
    "# non_dict_stats = {'number of reads':len(lengths),'minimum read length':min(lengths),'maximum read length':max(lengths),'mean read length':\"{:.0f}\".format(np.mean(lengths)),'median read length':\"{:.0f}\".format(np.median(lengths))}\n",
    "# stats = pd.DataFrame(non_dict_stats, index=['20171103_FAH15473/barcode02'])\n",
    "        \n",
    "              \n",
    "# ax = sns.distplot(lengths, color=\"k\", kde=False)\n",
    "# # ax.set(xlim=(250, 3500))\n",
    "# ax.set_title(\"non-frDNA reads for 20171103_FAH15473/barcode02\", fontsize=15)\n",
    "# ax.set_xlabel(\"Length of read\", fontsize=13)\n",
    "# ax.set_ylabel(\"Number of reads\", fontsize=13)\n",
    "# plt.show()\n",
    "# display(stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../summary_statistics.py\n",
    "\n",
    "\"\"\"\n",
    "The goal of this program is to examine the distribution of reads within\n",
    "each file, and for the files generated after homology analysis.\n",
    "The program will generate summary statistics for the result of the homology\n",
    "analysis and save figures illustrating the read distribution for the\n",
    "frDNA reads\n",
    "\"\"\"\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import argparse\n",
    "\n",
    "\n",
    "WHITE='\\033[1;37m'\n",
    "BLUE='\\033[0;34m'\n",
    "RED='\\033[0;31m'\n",
    "NC='\\033[0m'\n",
    "GREEN='\\033[0;32m'\n",
    "PURPLE='\\033[1;35m'\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "The goal of this program is to examine the distribution of reads within\n",
    "each file, and for the files generated after homology analysis.\n",
    "The program will generate summary statistics for the result of the homology\n",
    "analysis and save figures illustrating the read distribution for the\n",
    "frDNA reads\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"full_file\", help=\"The full, unfiltered file containing all reads for this barcode\")\n",
    "parser.add_argument(\"input_folder\", help=\"The destination folder within which the .paf files are generated\")\n",
    "parser.add_argument(\"output_folder\", help=\"The destination folder for any outputs from this script - including summary statistics file and plots\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "print('\\033[0;35m'+'START'+'\\033[1;37m')\n",
    "\n",
    "output_folder = args.output_folder.rsplit('/', 1)[-2]\n",
    "input_folder = args.input_folder.rsplit('/', 1)[-2]\n",
    "if args.verbose:\n",
    "    print('\\033[0;31m' + \"Input folder is \" + input_folder + '\\033[1;37m')\n",
    "    print('\\033[0;31m' + \"Output folder is \" + output_folder + '\\033[1;37m')\n",
    "    print('\\033[0;34m' + \"Loading \" + args.full_file + '\\033[1;37m')\n",
    "\n",
    "# Load the full file containing all reads for this barcode\n",
    "full_file_dict = SeqIO.to_dict(SeqIO.parse(args.full_file, \"fastq\"))\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Loaded \" + args.full_file + '\\033[1;37m')\n",
    "\n",
    "# Extract the information about the lengths of the sequence for each read in this barcode\n",
    "full_lengths = []\n",
    "for key in full_file_dict:\n",
    "    full_lengths.append(len(full_file_dict[key].seq))\n",
    "full_lengths_len = len(full_file_dict)\n",
    "\n",
    "\n",
    "# Plot the spread of read lengths for this barcode\n",
    "    # Expect to see two peaks - one for EF1a and one for frDNA\n",
    "ax = sns.distplot(full_lengths, color=\"k\", kde=False, bins=5000)\n",
    "ax.set(xlim=(250, 3500))\n",
    "ax.set_title(\"Read spread for %s\" % '/'.join(args.full_file.rsplit('/')[-3:-1]), fontsize=15)\n",
    "ax.set_xlabel(\"Length of read\", fontsize=13)\n",
    "ax.set_ylabel(\"Number of reads\", fontsize=13)\n",
    "figure1 = ax.get_figure()\n",
    "# Save this figure out\n",
    "figure1.savefig('/'.join([output_folder, 'full_read_spread.png']))\n",
    "figure1.clf()\n",
    "if args.verbose:\n",
    "    print('\\033[0;32m' + \"Full spread image file saved to \" + '/'.join([output_folder, 'full_read_spread.png']) + '\\033[1;37m')\n",
    "    print('\\033[0;34m' + \"Loading \" + input_folder+\"/combined_test.paf\" + '\\033[1;37m')\n",
    "\n",
    "# Import the PAF file resulting from the minimap2 homology filtering\n",
    "full_paf = pd.read_csv(input_folder+\"/combined_test.paf\", sep='\\t', header=None, engine='python')\n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Loaded \" + input_folder+\"/combined_test.paf\" + '\\033[1;37m')\n",
    "# Determine all the read ids present within the homology-filtered dataset\n",
    "# Then, create a dictionary extracting all the information from the full read file, but ONLY for reads present within the homology-filtered data\n",
    "full_dict = {}\n",
    "for key in full_paf[0].unique():\n",
    "    full_dict[key] = full_file_dict[key]\n",
    "\n",
    "# For each key in the homology-filtered dictionary, extract the sequence length and key\n",
    "full_paf_lengths = []\n",
    "full_keys = []\n",
    "for key in full_dict:\n",
    "    full_paf_lengths.append(len(full_dict[key].seq))\n",
    "    full_keys.append(key)\n",
    "\n",
    "mean = np.mean(full_paf_lengths)\n",
    "std = np.std(full_paf_lengths)\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[1;33m' + 'Mean read length is %s' % mean + '\\033[1;37m')\n",
    "    print('\\033[1;33m' + 'Standard deviation of read length is %s' % std + '\\033[1;37m')\n",
    "    \n",
    "    \n",
    "length_filt_dict = full_dict.copy()\n",
    "for key in full_keys:\n",
    "    if len(full_dict[key].seq) < (mean-1.645*std) or len(full_dict[key].seq) > (mean+1.645*std):\n",
    "        del length_filt_dict[key]\n",
    "\n",
    "        \n",
    "        \n",
    "SeqIO.write(length_filt_dict.values(), '/'.join([output_folder, 'length_restricted_reads.fasta']), \"fasta\")\n",
    "if args.verbose:\n",
    "    print('\\033[1;36m' + 'Saved %s' % ('/'.join([output_folder, 'length_restricted_reads.fasta'])) + '\\033[1;37m')     \n",
    "    \n",
    "    \n",
    "    \n",
    "length_filt_lens = []\n",
    "len_filt_keys = []\n",
    "for key in length_filt_dict:\n",
    "    length_filt_lens.append(len(length_filt_dict[key].seq))\n",
    "    len_filt_keys.append(key)\n",
    "\n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "    \n",
    "Extract the qscores\n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Loading \" + 'Basecalled/'+'/'.join(args.full_file.rsplit('/')[-3:-1])+'/sequencing_summary.txt' + '\\033[1;37m')\n",
    "summ_stats_csv = pd.read_csv('Basecalled/'+'/'.join(args.full_file.rsplit('/')[-3:-1])+'/sequencing_summary.txt', sep='\\t', header=None, names=['filename', 'read_id', 'run_id', 'batch_id', 'channel', 'mux', 'start_time', 'duration', 'num_events', 'passes_filtering', 'template_start', 'num_events_template', 'template_duration', 'sequence_length_template', 'mean_qscore_template', 'strand_score_template', 'median_template', 'mad_template'], engine='python')\n",
    "summ_stats_csv = pd.DataFrame(summ_stats_csv[1:])\n",
    "summary_list = []\n",
    "for column, row in summ_stats_csv.iterrows():\n",
    "    if row['read_id'] in full_keys:\n",
    "        summary_list.append([row['read_id'], row['mean_qscore_template']])\n",
    "summary_frame = pd.DataFrame(summary_list)\n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Finished with \" + 'Basecalled/'+'/'.join(args.full_file.rsplit('/')[-3:-1])+'/sequencing_summary.txt' + '\\033[1;37m')\n",
    "    \n",
    "Create a dictionary containing the statistics for the filtered dataset\n",
    "    Total no. frDNA reads, Min. read length, Max. read length, Mean read length, Median read length, Quality score\n",
    "\n",
    "stats_dict = {'number of frDNA reads':len(length_filt_lens),'minimum read length':min(length_filt_lens),'maximum read length':max(length_filt_lens),'mean read length':\"{:.0f}\".format(np.mean(length_filt_lens)),'std dev':\"{:.0f}\".format(np.std(length_filt_lens)),'median read length':\"{:.0f}\".format(np.median(length_filt_lens))\n",
    "              ,'min_qscore':\"{:.2f}\".format(min(summary_frame[1].astype(float))), 'max_qscore':\"{:.2f}\".format(max(summary_frame[1].astype(float))), 'mean_qscore':\"{:.2f}\".format(np.mean(summary_frame[1].astype(float))), 'median_qscore':\"{:.2f}\".format(np.median(summary_frame[1].astype(float)))\n",
    "             }\n",
    "stats = pd.DataFrame(stats_dict, index=['%s' % '/'.join(args.full_file.rsplit('/')[-3:-1])])    \n",
    "              \n",
    "bx = sns.distplot(length_filt_lens, color=\"k\", kde=False)\n",
    "bx.set(xlim=(250, 3500))\n",
    "bx.set_title(\"frDNA reads for %s\" % '/'.join(args.full_file.rsplit('/')[-3:-1]), fontsize=15)\n",
    "bx.set_xlabel(\"Length of read\", fontsize=13)\n",
    "bx.set_ylabel(\"Number of reads\", fontsize=13)\n",
    "figure2 = bx.get_figure()\n",
    "figure2.savefig('/'.join([output_folder, 'frDNA_len_filt_full.png']))\n",
    "figure2.clf()\n",
    "if args.verbose:\n",
    "    print('\\033[0;32m' + \"frDNA spread image file saved to \" + '/'.join([output_folder, 'frDNA_len_filt_full.png']) + '\\033[1;37m')\n",
    "\n",
    "cx = sns.distplot(length_filt_lens, color=\"k\", kde=False)\n",
    "cx.set(xlim=((mean-1.645*std)-100, (mean+1.645*std)+100))\n",
    "cx.set_title(\"frDNA reads for %s\" % '/'.join(args.full_file.rsplit('/')[-3:-1]), fontsize=15)\n",
    "cx.set_xlabel(\"Length of read\", fontsize=13)\n",
    "cx.set_ylabel(\"Number of reads\", fontsize=13)\n",
    "figure3 = cx.get_figure()\n",
    "figure3.savefig('/'.join([output_folder, 'frDNA_len_filt_limited.png']))\n",
    "figure3.clf()\n",
    "if args.verbose:\n",
    "    print('\\033[0;32m' + \"Zoomed-in frDNA spread image file saved to \" + '/'.join([output_folder, 'frDNA_len_filt_limited.png']) + '\\033[1;37m')\n",
    "\n",
    "stats.to_csv('/'.join([output_folder, 'frDNA_len_filt_statistics.csv']), index=False)\n",
    "if args.verbose:\n",
    "    print('\\033[0;32m' + \"Summary statistics file saved to \" + '/'.join([output_folder, 'frDNA_len_filt_statistics.csv']) + '\\033[1;37m')\n",
    "    \n",
    "print('\\033[0;35m'+'END'+'\\033[1;37m')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "path = \"/media/MassStorage/tmp/TE/honours/analysis/Stats/*/*/frDNA_statistics.csv\"\n",
    "path_names = glob.glob(path)\n",
    "full_stats = pd.DataFrame(data=None, columns = ['number of frDNA reads','minimum read length','maximum read length','mean read length','standard deviation','median read length','min_qscore','max_qscore','mean_qscore','median_qscore'])\n",
    "for path in path_names:\n",
    "    if path[67:-21] != 'unclassified':\n",
    "        if '20171212_FAH18688/barcode10' not in path and '20171207_FAH18654/barcode10' not in path:\n",
    "            path_stats_csv = pd.read_csv(path, header=0)\n",
    "            print(path_stats_csv)\n",
    "            full_stats = full_stats.append(path_stats_csv)\n",
    "            full_stats = full_stats.rename(index={0: path[49:-21]})\n",
    "full_stats = full_stats.sort_index(ascending=True)\n",
    "full_stats.to_csv('../../analysis/Stats/overall_frDNA_stats.csv')\n",
    "num_read_stats = pd.DataFrame(data=[\"{:.0f}\".format(min(full_stats['number of frDNA reads'])), \"{:.0f}\".format(max(full_stats['number of frDNA reads'])), \"{:.0f}\".format(np.mean(full_stats['number of frDNA reads'])), \"{:.0f}\".format(np.median(full_stats['number of frDNA reads']))], index=['Min', 'Max', 'Mean', 'Median'], columns=['Number of reads'])\n",
    "num_read_stats.to_csv('../../analysis/Stats/number_of_frDNA_reads_summary.csv')\n",
    "full_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(sum(full_stats['number of frDNA reads']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "path = \"/10tb/tmp/TE/honours/analysis/Length_Filtered/*/*/*.csv\"\n",
    "path_names = glob.glob(path)\n",
    "full_stats2 = pd.DataFrame(data=None, columns = ['number of frDNA reads','minimum read length','maximum read length','mean read length','standard deviation','median read length'])\n",
    "for path in path_names:\n",
    "    if path[64:-30] != 'unclassified':\n",
    "        if '20171212_FAH18688/barcode10' not in path and '20171207_FAH18654/barcode10' not in path:\n",
    "            path_stats_csv = pd.read_csv(path, header=0)\n",
    "            full_stats2 = full_stats2.append(path_stats_csv)\n",
    "            full_stats2 = full_stats2.rename(index={0: path[46:-30]})\n",
    "full_stats2 = full_stats2.sort_index(ascending=True)\n",
    "display(full_stats2)\n",
    "full_stats2.to_csv('../../analysis/Length_Filtered/overall_frDNA_stats.csv')\n",
    "num_read_stats = pd.DataFrame(data=[\"{:.0f}\".format(min(full_stats2['number of frDNA reads'])), \"{:.0f}\".format(max(full_stats2['number of frDNA reads'])), \"{:.0f}\".format(np.mean(full_stats2['number of frDNA reads'])), \"{:.0f}\".format(np.median(full_stats2['number of frDNA reads']))], index=['Min', 'Max', 'Mean', 'Median'], columns=['Number of reads'])\n",
    "num_read_stats.to_csv('../../analysis/Length_Filtered/number_of_frDNA_reads_summary.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create matrix for loss-of-reads when performing length filtering as percentage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_stats_1 = []\n",
    "full_stats_2 = []\n",
    "result = []\n",
    "\n",
    "for i in full_stats['number of frDNA reads']:\n",
    "    full_stats_1.append(i)\n",
    "for i in full_stats2['number of frDNA reads']:\n",
    "    full_stats_2.append(i)\n",
    "for i in range (0,len(full_stats_1)):\n",
    "    result.append(float(\"{:.2f}\".format((full_stats_1[i]-full_stats_2[i])/full_stats_1[i])))\n",
    "print(result)\n",
    "print('The maximum loss of reads is %s%% in %s' % (100*max(result), path_names[result.index(max(result))][46:-30]))\n",
    "print('The minimum loss of reads is %s%% in %s' % (100*min(result), path_names[result.index(min(result))][46:-30]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FAH18688_barcode10 = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Concatenated/20171212_FAH18688/barcode10/merged.fastq\", 'fastq'))\n",
    "lengths = []\n",
    "b10_keys = []\n",
    "for key in FAH18688_barcode10:\n",
    "    lengths.append(len(FAH18688_barcode10[key].seq))\n",
    "    b10_keys.append(key)\n",
    "\n",
    "non_dict_stats = {'number of reads':len(lengths),'minimum read length':min(lengths),'maximum read length':max(lengths),'mean read length':\"{:.0f}\".format(np.mean(lengths)),'median read length':\"{:.0f}\".format(np.median(lengths))}\n",
    "stats = pd.DataFrame(non_dict_stats, index=['20171212_FAH18688/barcode10'])\n",
    "              \n",
    "ax = sns.distplot(lengths, color=\"k\", kde=False)\n",
    "# ax.set(xlim=(250, 3500))\n",
    "ax.set_title(\"Miscalled read spread for 20171212_FAH18688/barcode10\", fontsize=15)\n",
    "ax.set_xlabel(\"Length of read\", fontsize=13)\n",
    "ax.set_ylabel(\"Number of reads\", fontsize=13)\n",
    "plt.show()\n",
    "display(stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FAH18654_barcode10 = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Concatenated/20171207_FAH18654/barcode10/merged.fastq\", 'fastq'))\n",
    "lengths = []\n",
    "b10_keys = []\n",
    "for key in FAH18654_barcode10:\n",
    "    lengths.append(len(FAH18654_barcode10[key].seq))\n",
    "    b10_keys.append(key)\n",
    "\n",
    "non_dict_stats = {'number of reads':len(lengths),'minimum read length':min(lengths),'maximum read length':max(lengths),'mean read length':\"{:.0f}\".format(np.mean(lengths)),'median read length':\"{:.0f}\".format(np.median(lengths))}\n",
    "stats = pd.DataFrame(non_dict_stats, index=['20171207_FAH18654/barcode10'])\n",
    "              \n",
    "ax = sns.distplot(lengths, color=\"k\", kde=False)\n",
    "# ax.set(xlim=(250, 3500))\n",
    "ax.set_title(\"Miscalled read spread for 20171207_FAH18654/barcode10\", fontsize=15)\n",
    "ax.set_xlabel(\"Length of read\", fontsize=13)\n",
    "ax.set_ylabel(\"Number of reads\", fontsize=13)\n",
    "plt.show()\n",
    "display(stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode02 = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Length_Filtered/20171103_FAH15473/barcode02/length_restricted_reads.fasta\", \"fasta\"))\n",
    "barcode06 = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Length_Filtered/20171103_FAH15473/barcode06/length_restricted_reads.fasta\", \"fasta\"))\n",
    "\n",
    "total_lens = []\n",
    "for key in barcode02:\n",
    "    total_lens.append(len(barcode02[key].seq))\n",
    "for key in barcode06:\n",
    "    total_lens.append(len(barcode06[key].seq))\n",
    "print(max(total_lens))\n",
    "print(min(total_lens))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode02_numbers = {}\n",
    "for key in barcode02:\n",
    "    seq = []\n",
    "    for element in barcode02[key].seq[30:-30]:\n",
    "        if element == \"A\":\n",
    "            seq.append(0)\n",
    "        elif element == \"C\":\n",
    "            seq.append(1)\n",
    "        elif element == \"G\":\n",
    "            seq.append(2)\n",
    "        elif element == \"T\":\n",
    "            seq.append(3)\n",
    "    if len(seq) < max(total_lens):\n",
    "        seq.extend([0]*(max(total_lens)-len(seq)))\n",
    "    barcode02_numbers[key] = seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq2 = np.array(random.choices(list(barcode02_numbers.values()),k=10000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode06_numbers = {}\n",
    "for key in barcode06:\n",
    "    seq = []\n",
    "    for element in barcode06[key].seq[30:-30]:\n",
    "        if element == \"A\":\n",
    "            seq.append(0)\n",
    "        elif element == \"C\":\n",
    "            seq.append(1)\n",
    "        elif element == \"G\":\n",
    "            seq.append(2)\n",
    "        elif element == \"T\":\n",
    "            seq.append(3)\n",
    "    if len(seq) < max(total_lens):\n",
    "        seq.extend([0]*(max(total_lens)-len(seq)))\n",
    "    barcode06_numbers[key] = seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq6 = np.array(random.choices(list(barcode06_numbers.values()),k=10000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_comb = np.concatenate((seq2, seq6), axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids2 = np.array([2]*(len(seq2)))\n",
    "print(len([2]*(len(seq2))))\n",
    "ids6 = np.array([6]*(len(seq6)))\n",
    "print(len([2]*(len(seq6))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids_comb = np.concatenate((ids2, ids6), axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(seq_comb))\n",
    "print(len(ids_comb))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savez_compressed('../../analysis/arrays_test/20171103_FAH15473_b2+b6_ids.csv', ids_comb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_comb[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savez_compressed('../../analysis/arrays_test/20171103_FAH15473_b2+b6_seqs.csv', seq_comb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs_test = np.load('../../analysis/arrays_test/20171103_FAH15473_b2+b6_seqs.csv.npz', allow_pickle=True)['arr_0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs_test[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids_test = np.load('../../analysis/arrays_test/20171103_FAH15473_b2+b6_ids.csv.npz', allow_pickle=True)['arr_0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids_test[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids_test[-10:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "path = \"/10tb/tmp/TE/honours/analysis/Concatenated/*/*/*.fastq\"\n",
    "path_names = glob.glob(path)\n",
    "total_count = 0\n",
    "can_count = 0\n",
    "mis_count = 0\n",
    "unc_count = 0\n",
    "for path in path_names:\n",
    "    temp_dict = SeqIO.to_dict(SeqIO.parse(path, \"fastq\"))\n",
    "    if path[61:-13] != 'unclassified':\n",
    "        if '20171212_FAH18688/barcode10' not in path and '20171207_FAH18654/barcode10' not in path:\n",
    "            total_count += len(temp_dict)\n",
    "            can_count += len(temp_dict)\n",
    "        else:\n",
    "            total_count += len(temp_dict)\n",
    "            mis_count += len(temp_dict)\n",
    "            can_count += len(temp_dict)\n",
    "    else:\n",
    "        total_count += len(temp_dict)\n",
    "        unc_count += len(temp_dict)\n",
    "print('Total number of reads assigned to a barcode by Deepbinner is %s' % can_count)\n",
    "print('Number of reads assigned to non-existent barcodes by Deepbinner is %s' % mis_count)\n",
    "print('Estimated number of misassigned reads in total is %s' % (23*mis_count))\n",
    "print('Percentage of misassigned reads based on estimated number is %s' % (2300*(mis_count)/can_count))\n",
    "print('Number of unclassified reads is %s' % unc_count)\n",
    "print('Percentage of unclassified reads is %s' % (100*unc_count/total_count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "path = \"/10tb/tmp/TE/honours/analysis/Length_Filtered/*/*/*.csv\"\n",
    "path_names = glob.glob(path)\n",
    "full_stats2 = pd.DataFrame(data=None, columns = ['number of frDNA reads','minimum read length','maximum read length','mean read length','std dev','median read length'])\n",
    "for path in path_names:\n",
    "    if path[64:-30] != 'unclassified':\n",
    "        if '20171212_FAH18688/barcode10' not in path and '20171207_FAH18654/barcode10' not in path:\n",
    "            path_stats_csv = pd.read_csv(path, header=0)\n",
    "            full_stats2 = full_stats2.append(path_stats_csv)\n",
    "            full_stats2 = full_stats2.rename(index={0: path[46:-30]})\n",
    "full_stats2 = full_stats2.sort_values('mean read length', ascending=False)\n",
    "\n",
    "sns.set(style=\"whitegrid\")\n",
    "fig, ax = plt.subplots(figsize=(10,17))\n",
    "ax = sns.barplot(x=full_stats2['mean read length'], y=full_stats2.index, palette=\"Blues_d\", xerr=full_stats2['std dev'],ci=True)\n",
    "ax.set_title(\"Mean Read Length by Sample\")\n",
    "figure1 = ax.get_figure()\n",
    "figure1.savefig(\"/10tb/tmp/TE/honours/analysis/Length_Filtered/mean_reads.png\",bbox_inches = \"tight\")\n",
    "figure1.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "path = \"/10tb/tmp/TE/honours/analysis/Length_Filtered/*/*/*.csv\"\n",
    "path_names = glob.glob(path)\n",
    "full_stats2 = pd.DataFrame(data=None, columns = ['number of frDNA reads','minimum read length','maximum read length','mean read length','std dev','median read length'])\n",
    "for path in path_names:\n",
    "    if path[64:-30] != 'unclassified':\n",
    "        if '20171212_FAH18688/barcode10' not in path and '20171207_FAH18654/barcode10' not in path:\n",
    "            path_stats_csv = pd.read_csv(path, header=0)\n",
    "            full_stats2 = full_stats2.append(path_stats_csv)\n",
    "            full_stats2 = full_stats2.rename(index={0: path[46:-30]})\n",
    "full_stats2 = full_stats2.sort_values('number of frDNA reads', ascending=False)\n",
    "\n",
    "sns.set(style=\"whitegrid\")\n",
    "fig, ax = plt.subplots(figsize=(10,17))\n",
    "ax = sns.barplot(x=full_stats2['number of frDNA reads'], y=full_stats2.index, palette=\"Blues_d\")\n",
    "ax.set_title(\"Number of Reads per Sample after Filtering\")\n",
    "figure1 = ax.get_figure()\n",
    "figure1.savefig(\"/10tb/tmp/TE/honours/analysis/Length_Filtered/num_reads.png\",bbox_inches = \"tight\")\n",
    "figure1.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Identify primers and orientation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = Mothur()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "m.help()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "m.pcr.seqs(fasta=\"../../analysis/Length_Filtered/20171103_FAH15473/barcode02/length_restricted_reads.fasta\",oligos=\"../../analysis/ITS_primers.oligos\",pdiffs=0,rdiffs=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pcr_dict = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Length_Filtered/20171103_FAH15473/barcode02/length_restricted_reads.pcr.fasta\",\"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = []\n",
    "\n",
    "for key in pcr_dict:\n",
    "    ids.append(key)\n",
    "with open('../../analysis/Length_Filtered/20171103_FAH15473/barcode02/ids.txt','w') as handle:\n",
    "    handle.writelines(\"%s\\n\" % name for name in ids)\n",
    "print(len(ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp_dict = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Length_Filtered/20171103_FAH15473/barcode02/length_restricted_reads.fasta\",\"fasta\"))\n",
    "new_dict = tmp_dict.copy()\n",
    "keys_list = random.sample(ids,k=200)\n",
    "print(len(keys_list))\n",
    "for key in new_dict:\n",
    "    if key not in keys_list:\n",
    "        del tmp_dict[key]\n",
    "print(len(tmp_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../get_cons_ids.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from mothur_py import Mothur\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "WHITE='\\033[1;37m'\n",
    "BLUE='\\033[0;34m'\n",
    "RED='\\033[0;31m'\n",
    "NC='\\033[0m'\n",
    "GREEN='\\033[0;32m'\n",
    "PURPLE='\\033[1;35m'\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "The goal of this program is to extract the read ids of reads that\n",
    "contain both the forward and reverse primer as exact matches\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;31m' + \"Input file is \" + args.input_file + '\\033[1;37m')\n",
    "\n",
    "m = Mothur()\n",
    "m.pcr.seqs(fasta=args.input_file,oligos='ITS_primers.oligos',pdiffs=2,rdiffs=2)\n",
    "pcr_dict = SeqIO.to_dict(SeqIO.parse(args.input_file[:-5]+\"pcr.fasta\",\"fasta\"))\n",
    "ids = []\n",
    "for key in pcr_dict:\n",
    "    ids.append(key)\n",
    "with open(args.input_file[:-29]+'ids.txt','w') as handle:\n",
    "    handle.writelines(\"%s\\n\" % name for name in ids)\n",
    "    \n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Ids file saved to \" + '\\033[0;35m' + (args.input_file[:-29]+'ids.txt') + '\\033[1;37m')\n",
    "    print('\\033[0;32m' + (\"The number of reads is %s\" % len(ids)) + '\\033[1;37m')\n",
    "    \n",
    "    \n",
    "tmp_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "new_dict = tmp_dict.copy()\n",
    "if len(ids) > 500:\n",
    "    keys_list = random.sample(ids,k=500)\n",
    "else:\n",
    "    print('\\033[1;37m' + \"LOW READS\")\n",
    "for key in new_dict:\n",
    "    if key not in keys_list:\n",
    "        del tmp_dict[key]\n",
    "SeqIO.write(tmp_dict.values(),('Consensus'+args.input_file[15:-29]+'for_consensus_500.fasta'),'fasta')\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Ids file saved to \" + '\\033[0;35m' + ('Consensus'+args.input_file[15:-29]+'for_consensus_500.fasta') + '\\033[1;37m')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting ../get_cons_ids3.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile ../get_cons_ids3.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from mothur_py import Mothur\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "WHITE='\\033[1;37m'\n",
    "BLUE='\\033[0;34m'\n",
    "RED='\\033[0;31m'\n",
    "NC='\\033[0m'\n",
    "GREEN='\\033[0;32m'\n",
    "PURPLE='\\033[1;35m'\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "The goal of this program is to extract the read ids of reads that\n",
    "contain both the forward and reverse primer as exact matches\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;31m' + \"Input file is \" + args.input_file + '\\033[1;37m')\n",
    "\n",
    "m = Mothur()\n",
    "m.pcr.seqs(fasta=args.input_file,oligos='ITS_primers.oligos',pdiffs=0,rdiffs=0)\n",
    "pcr_dict = SeqIO.to_dict(SeqIO.parse(args.input_file[:-5]+\"pcr.fasta\",\"fasta\"))\n",
    "ids = []\n",
    "for key in pcr_dict:\n",
    "    ids.append(key)\n",
    "with open(args.input_file[:-29]+'ids.txt','w') as handle:\n",
    "    handle.writelines(\"%s\\n\" % name for name in ids)\n",
    "    \n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Ids file saved to \" + '\\033[0;35m' + (args.input_file[:-29]+'ids.txt') + '\\033[1;37m')\n",
    "    print('\\033[0;32m' + (\"The number of reads is %s\" % len(ids)) + '\\033[1;37m')\n",
    "    \n",
    "    \n",
    "tmp_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "new_dict = tmp_dict.copy()\n",
    "if len(ids) > 100:\n",
    "    keys_list = random.sample(ids,k=100)\n",
    "else:\n",
    "    print('\\033[1;37m' + \"LOW READS\")\n",
    "for key in new_dict:\n",
    "    if key not in keys_list:\n",
    "        del tmp_dict[key]\n",
    "SeqIO.write(tmp_dict.values(),('Consensus'+args.input_file[15:-29]+'for_consensus.fasta'),'fasta')\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Ids file saved to \" + '\\033[0;35m' + ('Consensus'+args.input_file[15:-29]+'for_consensus.fasta') + '\\033[1;37m')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting ../get_cons_ids2.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile ../get_cons_ids2.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from mothur_py import Mothur\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "WHITE='\\033[1;37m'\n",
    "BLUE='\\033[0;34m'\n",
    "RED='\\033[0;31m'\n",
    "NC='\\033[0m'\n",
    "GREEN='\\033[0;32m'\n",
    "PURPLE='\\033[1;35m'\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "The goal of this program is to extract the read ids of reads that\n",
    "contain both the forward and reverse primer as exact matches\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;31m' + \"Input file is \" + args.input_file + '\\033[1;37m')\n",
    "\n",
    "m = Mothur()\n",
    "m.pcr.seqs(fasta=args.input_file,oligos='ITS_primers.oligos',pdiffs=2,rdiffs=2)\n",
    "pcr_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "ids = []\n",
    "for key in pcr_dict:\n",
    "    ids.append(key)\n",
    "with open(args.input_file[:-29]+'ids2.txt','w') as handle:\n",
    "    handle.writelines(\"%s\\n\" % name for name in ids)\n",
    "    \n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Ids file saved to \" + '\\033[0;35m' + (args.input_file[:-29]+'ids2.txt') + '\\033[1;37m')\n",
    "    print('\\033[0;32m' + (\"The number of reads is %s\" % len(ids)) + '\\033[1;37m')\n",
    "    \n",
    "    \n",
    "tmp_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "new_dict = tmp_dict.copy()\n",
    "if len(ids) > 1000:\n",
    "    keys_list = random.sample(ids,k=1000)\n",
    "else:\n",
    "    print('\\033[1;37m' + \"LOW READS\")\n",
    "for key in new_dict:\n",
    "    if key not in keys_list:\n",
    "        del tmp_dict[key]\n",
    "SeqIO.write(tmp_dict.values(),('Consensus'+args.input_file[15:-29]+'for_consensus_1000.fasta'),'fasta')\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Ids file saved to \" + '\\033[0;35m' + ('Consensus'+args.input_file[15:-29]+'for_consensus_1000.fasta') + '\\033[1;37m')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../get_final_cons.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from mothur_py import Mothur\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "WHITE='\\033[1;37m'\n",
    "BLUE='\\033[0;34m'\n",
    "RED='\\033[0;31m'\n",
    "NC='\\033[0m'\n",
    "GREEN='\\033[0;32m'\n",
    "PURPLE='\\033[1;35m'\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "The goal of this program is to extract the read ids of reads that\n",
    "contain both the forward and reverse primer as exact matches\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;31m' + \"Input file is \" + args.input_file + '\\033[1;37m')\n",
    "\n",
    "m = Mothur()\n",
    "m.pcr.seqs(fasta=args.input_file,oligos='ITS_primers.oligos',pdiffs=2,rdiffs=2)\n",
    "pcr_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "ids = []\n",
    "for key in pcr_dict:\n",
    "    ids.append(key)\n",
    "with open(args.input_file[:-29]+'ids2.txt','w') as handle:\n",
    "    handle.writelines(\"%s\\n\" % name for name in ids)\n",
    "    \n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Ids file saved to \" + '\\033[0;35m' + (args.input_file[:-29]+'ids2.txt') + '\\033[1;37m')\n",
    "    print('\\033[0;32m' + (\"The number of reads is %s\" % len(ids)) + '\\033[1;37m')\n",
    "    \n",
    "    \n",
    "tmp_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "new_dict = tmp_dict.copy()\n",
    "if len(ids) > 100:\n",
    "    keys_list = random.sample(ids,k=100)\n",
    "else:\n",
    "    print('\\033[1;37m' + \"LOW READS\")\n",
    "for key in new_dict:\n",
    "    if key not in keys_list:\n",
    "        del tmp_dict[key]\n",
    "SeqIO.write(tmp_dict.values(),('Consensus'+args.input_file[15:-29]+'final_cons.fasta'),'fasta')\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Ids file saved to \" + '\\033[0;35m' + ('Consensus'+args.input_file[15:-29]+'final_cons.fasta') + '\\033[1;37m')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Align import AlignInfo\n",
    "alignment = Bio.AlignIO.read(\"../../analysis/Consensus/20171103_FAH15473/barcode02/consensus_100.fasta\",\"fasta\")\n",
    "summary_align = AlignInfo.SummaryInfo(alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus = summary_align.dumb_consensus(threshold=0.7, ambiguous='N')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print(consensus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "database = SeqIO.to_dict(SeqIO.parse(\"../../database/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta\", \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lens = []\n",
    "for key in database:\n",
    "    lens.append(len(database[key].seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(max(lens))\n",
    "print(min(lens))\n",
    "print(np.mean(lens))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"The number of reads in this file is\", len(lens))\n",
    "ax = sns.distplot(lens, color=\"k\", kde=False)\n",
    "\n",
    "\n",
    "# ax.set(xlim=(250, 3500))\n",
    "ax.set_title(\"Reads spread 20171103_FAH15473/barcode02\", fontsize=15)\n",
    "ax.set_xlabel(\"Length of read\", fontsize=13)\n",
    "ax.set_ylabel(\"Number of reads\", fontsize=13)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len([1 for i in lens if i > 1000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len([1 for i in lens if i > 2000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len([1 for i in lens if i > 3000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len([1 for i in lens if i > 500])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len([1 for i in lens if i < 500])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../get_align_seqs.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "WHITE='\\033[1;37m'\n",
    "BLUE='\\033[0;34m'\n",
    "RED='\\033[0;31m'\n",
    "NC='\\033[0m'\n",
    "GREEN='\\033[0;32m'\n",
    "PURPLE='\\033[1;35m'\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program extracts a specified number of reads from a fasta\n",
    "file and saves them to a new fasta file\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "parser.add_argument(\"num_reads\", help=\"The number of reads to extract\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;31m' + \"Input file is \" + args.input_file + '\\033[1;37m')\n",
    "    print('\\033[0;31m' + \"The number of extracted reads is \" + '\\033[0;32m' + args.num_reads + '\\033[1;37m')\n",
    "    \n",
    "tmp_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "new_dict = tmp_dict.copy()\n",
    "\n",
    "ids=[]\n",
    "for key in tmp_dict:\n",
    "    ids.append(key)\n",
    "    \n",
    "keys_list = random.sample(ids,k=int(args.num_reads))\n",
    "\n",
    "for key in new_dict:\n",
    "    if key not in keys_list:\n",
    "        del tmp_dict[key]\n",
    "SeqIO.write(tmp_dict.values(),(args.input_file[:9]+'Alignment'+args.input_file[24:-29]+ args.num_reads + '_reads.fasta'),'fasta')\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Ids file saved to \" + '\\033[0;35m' + (args.input_file[:9]+'Alignment'+args.input_file[24:-29]+ args.num_reads + '_reads.fasta') + '\\033[1;37m')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "emboss_test = \"\"\"nnnnnnnnnnnnnnnnnnnnnnnnnnnnnnAGnnnnnnnnnnnnnGTTAACnACnAnnAA\n",
    "GACAnCnCGACnnnAACnnTTTCTTnCAGnnnCACCTGCAAGTCTGGTGCCAGCAGCCnG\n",
    "CnnnnnnGnnTAAnTnTCCAGCTCnnCAAnTAnnnGCnGTATAnTTnnnAAAGnTTGnnT\n",
    "TGAnnCGTTnnAAAAAGCnTCnnnGTnAGnnnTCGAnnAnCTTnnnnCGGCnnCTnCTGG\n",
    "CnnnnnAGTTnnGGTCnnnCGCnnnnCTnTTnTGnnGTGnTnGTAnnnnCTGnATTTGTT\n",
    "GnnnGAGnGCTTnnACCTnCTnnnnTnnGGTnGAACTTnCnnnAATGCACnTTTnnACnT\n",
    "GGGnnTGTTnGGAnnGGGAACCnnAGGATTTTnnnnnnnnTnnAAAAATTAnGnAGTGTT\n",
    "CnnAAAGCAnnGnnGCTTAnTGCCTnnnGAAnnTAnnnCATTAGCnATnnGGAAnnTnnn\n",
    "AATnnAAAATnnAGGACnGnnnnnnnTGnnnTGAnnnTTCTAnTTTnnnnTGnTTGGnTT\n",
    "TCTnnAGGATnnTACCnnnnnGTAnATGATnnnnGAATnnnnnAGGGTnnCAGnnTTGnn\n",
    "nnGGGnnnnGCAnnTTTnnGTAnnnnTnnTACATCnGTnCAGAGnnnGTnnGnAAATTnn\n",
    "CTTnnnnnGGnATTGATnGnnTAAGnnAnCAAACTACnnnTGCGAAnnnnnGCATCnTGC\n",
    "CAAnnGnGATGACTTnnCATTGATCnnAAGAACGnnAAGnGTTAAGnnnGnGTTCAAAAA\n",
    "nnCGATnnCnnnAGATnnAnnnCCnnGnnnTTGnTnnnnAGTnnCTTAACnnnAGTnnAA\n",
    "ACTnnAnnnTGCnnCGACnnTGnnGGGATCAGnACnnnAAGGATTTATnAnnnATGnnAC\n",
    "TTGTTTGnnGCACCCAAnnnnnnnnnGGAAACCTnnnGAAGnnnnTTTAnGGnTTCGTGG\n",
    "nnnnGGnnAGTACGGTnCACnnAAGnGCnnnnnnnTGAAACnTTnnAAAGGAATTGAnCG\n",
    "GAAnnnGGGCACCnnnACCAnnGGTnGTGnnGAGCnnnnCTGCannGCTTnAATTnnTGA\n",
    "nCTCAACACnnnnnGGGAAnACTnnnCAnCCAnGGnnnTCCAGAnnnnnnCACnAGTnnn\n",
    "AAGGATnTGAnnnCAnGAnnnnTTGnnATAnGCnnTTTTTnCTTnGAnTTTTGnnTGGnn\n",
    "TTGnnnnGTGnGTGCAnTGnnnnGCnnCGnnnTTCnnnnTTAnnGnTTGGTGnGnnAGTG\n",
    "nnATTnnTGTCnnnnTnGGnnTTAATTnCCGATAnnnACnnGAACnnnGAGAnCnnCTTC\n",
    "TCnnCTGnCTnnAAATAGTnnnCCAGCnnTGGCnnTACGnnGnnCnTGGCnTnnGCnTGA\n",
    "nnCnnTTCTTnAGnnAGGGACTnnATCAnnACGnTTnTAGTnnTGATGnGAnAGTTGGAn\n",
    "nnnGCAAnTAnnACnAGGTCTnnnGTGATGnnCCCTTnnAGATGnTTCnnTGnnnnnGGC\n",
    "nCGCAnCGCnnnnnnGCnTACAnnCTGACnnCnAAGCnnCAGCnnGAGnnnTAnTATCAC\n",
    "nnnCnTTATCTnnnnAAAAGATTGnnnGGTnnnAATCnTTnGTGAAAnCnnTTGnnGTCn\n",
    "nGTGAnTnnGGGGAnnTnnnAGAGCnnATTGCnAnnATTATTGCnnnnTCTTCAnACnnG\n",
    "AnnnGGAAnTACCnnnnTnAGTAnnAGnCGTATnnGTCnnATnCnnnAGCnATnnGCGnT\n",
    "TnnnnnnGATTnnACGnnTnCCCTGnCCnCTTTGnnnnTACAnnnnnCACCGnCCCGnnn\n",
    "TCGCTnnnnACTACCnGnnATTGGCnnnAGGCnnnnnTTTTTnnGAnnnGGCGTTnnCGG\n",
    "AGAGTnCTAnTAAGnGAnnnnGCTnGnnGCAACnAnGCAnnTnnCTTACTnnnGATTCAA\n",
    "AGnnnnTTCTnACGAnnAAATnnnGGnnnTCTGCTnnAGAGGAAnnnGTAAAnnnnAGTC\n",
    "GTnAACnnnnnAAGGTTTnCCGTnAnnnnGGTGAnACCnTGCnnGGAAnnnGGATCAnTT\n",
    "ATTnnnnAAAAGAACTAnGAGnnnnnnTGCACnTTnnnnTATTGnnnnnnnnnnTGGCTn\n",
    "CGACCCCnnTTTTnAAAAnnnATCTCAnnnCCCnnAAACnTTTTnnAAGACTTGGnnnTT\n",
    "GCATnnGATTTnGAAAGAnnnnnATnnnnCATnTGnnCAATTnnGnAGTnAGnnnACnnG\n",
    "TnnnnAACTTCnnTnTTATTnGAATGnTnnTGCATTACnCCTCCCnnnnnnnnnnnTTTT\n",
    "TTTTTATTAAAAATTnnACACAAAACnACnnAAGTTTAAAnnTnnnGAnATGTAAnnnCC\n",
    "AAACnnCnTTTAATnnnTATnAAATAnnACnnTTTTAACAnnnnATGnGATCnnTCTnnA\n",
    "GnnGCTCTnnnCACATnnCGATnnnGAnAGAACnnACAGnnTnGAAnnATGnnnTnGATn\n",
    "nnAAGnTAATGnnnTnGAATnnnnnnTGCnnAnnnGAATnnnnnnTCnAnnGTGAnnnnA\n",
    "TnCnnATCnGnnAATCTTTnnnnGAACGCnnACCnTTGCGCnnCTTTTnnnGGTAnTTCC\n",
    "nnnAAAGGCAnnCnACCTGTTnnnnnnnGTGTCAnnTGAAAnnCCnnnnnCTnCTCATnn\n",
    "nnnTnAAATAATTnnTTGnATTnnAATTnATnnTTTCnnnnAAnTGGATnGTTGnnAnGT\n",
    "GCnnnnnnnTGnCTGTnnnnAATTnnAGnCTCACnnTTTAAnnnATATATAnAGnnnnTn\n",
    "CACnnTTTTCTAnTAnnnnAGTTGGnAnnnTTGnnACTTGGnnnTGTAATnAAnTTTTAT\n",
    "CnATCnACAnnTCnnAAGnnnnGATnTGnnTAnGCAnnnnnATAnnnCTGCnnnCAnTCT\n",
    "nnnTATTTnnnnAAGGAGACTCCnnnAAAAACCCnnAATTTTnnAAnnnnnnCCTnnnTA\n",
    "AGACnnCTnnCAAATnCnAGGTnGnnGGACTAnCCCnnGCTnnnnGAnnACTTnnnAAGn\n",
    "nCATnATnCAATnnAAnnnGCnnGGAGGnnnAAAAGAAACnnTAACnnnAAGGAnnnnnn\n",
    "nnnTTCCnnnnCTAGnTAACGGnnCnnnnAGTnnGAAGAGGnnGAAnnAAGCCCnAAATT\n",
    "nnnTGTAnnnATnCnnnTGGCnnTCTTTnCnAGAGTCCnnnAGTTGTnnnnAAnnTTTTG\n",
    "nnnnAGnnAACTGnnnnnTTTTCnnAGTnnnnGCnTGGnnnACCAnnTGnnTAnnnnnTA\n",
    "AGnTCnnnTGnTTGAAAnnAGCnnAGCAnnnnnnnTnnCATTGAGGnnnnGnnnATAATC\n",
    "CnCGTTTnnnnAnTGATAnnTGnGACTACnnCAGnnnTGCnnATTAnnnnnnnTGAnnnT\n",
    "AnCnAnnnnnGTCTnnnCTAAnnGAGTnnnCnnnGAGTTGTTTnnGnnGGAATGnnnCAn\n",
    "GCTCnnAAAGnnnnTGGnGTGGnnnnnTAnnnAATTCnnCATnCTnnAAnnnGGnnnnnn\n",
    "nCTAnnnAAnTATAnnnGGTGnAnGAGACCnGAnnnTAGCnnAAnnACAAGnTnnnnACn\n",
    "CGTGAnnGGGAAAnnnnATGAAAnnnAGAACTTTGGAAnnAGAGnnnnAGTTAACnAGnn\n",
    "nnTACnGTnnnnGAAAnnnnTTGnnTTAAAnnnnnGAAACnnAnnnCTTGnnAAGnTTnn\n",
    "nAnGACTnTGTTATTnnATTAGnTTCAAnnCnnCTTTTTGACnGnnAGGnnAGnnnnTAT\n",
    "TnCTnnnAATnnGATTAnnACnAGAnCCAAnnnnnCATCnAAnTTTnnTTGGnGnnnTGn\n",
    "TTGGAGAAGnGGnnnnTTTAAGnGAAATGnnTAGCnnAnGTCTnCnnTGAnnnnCnnnTG\n",
    "TGTnnnTATnnnnAGnTCnCTGAGCnTTTGAnTnnACAATnGCTTnnnnnAAGATTGAnn\n",
    "GGAnAGGCnAGTAnnnAGCGCnnnAATnTTATnnTGTnGTGGAnnnnACAnnAATTnAAT\n",
    "GTTCnTnnnnTCTTACnnTnnnGAnGGnnnATGnnTTGGTnGTnAATAGCnnTnTTAAAT\n",
    "nnGAnCCnnnCGTnnnnnnnnCTTnnGnAAACnnnACnnGGACnnCAAGGAnGTnnCTAn\n",
    "ACATnGCTTGnCAnnnnnAGTATTTnGGGTGnnnnCTTnnnnnGAAACCnnCTTATGnnC\n",
    "GnnnnTAATGAAnnnAGTnnAAATnGTnnnAAATnnGGGAnnTCTGTTAAnnnAAGTGCA\n",
    "CCAnnnnTTGAnnCCnAGTCCnnnnAGATTATnnTnnTAnTATGATGGnnnnnTACTGnA\n",
    "GTnnAAGAGnnnnCAAnGnnTATGnnnTTGnnGGACCCGAnnAAGAnnnTGnGTnnGAnA\n",
    "CnnnTAnnTGnCCTGnnAAnnnTnAGGnGTnnnnGAAGCCnnAnGAGGAAACTnnnnnCn\n",
    "nnTGGTnnnnnGGAnnAGCTnnCGnTAGCnnnGTnnnnnTCTGAnnnnnCGnTGnCAnnn\n",
    "AATCnGATnCnnGTCnnAnnAATTTGGnnnnnGTATAGGGnnGCGAnnnAnnAGACTnAA\n",
    "TCnGnnnnnAnnnnCCATnnCnnnTAGTnAGCnnTGGnnnnTTCnnCTGCCnnnGAAGnT\n",
    "TTnnnnCCnCTCAGGATAnnnnnnGCAAAnnnnnGACTCnnGTATnCAGnTTTnTATnnn\n",
    "GAGGnTAAAGnnnnCGAnAnnnTGATTnnAGAGGCCTTGGGGnnATGAnAnnACnnnnAT\n",
    "CnnCTTAACCTnnnATTnnnnCTCnAAACnnTTTAAATnnnATGTnnAAGACnnnGCTnC\n",
    "nnCTGTTTCTTnnAATnnnnnTGAnnAnnnnCnnGTGGnnGCATnGnTGnnnnnAATnnn\n",
    "GAGAnGTCTTTnnAGTnnnGGGCCnAnTTTTTnnGGTAAGCAGAACTGnnGCGAnGGTGC\n",
    "nnTGnAAGAAnAGTTGTnCnnnGGTGTnnnnnCTTnnTGnnnnnnTGTTAACnCTnTAnn\n",
    "GCAAnnnTACnnnnGTAACnnnnnnnnnnnnn\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "emboss_new = emboss_test.replace('N','').replace('n','').replace('\\n','')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(emboss_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "emboss_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Run loop over EMBOSS cons -identity option from 0 to 100 to get a range of different consensus sequences [x]\n",
    "##### Compare each consensus with the GENEIOUS consensus and with ~20 randomly selected reads (same reads for each consensus)\n",
    "###### Extract the score and determine what identity gives the best mean match score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_dict = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Consensus/20171103_FAH15473/barcode02/test1.fasta\",\"fasta\"))\n",
    "emboss_test = str(test_dict['test'].seq)\n",
    "emboss_new = emboss_test.replace('N','').replace('n','').replace('\\n','')\n",
    "\n",
    "scores_matrix = []\n",
    "test100 = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Alignment/20171103_FAH15473/barcode02/100_reads.fasta\",\"fasta\"))\n",
    "for key in test100:\n",
    "    tmp = []\n",
    "    alignments = pairwise2.align.globalxx(emboss_new, test100[key].seq, score_only=True)\n",
    "    tmp.append(alignments)\n",
    "    scores_matrix.append(int(alignments))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../consensus_test.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "WHITE='\\033[1;37m'\n",
    "BLUE='\\033[0;34m'\n",
    "RED='\\033[0;31m'\n",
    "NC='\\033[0m'\n",
    "GREEN='\\033[0;32m'\n",
    "PURPLE='\\033[1;35m'\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program extracts a specified number of reads from a fasta\n",
    "file and saves them to a new fasta file\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;31m' + \"Input file is \" + args.input_file + '\\033[1;37m')\n",
    "\n",
    "keys = []\n",
    "test_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "for key in test_dict:\n",
    "    keys.append(key)\n",
    "emboss_test = str(test_dict[keys[0]].seq)\n",
    "emboss_new = emboss_test.replace('N','').replace('n','').replace('\\n','')\n",
    "print(emboss_new)\n",
    "\n",
    "scores_matrix = []\n",
    "test100 = SeqIO.to_dict(SeqIO.parse(\"analysis/Alignment/20171103_FAH15473/barcode02/100_reads.fasta\",\"fasta\"))\n",
    "for key in test100:\n",
    "    alignments = pairwise2.align.globalxx(emboss_new, test100[key].seq, score_only=True)\n",
    "    scores_matrix.append(int(alignments))\n",
    "\n",
    "print(scores_matrix)\n",
    "mean = np.mean(scores_matrix)\n",
    "median=np.median(scores_matrix)\n",
    "f=open((args.input_file[:47]+\"scores.txt\"),\"a+\")\n",
    "f.write(\">%s\\t%s\\t%s\\t\\n\" % (keys[0], mean, median))\n",
    "\n",
    "tmp = pd.DataFrame(scores_matrix)\n",
    "tmp.to_csv(args.input_file[:47]+\"%s.csv\" % keys[0],index=False,header=False)\n",
    "\n",
    "\n",
    "with open((args.input_file[:47]+\"%s_new.fasta\" % keys[0]),\"w+\") as handle:\n",
    "    handle.write(\">%s\\n\" % keys[0] +emboss_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(args.input_file[:-29]+'ids.txt','w') as handle:\n",
    "    handle.writelines(\"%s\\n\" % name for name in ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f=open(\"../../analysis/Consensus/20171103_FAH15473/barcode02/scores.txt\",\"r\")\n",
    "if f.mode == \"r\":\n",
    "    contents=f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp=contents.replace(\"\\t\",\",\").replace(\"\\n\",\"\").replace(\",>\",\">\").split(\">\")[1:]\n",
    "print(len(tmp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data=[]\n",
    "indices=[]\n",
    "for item in tmp:\n",
    "    indices.append(int(item.split(\",\")[0][4:]))\n",
    "    data.append(((float(item.split(\",\")[1])),float(item.split(\",\")[2])))\n",
    "frame = pd.DataFrame(index=indices,data=data,columns=[\"Mean Score\",\"Median Score\"])\n",
    "print(frame.loc[frame['Mean Score'].idxmax()].name)\n",
    "print(len(frame))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "frame = frame.sort_index(ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with pd.option_context('display.max_rows', None, 'display.max_columns', None):\n",
    "#     print(frame)\n",
    "    \n",
    "#     # Index 0 is the Geneious-generated Consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import ranksums\n",
    "geneious = pd.read_csv(\"../../analysis/Consensus/20171103_FAH15473/barcode02/test0.csv\",header=None)\n",
    "test20 = pd.read_csv(\"../../analysis/Consensus/20171103_FAH15473/barcode02/test20.csv\",header=None)\n",
    "test1 = pd.read_csv(\"../../analysis/Consensus/20171103_FAH15473/barcode02/test1.csv\",header=None)\n",
    "test52 = pd.read_csv(\"../../analysis/Consensus/20171103_FAH15473/barcode02/test52.csv\",header=None)\n",
    "test75 = pd.read_csv(\"../../analysis/Consensus/20171103_FAH15473/barcode02/test75.csv\",header=None)\n",
    "\n",
    "print(ranksums(geneious,test1))\n",
    "print(ranksums(geneious,test20))\n",
    "print(ranksums(geneious,test52))\n",
    "print(ranksums(geneious,test75))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot = pd.DataFrame(columns=['geneious','test1','test20','test52','test75'])\n",
    "plot['geneious'] = geneious[0]\n",
    "plot['test1'] = test1[0]\n",
    "plot['test20'] = test20[0]\n",
    "plot['test52'] = test52[0]\n",
    "plot['test75'] = test75[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = sns.violinplot(data=plot)\n",
    "ax.set_title(\"Different consensus sequence generation methos for 20171103_FAH15473/barcode02\", fontsize=12)\n",
    "ax.set_xlabel(\"Consensus Generation Method\", fontsize=10)\n",
    "ax.set_ylabel(\"Score\", fontsize=10)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test100 = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Alignment/20171103_FAH15473/barcode02/100_reads.fasta\",\"fasta\"))\n",
    "test_dict = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Consensus/20171103_FAH15473/barcode02/test0_new.fasta\",\"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scores_matrix = []\n",
    "tmp = []\n",
    "for key in test100:\n",
    "    alignments = pairwise2.align.globalxx(emboss_new, test100[key].seq)\n",
    "    scores_matrix.append(alignments)\n",
    "    tmp.append(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Import paf\n",
    "2. Extract columns 10/11 for each row\n",
    "3. Save the alignment percentage identity to a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f=open(\"../../analysis/Consensus/20171103_FAH15473/barcode02/test0.paf\",\"r\")\n",
    "if f.mode == \"r\":\n",
    "    contents=f.read()  \n",
    "tmp=contents.replace(\"\\t\",\",\").split('\\n')\n",
    "tmp_dict = {}\n",
    "for line in tmp[:-1]:\n",
    "    tmp_dict[line.split(\",\")[0]] = [int(line.split(\",\")[9]),int(line.split(\",\")[10])]\n",
    "geneious = pd.DataFrame.from_dict(tmp_dict,orient='index',columns=['matching bases','total bases'])\n",
    "geneious['alignment identity'] = 100*(geneious['matching bases']/geneious['total bases'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f=open(\"../../analysis/Consensus/20171103_FAH15473/barcode02/test20.paf\",\"r\")\n",
    "if f.mode == \"r\":\n",
    "    contents=f.read()  \n",
    "tmp=contents.replace(\"\\t\",\",\").split('\\n')\n",
    "tmp_dict = {}\n",
    "for line in tmp[:-1]:\n",
    "    tmp_dict[line.split(\",\")[0]] = [int(line.split(\",\")[9]),int(line.split(\",\")[10])]\n",
    "test20 = pd.DataFrame.from_dict(tmp_dict,orient='index',columns=['matching bases','total bases'])\n",
    "test20['alignment identity'] = 100*(test20['matching bases']/test20['total bases'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot = pd.DataFrame(columns=['geneious','test20'])\n",
    "plot['geneious'] = geneious['alignment identity']\n",
    "plot['test20'] = test20['alignment identity']\n",
    "\n",
    "ax = sns.violinplot(data=plot)\n",
    "ax.set_title(\"Different consensus sequence generation methods for 20171103_FAH15473/barcode02\", fontsize=12)\n",
    "ax.set_xlabel(\"Consensus Generation Method\", fontsize=10)\n",
    "ax.set_ylabel(\"Alignment Percentage Identity\", fontsize=10)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../cleanup.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program cleans a consensus sequence by removing any\n",
    "'n' or 'N' characters\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "keys = []\n",
    "test_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "for key in test_dict:\n",
    "    keys.append(key)\n",
    "emboss_test = str(test_dict[keys[0]].seq)\n",
    "emboss_new = emboss_test.replace('N','').replace('n','').replace('\\n','')\n",
    "\n",
    "with open(args.input_file[:47]+\"clean_consensus_1000.fasta\",\"w+\") as handle:\n",
    "    handle.write(\">%s\\n\" % keys[0] + emboss_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../cleanup_final.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program cleans a consensus sequence by removing any\n",
    "'n' or 'N' characters\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "keys = []\n",
    "test_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "for key in test_dict:\n",
    "    keys.append(key)\n",
    "emboss_test = str(test_dict[keys[0]].seq)\n",
    "emboss_new = emboss_test.replace('N','').replace('n','').replace('\\n','')\n",
    "\n",
    "with open(args.input_file[:47]+\"final_clean_consensus.fasta\",\"w+\") as handle:\n",
    "    handle.write(\">%s\\n\" % keys[0] + emboss_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../qscore_cleanup.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program cleans a consensus sequence by removing any\n",
    "'n' or 'N' characters\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "keys = []\n",
    "test_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "for key in test_dict:\n",
    "    keys.append(key)\n",
    "emboss_test = str(test_dict[keys[0]].seq)\n",
    "emboss_new = emboss_test.replace('N','').replace('n','').replace('\\n','')\n",
    "\n",
    "with open(args.input_file[:47]+\"clean_qscore_consensus.fasta\",\"w+\") as handle:\n",
    "    handle.write(\">%s\\n\" % keys[0] + emboss_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../cleanup_100.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program cleans a consensus sequence by removing any\n",
    "'n' or 'N' characters\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "keys = []\n",
    "test_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "for key in test_dict:\n",
    "    keys.append(key)\n",
    "emboss_test = str(test_dict[keys[0]].seq)\n",
    "emboss_new = emboss_test.replace('N','').replace('n','').replace('\\n','')\n",
    "\n",
    "with open(args.input_file[:47]+\"clean_consensus_100.fasta\",\"w+\") as handle:\n",
    "    handle.write(\">%s\\n\" % keys[0] + emboss_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../cleanup_101.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program cleans a consensus sequence by removing any\n",
    "'n' or 'N' characters\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "keys = []\n",
    "test_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "for key in test_dict:\n",
    "    keys.append(key)\n",
    "emboss_test = str(test_dict[keys[0]].seq)\n",
    "emboss_new = emboss_test.replace('N','').replace('n','').replace('\\n','')\n",
    "\n",
    "with open(args.input_file[:47]+\"clean_consensus_101.fasta\",\"w+\") as handle:\n",
    "    handle.write(\">%s\\n\" % keys[0] + emboss_new)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Steps to generate consensus\n",
    "\n",
    "    1. Subsample a set number of reads using the get_cons_ids python script line\n",
    "    2. From these reads, generate an alignment file using muscle via the create_consensus script line, on the Consensus/ folder\n",
    "    3. Use the emboss_cons script line to create a consensus from the alignment file, from the analysis folder\n",
    "        3.5 This also involves a cleanup phase where any 'n' or 'N' character is removed from the resulting file\n",
    "    4. Collect all cleaned consensus files in a single fasta file using the make_database.sh script\n",
    "    5. Rename each file???"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../make_database_1000.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program writes a cleaned consensus to a database file\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "keys = []\n",
    "test_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "for key in test_dict:\n",
    "    keys.append(key)\n",
    "emboss_test = str(test_dict[keys[0]].seq)\n",
    "emboss_new = emboss_test.replace('N','').replace('n','').replace('\\n','')\n",
    "    \n",
    "with open(\"database/custom_database_1000.fasta\",\"a\") as handle:\n",
    "    handle.write(\">%s\\n\" % keys[0] + emboss_new + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../make_database_final.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program writes a cleaned consensus to a database file\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.input_file != 'analysis/Consensus/20171207_FAH18654/barcode10/clean_qscore_consensus.fasta' and args.input_file != 'analysis/Consensus/20171212_FAH18688/barcode10/clean_qscore_consensus.fasta':\n",
    "    if args.verbose:\n",
    "        print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "    keys = []\n",
    "    test_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "    for key in test_dict:\n",
    "        keys.append(key)\n",
    "    emboss_test = str(test_dict[keys[0]].seq)\n",
    "    emboss_new = emboss_test.replace('N','').replace('n','').replace('\\n','')\n",
    "\n",
    "    with open(\"database/make_database_final.fasta\",\"a\") as handle:\n",
    "        handle.write(\">%s\\n\" % keys[0] + emboss_new + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../make_qscore_database.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program writes a cleaned consensus to a database file\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.input_file != 'analysis/Consensus/20171207_FAH18654/barcode10/clean_qscore_consensus.fasta' and args.input_file != 'analysis/Consensus/20171212_FAH18688/barcode10/clean_qscore_consensus.fasta':\n",
    "    if args.verbose:\n",
    "        print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "    keys = []\n",
    "    test_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "    for key in test_dict:\n",
    "        keys.append(key)\n",
    "    emboss_test = str(test_dict[keys[0]].seq)\n",
    "    emboss_new = emboss_test.replace('N','').replace('n','').replace('\\n','')\n",
    "\n",
    "    with open(\"database/qscore_database.fasta\",\"a\") as handle:\n",
    "        handle.write(\">%s\\n\" % keys[0] + emboss_new + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../make_database_100.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program writes a cleaned consensus to a database file\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "keys = []\n",
    "test_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "for key in test_dict:\n",
    "    keys.append(key)\n",
    "emboss_test = str(test_dict[keys[0]].seq)\n",
    "emboss_new = emboss_test.replace('N','').replace('n','').replace('\\n','')\n",
    "    \n",
    "with open(\"database/custom_database_100.fasta\",\"a\") as handle:\n",
    "    handle.write(\">%s\\n\" % keys[0] + emboss_new + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../make_database_101.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program writes a cleaned consensus to a database file\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "keys = []\n",
    "test_dict = SeqIO.to_dict(SeqIO.parse(args.input_file,\"fasta\"))\n",
    "for key in test_dict:\n",
    "    keys.append(key)\n",
    "emboss_test = str(test_dict[keys[0]].seq)\n",
    "emboss_new = emboss_test.replace('N','').replace('n','').replace('\\n','')\n",
    "    \n",
    "with open(\"database/custom_database_101.fasta\",\"a\") as handle:\n",
    "    handle.write(\">%s\\n\" % keys[0] + emboss_new + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../1000_minimap_result.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program extracts a specified number of reads from a fasta\n",
    "file and saves them to a new fasta file\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "f=open(args.input_file,\"r\")\n",
    "if f.mode == \"r\":\n",
    "    contents=f.read()\n",
    "\n",
    "tmp=contents.replace(\"\\t\",\",\").split('\\n')\n",
    "tmp_dict = {}\n",
    "for line in tmp[:-1]:\n",
    "    tmp_dict[line.split(\",\")[0]] = str(line.split(\",\")[5])\n",
    "\n",
    "count_dict = {}\n",
    "correct = 0\n",
    "incorrect = 0\n",
    "for item in tmp_dict:\n",
    "    if tmp_dict[item] not in count_dict:\n",
    "        count_dict[tmp_dict[item]] = 1\n",
    "    else:\n",
    "        count_dict[tmp_dict[item]] = count_dict[tmp_dict[item]] + 1\n",
    "        \n",
    "tmp = pd.DataFrame.from_dict(count_dict,orient='index',columns=[\"Count\"])\n",
    "tmp[\"Percentage Match\"] = tmp.apply(lambda row: 100*row.Count/1000,axis=1)\n",
    "tmp.index.names = ['analysis/Consensus/'+args.input_file[19:-18]]\n",
    "tmp = tmp.sort_values(by=\"Count\",ascending=False)\n",
    "\n",
    "if args.verbose:\n",
    "    print(tmp)\n",
    "\n",
    "tmp.to_csv(args.input_file[:-18]+'match_distribution.csv',sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../record_name.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program alters the record.name of entries in a fasta file\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "    \n",
    "with open(args.input_file) as original, open(args.input_file[:-15]+'labelled_read_pool.fasta', 'w') as corrected:\n",
    "    records = SeqIO.parse(original, 'fasta')\n",
    "    for record in records:\n",
    "        record.description = (args.input_file.split('/')[-3]+'_'+args.input_file.split('/')[-2])\n",
    "        SeqIO.write(record, corrected, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f=open(\"../../analysis/Alignment/20171207_FAH18654/barcode12/1000_reads_qiime.paf\",\"r\")\n",
    "if f.mode == \"r\":\n",
    "    contents=f.read()  \n",
    "tmp=contents.replace(\"\\t\",\",\").split('\\n')\n",
    "tmp\n",
    "tmp_dict = {}\n",
    "for line in tmp[:-1]:\n",
    "    tmp_dict[line.split(\",\")[0]] = str(line.split(\",\")[5].split('_')[0])\n",
    "tmp2 = pd.DataFrame.from_dict(data=tmp_dict, orient='index')\n",
    "counts = tmp2[0].value_counts()\n",
    "tmp3 = counts.to_frame().reset_index(drop=False)\n",
    "g=open(\"../../database/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt\", \"r\")\n",
    "if g.mode == \"r\":\n",
    "    contents2=g.read()\n",
    "tmp4=contents2.replace(\"\\t\",\",\").split(\"\\n\")\n",
    "tmp5 = []\n",
    "for i in range(0,len(tmp4)):\n",
    "    tmp5.append(tmp4[i].split(','))\n",
    "\n",
    "tmp6 = pd.DataFrame(tmp5)\n",
    "print(tmp3)\n",
    "\n",
    "tmp_dict2 = {}\n",
    "for i in range(0,len(tmp6[0])):\n",
    "    tmp_dict2[tmp6[0][i].split(\"_\")[0]]=tmp6[1][i]\n",
    "\n",
    "for a,b in tmp3[:7].itertuples(index=False):\n",
    "    print(a)\n",
    "    print(b)\n",
    "    if a in tmp_dict2:\n",
    "        print(tmp_dict2[a]+'\\n')\n",
    "    else:\n",
    "        print(\"not found\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../phylogenetic_naming_qscore.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program alters the title of entries in the consensus fasta\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "true_names = {'20171103_FAH15473/barcode01': 'Puccinia_striiformis_tritici',\n",
    "             '20171103_FAH15473/barcode02': 'Zymoseptoria_tritici',\n",
    "             '20171103_FAH15473/barcode03': 'Pyrenophora_tritici-repentis',\n",
    "             '20171103_FAH15473/barcode04': 'Fusarium_oxysporum',\n",
    "             '20171103_FAH15473/barcode05': 'Tuber_brumale',\n",
    "             '20171103_FAH15473/barcode06': 'Cortinarius_globuliformis',\n",
    "             '20171103_FAH15473/barcode07': 'Aspergillus_niger',\n",
    "             '20171103_FAH15473/barcode08': 'Clavispora_lusitaniae',\n",
    "             '20171103_FAH15473/barcode09': 'Cryptococcus_neoformans',\n",
    "             '20171103_FAH15473/barcode10': 'Penicillium_chrysogenum',\n",
    "             '20171103_FAH15473/barcode11': 'Rhodotorula_mucilaginosa',\n",
    "             '20171103_FAH15473/barcode12': 'Scedosporium_boydii',\n",
    "             '20171207_FAH18654/barcode01': 'Blastobotrys_proliferans',\n",
    "             '20171207_FAH18654/barcode02': 'Candida_zeylanoides',\n",
    "             '20171207_FAH18654/barcode03': 'Galactomyces_geotrichum',\n",
    "             '20171207_FAH18654/barcode04': 'Kodamaea_ohmeri',\n",
    "             '20171207_FAH18654/barcode05': 'Meyerozyma_guillermondii',\n",
    "             '20171207_FAH18654/barcode06': 'Wickerhamomyces_anomalus',\n",
    "             '20171207_FAH18654/barcode07': 'Yamadazyma_mexicana',\n",
    "             '20171207_FAH18654/barcode08': 'Yamadazyma_scolyti',\n",
    "             '20171207_FAH18654/barcode09': 'Yarrowia_lipolytica',\n",
    "             '20171207_FAH18654/barcode11': 'Zygoascus_hellenicus',\n",
    "             '20171207_FAH18654/barcode12': 'Aspergillus_flavus',\n",
    "             '20171212_FAH18688/barcode01': 'Cryptococcus_zero',\n",
    "             '20171212_FAH18688/barcode02': 'Aspergillus_sp.',\n",
    "             '20171212_FAH18688/barcode03': 'CCL067',\n",
    "             '20171212_FAH18688/barcode04': 'Diaporthe_sp.',\n",
    "             '20171212_FAH18688/barcode05': 'Tapesia_yallundae_CCL031',\n",
    "             '20171212_FAH18688/barcode06': 'Tapesia_yallundae_CCL029',\n",
    "             '20171212_FAH18688/barcode07': 'Dothiorella_vidmadera',\n",
    "             '20171212_FAH18688/barcode08': 'Quambalaria_cyanescens',\n",
    "             '20171212_FAH18688/barcode09': 'Entoleuca_sp.',\n",
    "             '20171212_FAH18688/barcode11': 'CCL060',\n",
    "             '20171212_FAH18688/barcode12': 'CCL068',\n",
    "             '20180108_FAH18647/barcode01': 'Saccharomyces_cerevisiae',\n",
    "             '20180108_FAH18647/barcode02': 'Cladophialophora_sp.',\n",
    "             '20180108_FAH18647/barcode03': 'Candida_albicans',\n",
    "             '20180108_FAH18647/barcode04': 'Candida_metapsilosis',\n",
    "             '20180108_FAH18647/barcode05': 'Candida_orthopsilosis',\n",
    "             '20180108_FAH18647/barcode06': 'Candida_parapsilosis',\n",
    "             '20180108_FAH18647/barcode07': 'Cryptococcus_gattii',\n",
    "             '20180108_FAH18647/barcode08': 'Geotrichum_candidum',\n",
    "             '20180108_FAH18647/barcode09': 'Kluyveromyces_lactis',\n",
    "             '20180108_FAH18647/barcode10': 'Kluyveromyces_marxianus',\n",
    "             '20180108_FAH18647/barcode11': 'Pichia_kudriavzevii',\n",
    "             '20180108_FAH18647/barcode12': 'Pichia_membranifaciens'}\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "with open(args.input_file) as original, open(args.input_file[:-21]+'qscore_database_labelled.fasta', 'w+') as labelled:\n",
    "    for line in original:\n",
    "        if line.startswith(\">\"):\n",
    "            barcode = line.split('/')[2]+'/'+line.split('/')[3]\n",
    "            species = true_names[barcode]\n",
    "            newline = '>'+species+'\\n'\n",
    "        else:\n",
    "            newline = line\n",
    "        labelled.write(newline)\n",
    "original.close()\n",
    "labelled.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../phylogenetic_naming_100.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program alters the title of entries in the consensus fasta\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "true_names = {'20171103_FAH15473/barcode01': 'Puccinia_striiformis_tritici',\n",
    "             '20171103_FAH15473/barcode02': 'Zymoseptoria_tritici',\n",
    "             '20171103_FAH15473/barcode03': 'Pyrenophora_tritici-repentis',\n",
    "             '20171103_FAH15473/barcode04': 'Fusarium_oxysporum',\n",
    "             '20171103_FAH15473/barcode05': 'Tuber_brumale',\n",
    "             '20171103_FAH15473/barcode06': 'Cortinarius_globuliformis',\n",
    "             '20171103_FAH15473/barcode07': 'Aspergillus_niger',\n",
    "             '20171103_FAH15473/barcode08': 'Clavispora_lusitaniae',\n",
    "             '20171103_FAH15473/barcode09': 'Cryptococcus_neoformans',\n",
    "             '20171103_FAH15473/barcode10': 'Penicillium_chrysogenum',\n",
    "             '20171103_FAH15473/barcode11': 'Rhodotorula_mucilaginosa',\n",
    "             '20171103_FAH15473/barcode12': 'Scedosporium_boydii',\n",
    "             '20171207_FAH18654/barcode01': 'Blastobotrys_proliferans',\n",
    "             '20171207_FAH18654/barcode02': 'Candida_zeylanoides',\n",
    "             '20171207_FAH18654/barcode03': 'Galactomyces_geotrichum',\n",
    "             '20171207_FAH18654/barcode04': 'Kodamaea_ohmeri',\n",
    "             '20171207_FAH18654/barcode05': 'Meyerozyma_guillermondii',\n",
    "             '20171207_FAH18654/barcode06': 'Wickerhamomyces_anomalus',\n",
    "             '20171207_FAH18654/barcode07': 'Yamadazyma_mexicana',\n",
    "             '20171207_FAH18654/barcode08': 'Yamadazyma_scolyti',\n",
    "             '20171207_FAH18654/barcode09': 'Yarrowia_lipolytica',\n",
    "             '20171207_FAH18654/barcode11': 'Zygoascus_hellenicus',\n",
    "             '20171207_FAH18654/barcode12': 'Aspergillus_flavus',\n",
    "             '20171212_FAH18688/barcode01': 'Cryptococcus_zero',\n",
    "             '20171212_FAH18688/barcode02': 'Aspergillus_sp.',\n",
    "             '20171212_FAH18688/barcode03': 'CCL067',\n",
    "             '20171212_FAH18688/barcode04': 'Diaporthe_sp.',\n",
    "             '20171212_FAH18688/barcode05': 'Tapesia_yallundae_CCL031',\n",
    "             '20171212_FAH18688/barcode06': 'Tapesia_yallundae_CCL029',\n",
    "             '20171212_FAH18688/barcode07': 'Dothiorella_vidmadera',\n",
    "             '20171212_FAH18688/barcode08': 'Quambalaria_cyanescens',\n",
    "             '20171212_FAH18688/barcode09': 'Entoleuca_sp.',\n",
    "             '20171212_FAH18688/barcode11': 'CCL060',\n",
    "             '20171212_FAH18688/barcode12': 'CCL068',\n",
    "             '20180108_FAH18647/barcode01': 'Saccharomyces_cerevisiae',\n",
    "             '20180108_FAH18647/barcode02': 'Cladophialophora_sp.',\n",
    "             '20180108_FAH18647/barcode03': 'Candida_albicans',\n",
    "             '20180108_FAH18647/barcode04': 'Candida_metapsilosis',\n",
    "             '20180108_FAH18647/barcode05': 'Candida_orthopsilosis',\n",
    "             '20180108_FAH18647/barcode06': 'Candida_parapsilosis',\n",
    "             '20180108_FAH18647/barcode07': 'Cryptococcus_gattii',\n",
    "             '20180108_FAH18647/barcode08': 'Geotrichum_candidum',\n",
    "             '20180108_FAH18647/barcode09': 'Kluyveromyces_lactis',\n",
    "             '20180108_FAH18647/barcode10': 'Kluyveromyces_marxianus',\n",
    "             '20180108_FAH18647/barcode11': 'Pichia_kudriavzevii',\n",
    "             '20180108_FAH18647/barcode12': 'Pichia_membranifaciens'}\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "with open(args.input_file) as original, open(args.input_file[:-25]+'custom_database_100_labelled.fasta', 'w+') as labelled:\n",
    "    for line in original:\n",
    "        if line.startswith(\">\"):\n",
    "            barcode = line.split('/')[2]+'/'+line.split('/')[3]\n",
    "            species = true_names[barcode]\n",
    "            newline = '>'+species+'\\n'\n",
    "        else:\n",
    "            newline = line\n",
    "        labelled.write(newline)\n",
    "original.close()\n",
    "labelled.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../phylogenetic_naming_101.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program alters the title of entries in the consensus fasta\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "true_names = {'20171103_FAH15473/barcode01': 'Puccinia_striiformis_tritici',\n",
    "             '20171103_FAH15473/barcode02': 'Zymoseptoria_tritici',\n",
    "             '20171103_FAH15473/barcode03': 'Pyrenophora_tritici-repentis',\n",
    "             '20171103_FAH15473/barcode04': 'Fusarium_oxysporum',\n",
    "             '20171103_FAH15473/barcode05': 'Tuber_brumale',\n",
    "             '20171103_FAH15473/barcode06': 'Cortinarius_globuliformis',\n",
    "             '20171103_FAH15473/barcode07': 'Aspergillus_niger',\n",
    "             '20171103_FAH15473/barcode08': 'Clavispora_lusitaniae',\n",
    "             '20171103_FAH15473/barcode09': 'Cryptococcus_neoformans',\n",
    "             '20171103_FAH15473/barcode10': 'Penicillium_chrysogenum',\n",
    "             '20171103_FAH15473/barcode11': 'Rhodotorula_mucilaginosa',\n",
    "             '20171103_FAH15473/barcode12': 'Scedosporium_boydii',\n",
    "             '20171207_FAH18654/barcode01': 'Blastobotrys_proliferans',\n",
    "             '20171207_FAH18654/barcode02': 'Candida_zeylanoides',\n",
    "             '20171207_FAH18654/barcode03': 'Galactomyces_geotrichum',\n",
    "             '20171207_FAH18654/barcode04': 'Kodamaea_ohmeri',\n",
    "             '20171207_FAH18654/barcode05': 'Meyerozyma_guillermondii',\n",
    "             '20171207_FAH18654/barcode06': 'Wickerhamomyces_anomalus',\n",
    "             '20171207_FAH18654/barcode07': 'Yamadazyma_mexicana',\n",
    "             '20171207_FAH18654/barcode08': 'Yamadazyma_scolyti',\n",
    "             '20171207_FAH18654/barcode09': 'Yarrowia_lipolytica',\n",
    "             '20171207_FAH18654/barcode11': 'Zygoascus_hellenicus',\n",
    "             '20171207_FAH18654/barcode12': 'Aspergillus_flavus',\n",
    "             '20171212_FAH18688/barcode01': 'Cryptococcus_zero',\n",
    "             '20171212_FAH18688/barcode02': 'Aspergillus_sp.',\n",
    "             '20171212_FAH18688/barcode03': 'CCL067',\n",
    "             '20171212_FAH18688/barcode04': 'Diaporthe_sp.',\n",
    "             '20171212_FAH18688/barcode05': 'Tapesia_yallundae_CCL031',\n",
    "             '20171212_FAH18688/barcode06': 'Tapesia_yallundae_CCL029',\n",
    "             '20171212_FAH18688/barcode07': 'Dothiorella_vidmadera',\n",
    "             '20171212_FAH18688/barcode08': 'Quambalaria_cyanescens',\n",
    "             '20171212_FAH18688/barcode09': 'Entoleuca_sp.',\n",
    "             '20171212_FAH18688/barcode11': 'CCL060',\n",
    "             '20171212_FAH18688/barcode12': 'CCL068',\n",
    "             '20180108_FAH18647/barcode01': 'Saccharomyces_cerevisiae',\n",
    "             '20180108_FAH18647/barcode02': 'Cladophialophora_sp.',\n",
    "             '20180108_FAH18647/barcode03': 'Candida_albicans',\n",
    "             '20180108_FAH18647/barcode04': 'Candida_metapsilosis',\n",
    "             '20180108_FAH18647/barcode05': 'Candida_orthopsilosis',\n",
    "             '20180108_FAH18647/barcode06': 'Candida_parapsilosis',\n",
    "             '20180108_FAH18647/barcode07': 'Cryptococcus_gattii',\n",
    "             '20180108_FAH18647/barcode08': 'Geotrichum_candidum',\n",
    "             '20180108_FAH18647/barcode09': 'Kluyveromyces_lactis',\n",
    "             '20180108_FAH18647/barcode10': 'Kluyveromyces_marxianus',\n",
    "             '20180108_FAH18647/barcode11': 'Pichia_kudriavzevii',\n",
    "             '20180108_FAH18647/barcode12': 'Pichia_membranifaciens'}\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "with open(args.input_file) as original, open(args.input_file[:-25]+'custom_database_101_labelled.fasta', 'w+') as labelled:\n",
    "    for line in original:\n",
    "        if line.startswith(\">\"):\n",
    "            barcode = line.split('/')[2]+'/'+line.split('/')[3]\n",
    "            species = true_names[barcode]\n",
    "            newline = '>'+species+'\\n'\n",
    "        else:\n",
    "            newline = line\n",
    "        labelled.write(newline)\n",
    "original.close()\n",
    "labelled.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../phylogenetic_naming_1000.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program alters the title of entries in the consensus fasta\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "true_names = {'20171103_FAH15473/barcode01': 'Puccinia_striiformis_tritici',\n",
    "             '20171103_FAH15473/barcode02': 'Zymoseptoria_tritici',\n",
    "             '20171103_FAH15473/barcode03': 'Pyrenophora_tritici-repentis',\n",
    "             '20171103_FAH15473/barcode04': 'Fusarium_oxysporum',\n",
    "             '20171103_FAH15473/barcode05': 'Tuber_brumale',\n",
    "             '20171103_FAH15473/barcode06': 'Cortinarius_globuliformis',\n",
    "             '20171103_FAH15473/barcode07': 'Aspergillus_niger',\n",
    "             '20171103_FAH15473/barcode08': 'Clavispora_lusitaniae',\n",
    "             '20171103_FAH15473/barcode09': 'Cryptococcus_neoformans',\n",
    "             '20171103_FAH15473/barcode10': 'Penicillium_chrysogenum',\n",
    "             '20171103_FAH15473/barcode11': 'Rhodotorula_mucilaginosa',\n",
    "             '20171103_FAH15473/barcode12': 'Scedosporium_boydii',\n",
    "             '20171207_FAH18654/barcode01': 'Blastobotrys_proliferans',\n",
    "             '20171207_FAH18654/barcode02': 'Candida_zeylanoides',\n",
    "             '20171207_FAH18654/barcode03': 'Galactomyces_geotrichum',\n",
    "             '20171207_FAH18654/barcode04': 'Kodamaea_ohmeri',\n",
    "             '20171207_FAH18654/barcode05': 'Meyerozyma_guillermondii',\n",
    "             '20171207_FAH18654/barcode06': 'Wickerhamomyces_anomalus',\n",
    "             '20171207_FAH18654/barcode07': 'Yamadazyma_mexicana',\n",
    "             '20171207_FAH18654/barcode08': 'Yamadazyma_scolyti',\n",
    "             '20171207_FAH18654/barcode09': 'Yarrowia_lipolytica',\n",
    "             '20171207_FAH18654/barcode11': 'Zygoascus_hellenicus',\n",
    "             '20171207_FAH18654/barcode12': 'Aspergillus_flavus',\n",
    "             '20171212_FAH18688/barcode01': 'Cryptococcus_zero',\n",
    "             '20171212_FAH18688/barcode02': 'Aspergillus_sp.',\n",
    "             '20171212_FAH18688/barcode03': 'CCL067',\n",
    "             '20171212_FAH18688/barcode04': 'Diaporthe_sp.',\n",
    "             '20171212_FAH18688/barcode05': 'Tapesia_yallundae_CCL031',\n",
    "             '20171212_FAH18688/barcode06': 'Tapesia_yallundae_CCL029',\n",
    "             '20171212_FAH18688/barcode07': 'Dothiorella_vidmadera',\n",
    "             '20171212_FAH18688/barcode08': 'Quambalaria_cyanescens',\n",
    "             '20171212_FAH18688/barcode09': 'Entoleuca_sp.',\n",
    "             '20171212_FAH18688/barcode11': 'CCL060',\n",
    "             '20171212_FAH18688/barcode12': 'CCL068',\n",
    "             '20180108_FAH18647/barcode01': 'Saccharomyces_cerevisiae',\n",
    "             '20180108_FAH18647/barcode02': 'Cladophialophora_sp.',\n",
    "             '20180108_FAH18647/barcode03': 'Candida_albicans',\n",
    "             '20180108_FAH18647/barcode04': 'Candida_metapsilosis',\n",
    "             '20180108_FAH18647/barcode05': 'Candida_orthopsilosis',\n",
    "             '20180108_FAH18647/barcode06': 'Candida_parapsilosis',\n",
    "             '20180108_FAH18647/barcode07': 'Cryptococcus_gattii',\n",
    "             '20180108_FAH18647/barcode08': 'Geotrichum_candidum',\n",
    "             '20180108_FAH18647/barcode09': 'Kluyveromyces_lactis',\n",
    "             '20180108_FAH18647/barcode10': 'Kluyveromyces_marxianus',\n",
    "             '20180108_FAH18647/barcode11': 'Pichia_kudriavzevii',\n",
    "             '20180108_FAH18647/barcode12': 'Pichia_membranifaciens'}\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "with open(args.input_file) as original, open(args.input_file[:-26]+'custom_database_1000_labelled.fasta', 'w+') as labelled:\n",
    "    for line in original:\n",
    "        if line.startswith(\">\"):\n",
    "            barcode = line.split('/')[2]+'/'+line.split('/')[3]\n",
    "            species = true_names[barcode]\n",
    "            newline = '>'+species+'\\n'\n",
    "        else:\n",
    "            newline = line\n",
    "        labelled.write(newline)\n",
    "original.close()\n",
    "labelled.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../phylogenetic_naming.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program alters the title of entries in the consensus fasta\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "true_names = {'20171103_FAH15473/barcode01': 'Puccinia_striiformis-tritici',\n",
    "             '20171103_FAH15473/barcode02': 'Zymoseptoria_tritici',\n",
    "             '20171103_FAH15473/barcode03': 'Pyrenophora_tritici-repentis',\n",
    "             '20171103_FAH15473/barcode04': 'Fusarium_oxysporum',\n",
    "             '20171103_FAH15473/barcode05': 'Tuber_brumale',\n",
    "             '20171103_FAH15473/barcode06': 'Cortinarius_globuliformis',\n",
    "             '20171103_FAH15473/barcode07': 'Aspergillus_niger',\n",
    "             '20171103_FAH15473/barcode08': 'Clavispora_lusitaniae',\n",
    "             '20171103_FAH15473/barcode09': 'Kluyveromyces_unidentified',\n",
    "             '20171103_FAH15473/barcode10': 'Penicillium_chrysogenum',\n",
    "             '20171103_FAH15473/barcode11': 'Rhodotorula_mucilaginosa',\n",
    "             '20171103_FAH15473/barcode12': 'Scedosporium_boydii',\n",
    "             '20171207_FAH18654/barcode01': 'Blastobotrys_proliferans',\n",
    "             '20171207_FAH18654/barcode02': 'Debaryomyces_unidentified',\n",
    "             '20171207_FAH18654/barcode03': 'Galactomyces_geotrichum',\n",
    "             '20171207_FAH18654/barcode04': 'Kodamaea_ohmeri',\n",
    "             '20171207_FAH18654/barcode05': 'Meyerozyma_guillermondii',\n",
    "             '20171207_FAH18654/barcode06': 'Wickerhamomyces_anomalus',\n",
    "             '20171207_FAH18654/barcode07': 'Yamadazyma_mexicana',\n",
    "             '20171207_FAH18654/barcode08': 'Yamadazyma_scolyti',\n",
    "             '20171207_FAH18654/barcode09': 'Yarrowia_lipolytica',\n",
    "             '20171207_FAH18654/barcode11': 'Zygoascus_hellenicus',\n",
    "             '20171207_FAH18654/barcode12': 'Aspergillus_flavus',\n",
    "             '20171212_FAH18688/barcode01': 'Cryptococcus_zero',\n",
    "             '20171212_FAH18688/barcode02': 'Aspergillus_unidentified',\n",
    "             '20171212_FAH18688/barcode03': 'Diaporthe_CCL067',\n",
    "             '20171212_FAH18688/barcode04': 'Diaporthe_unidentified',\n",
    "             '20171212_FAH18688/barcode05': 'Oculimacula_yallundae-CCL031',\n",
    "             '20171212_FAH18688/barcode06': 'Oculimacula_yallundae-CCL029',\n",
    "             '20171212_FAH18688/barcode07': 'Dothiorella_vidmadera',\n",
    "             '20171212_FAH18688/barcode08': 'Quambalaria_cyanescens',\n",
    "             '20171212_FAH18688/barcode09': 'Entoleuca_unidentified',\n",
    "             '20171212_FAH18688/barcode11': 'Asteroma_CCL060',\n",
    "             '20171212_FAH18688/barcode12': 'Asteroma_CCL068',\n",
    "             '20180108_FAH18647/barcode01': 'Saccharomyces_cerevisiae',\n",
    "             '20180108_FAH18647/barcode02': 'Cladophialophora_unidentified',\n",
    "             '20180108_FAH18647/barcode03': 'Candida_albicans',\n",
    "             '20180108_FAH18647/barcode04': 'Candida_metapsilosis',\n",
    "             '20180108_FAH18647/barcode05': 'Candida_orthopsilosis',\n",
    "             '20180108_FAH18647/barcode06': 'Candida_parapsilosis',\n",
    "             '20180108_FAH18647/barcode07': 'Cryptococcus_gattii',\n",
    "             '20180108_FAH18647/barcode08': 'Geotrichum_candidum',\n",
    "             '20180108_FAH18647/barcode09': 'Candida_unidentified',\n",
    "             '20180108_FAH18647/barcode10': 'Kluyveromyces_marxianus',\n",
    "             '20180108_FAH18647/barcode11': 'Pichia_kudriavzevii',\n",
    "             '20180108_FAH18647/barcode12': 'Pichia_membranifaciens'}\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "with open(args.input_file) as original, open(args.input_file[:-21]+'custom_database_labelled.fasta', 'w+') as labelled:\n",
    "    for line in original:\n",
    "        if line.startswith(\">\"):\n",
    "            barcode = line.split('/')[2]+'/'+line.split('/')[3]\n",
    "            species = true_names[barcode]\n",
    "            newline = '>'+species+'\\n'\n",
    "        else:\n",
    "            newline = line\n",
    "        labelled.write(newline)\n",
    "original.close()\n",
    "labelled.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../phylogenetic_naming_final.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This program alters the title of entries in the consensus fasta\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "true_names = {'20171103_FAH15473/barcode01': 'Puccinia_striiformis_tritici',\n",
    "             '20171103_FAH15473/barcode02': 'Zymoseptoria_tritici',\n",
    "             '20171103_FAH15473/barcode03': 'Pyrenophora_tritici-repentis',\n",
    "             '20171103_FAH15473/barcode04': 'Fusarium_oxysporum',\n",
    "             '20171103_FAH15473/barcode05': 'Tuber_brumale',\n",
    "             '20171103_FAH15473/barcode06': 'Cortinarius_globuliformis',\n",
    "             '20171103_FAH15473/barcode07': 'Aspergillus_niger',\n",
    "             '20171103_FAH15473/barcode08': 'Clavispora_lusitaniae',\n",
    "             '20171103_FAH15473/barcode09': 'Cryptococcus_neoformans',\n",
    "             '20171103_FAH15473/barcode10': 'Penicillium_chrysogenum',\n",
    "             '20171103_FAH15473/barcode11': 'Rhodotorula_mucilaginosa',\n",
    "             '20171103_FAH15473/barcode12': 'Scedosporium_boydii',\n",
    "             '20171207_FAH18654/barcode01': 'Blastobotrys_proliferans',\n",
    "             '20171207_FAH18654/barcode02': 'Candida_zeylanoides',\n",
    "             '20171207_FAH18654/barcode03': 'Galactomyces_geotrichum',\n",
    "             '20171207_FAH18654/barcode04': 'Kodamaea_ohmeri',\n",
    "             '20171207_FAH18654/barcode05': 'Meyerozyma_guillermondii',\n",
    "             '20171207_FAH18654/barcode06': 'Wickerhamomyces_anomalus',\n",
    "             '20171207_FAH18654/barcode07': 'Yamadazyma_mexicana',\n",
    "             '20171207_FAH18654/barcode08': 'Yamadazyma_scolyti',\n",
    "             '20171207_FAH18654/barcode09': 'Yarrowia_lipolytica',\n",
    "             '20171207_FAH18654/barcode11': 'Zygoascus_hellenicus',\n",
    "             '20171207_FAH18654/barcode12': 'Aspergillus_flavus',\n",
    "             '20171212_FAH18688/barcode01': 'Cryptococcus_zero',\n",
    "             '20171212_FAH18688/barcode02': 'Aspergillus_sp.',\n",
    "             '20171212_FAH18688/barcode03': 'CCL067',\n",
    "             '20171212_FAH18688/barcode04': 'Diaporthe_sp.',\n",
    "             '20171212_FAH18688/barcode05': 'Tapesia_yallundae_CCL031',\n",
    "             '20171212_FAH18688/barcode06': 'Tapesia_yallundae_CCL029',\n",
    "             '20171212_FAH18688/barcode07': 'Dothiorella_vidmadera',\n",
    "             '20171212_FAH18688/barcode08': 'Quambalaria_cyanescens',\n",
    "             '20171212_FAH18688/barcode09': 'Entoleuca_sp.',\n",
    "             '20171212_FAH18688/barcode11': 'CCL060',\n",
    "             '20171212_FAH18688/barcode12': 'CCL068',\n",
    "             '20180108_FAH18647/barcode01': 'Saccharomyces_cerevisiae',\n",
    "             '20180108_FAH18647/barcode02': 'Cladophialophora_sp.',\n",
    "             '20180108_FAH18647/barcode03': 'Candida_albicans',\n",
    "             '20180108_FAH18647/barcode04': 'Candida_metapsilosis',\n",
    "             '20180108_FAH18647/barcode05': 'Candida_orthopsilosis',\n",
    "             '20180108_FAH18647/barcode06': 'Candida_parapsilosis',\n",
    "             '20180108_FAH18647/barcode07': 'Cryptococcus_gattii',\n",
    "             '20180108_FAH18647/barcode08': 'Geotrichum_candidum',\n",
    "             '20180108_FAH18647/barcode09': 'Kluyveromyces_lactis',\n",
    "             '20180108_FAH18647/barcode10': 'Kluyveromyces_marxianus',\n",
    "             '20180108_FAH18647/barcode11': 'Pichia_kudriavzevii',\n",
    "             '20180108_FAH18647/barcode12': 'Pichia_membranifaciens'}\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "with open(args.input_file) as original, open(args.input_file[:-25]+'final_custom_database_labelled.fasta', 'w+') as labelled:\n",
    "    for line in original:\n",
    "        if line.startswith(\">\"):\n",
    "            barcode = line.split('/')[2]+'/'+line.split('/')[3]\n",
    "            species = true_names[barcode]\n",
    "            newline = '>'+species+'\\n'\n",
    "        else:\n",
    "            newline = line\n",
    "        labelled.write(newline)\n",
    "original.close()\n",
    "labelled.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Steps to generate consensus and tree\n",
    "\n",
    "    1. Subsample a set number of reads using the get_cons_ids python script line\n",
    "    2. From these reads, generate an alignment file using muscle via the create_consensus script line, on the Consensus/ folder\n",
    "    3. Use the emboss_cons script line to create a consensus from the alignment file, from the analysis folder\n",
    "        3.5 This also involves a cleanup phase where any 'n' or 'N' character is removed from the resulting file\n",
    "    4. Collect all cleaned consensus files in a single fasta file using the make_database.sh script\n",
    "    5. Rename each file using the phylogenetic_naming script line\n",
    "    6. Use muscle to create an alignment file (muscle -in labelled_consensus.fasta -out labelled_consensus.aln)\n",
    "    7. Download file and export to Geneious\n",
    "    8. Utilise the Mask Alignment tool to remove gaps\n",
    "    9. Create tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylogeny = {'20171103_FAH15473/barcode01': 'Basidiomycota/Pucciniomycetes/Pucciniales/Pucciniaceae/Puccinia/striiformis_tritici',\n",
    "             '20171103_FAH15473/barcode02': 'Ascomycota//Dothideomycetes/Capnodiales/Mycosphaerellaceae/Zymoseptoria/tritici',\n",
    "             '20171103_FAH15473/barcode03': 'Ascomycota/Dothideomycetes/Pleosporales/Pleospraceae/Pyrenophora/tritici-repentis',\n",
    "             '20171103_FAH15473/barcode04': 'Ascomycota/Sordariomycetes/Hypocreales/Nectriaceae/Fusarium/oxysporum',\n",
    "             '20171103_FAH15473/barcode05': 'Ascomycota/Pezizomycetes/Pezizales/Tuberaceae/Tuber/brumale',\n",
    "             '20171103_FAH15473/barcode06': 'Basidiomycota/Agaricales/Cortinariaceae/Cortinarius/globuliformis',\n",
    "             '20171103_FAH15473/barcode07': 'Ascomycota/Eurotiomycetes/Eurotiales/Trichocomaceae/Aspergillus/niger',\n",
    "             '20171103_FAH15473/barcode08': 'Ascomycota/Saccharomycetes/Saccharomycetales/Metschnikowiaceae/Clavispora/lusitaniae',\n",
    "                 '20171103_FAH15473/barcode09': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Kluyveromyces/sp. (Cryptococcus_neoformans)',\n",
    "             '20171103_FAH15473/barcode10': 'Ascomycota/Eurotiomycetes/Eurotiales/Trichocomaceae/Penicillium/chrysogenum',\n",
    "             '20171103_FAH15473/barcode11': 'Basidiomycota/Microbotryomycetes/Sporidiobolales/Sporidiobolaceae/Rhodotorula/mucilaginosa',\n",
    "             '20171103_FAH15473/barcode12': 'Ascomycota/Sordariomycetes/Microascales/Microascaceae/Scedosporium/boydii',\n",
    "             '20171207_FAH18654/barcode01': 'Ascomycota/Saccharomycetes/Saccharomycetales/Trichomonascaceae/Blastobotrys/proliferans',\n",
    "                 '20171207_FAH18654/barcode02': 'Ascomycota/Saccharomycetes/Saccharomycetales/Debaryomycetaceae/Meyerozyma/sp. (Candida_zeylanoides)',\n",
    "                 '20171207_FAH18654/barcode03': 'Ascomycota/Saccharomycetes/Saccharomycetales/Dipodascaceae/Galactomyces/geotrichum',\n",
    "             '20171207_FAH18654/barcode04': 'Ascomycota/Saccharomycetes/Saccharomycetales/Metschnikowiaceae/Kodamaea/ohmeri',\n",
    "             '20171207_FAH18654/barcode05': 'Ascomycota/Saccharomycetes/Saccharomycetales/Debaryomycetaceae/Meyerozyma/guillermondii',\n",
    "             '20171207_FAH18654/barcode06': 'Ascomycota/Saccharomycetes/Saccharomycetales/Phaffomycetaceae/Wickerhamomyces/anomalus',\n",
    "             '20171207_FAH18654/barcode07': 'Ascomycota/Saccharomycetes/Saccharomycetales/Debaryomycetaceae/Yamadazyma/mexicana',\n",
    "             '20171207_FAH18654/barcode08': 'Ascomycota/Saccharomycetes/Saccharomycetales/Debaryomycetaceae/Yamadazyma/scolyti',\n",
    "             '20171207_FAH18654/barcode09': 'Ascomycota/Saccharomycetes/Saccharomycetales/Dipodascaceae/Yarrowia/lipolytica',\n",
    "             '20171207_FAH18654/barcode11': 'Ascomycota/Saccharomycetes/Saccharomycetales/Trichomonascaceae/Zygoascus/hellenicus',\n",
    "             '20171207_FAH18654/barcode12': 'Ascomycota/Eurotiomycetes/Eurotiales/Trichocomaceae/Aspergillus/flavus',\n",
    "             '20171212_FAH18688/barcode01': 'Basidiomycota/Tremellomycetes/Tremellales/Tremellaceae/Cryptococcus/zero',\n",
    "             '20171212_FAH18688/barcode02': 'Ascomycota/Eurotiomycetes/Eurotiales/Trichocomaceae/Aspergillus/sp.',\n",
    "                 '20171212_FAH18688/barcode03': 'Ascomycota/Sordariomycetes/Diaporthales/Diaporthaceae/Diaporthe/sp._CCL067',\n",
    "             '20171212_FAH18688/barcode04': 'Ascomycota/Sordariomycetes/Diaporthales/Diaporthaceae/Diaporthe/sp.',\n",
    "             '20171212_FAH18688/barcode05': 'Ascomycota/Leotiomycetes/Helotiales/Dermateaceae/Tapesia/yallundae_CCL031',\n",
    "             '20171212_FAH18688/barcode06': 'Ascomycota/Leotiomycetes/Helotiales/Dermateaceae/Tapesia/yallundae_CCL029',\n",
    "             '20171212_FAH18688/barcode07': 'Ascomycota/Dothideomycetes/Botryosphaeriales/Botryosphaeriaceae/Dothiorella/vidmadera',\n",
    "             '20171212_FAH18688/barcode08': 'Exobasidiomycetes/Microstromatales/Quambalariaceae/Quambalaria/cyanescens',\n",
    "             '20171212_FAH18688/barcode09': 'Ascomycota/Sordariomycetes/Xylariales.Xylariaceae/Entoleuca/sp.',\n",
    "                 '20171212_FAH18688/barcode11': 'Ascomycota/Sordariomycetes/Diaporthales/Diaporthaceae/Diaporthe/sp._CCL060',\n",
    "                 '20171212_FAH18688/barcode12': 'Ascomycota/Sordariomycetes/Diaporthales/Diaporthaceae/Diaporthe/sp._CCL068',\n",
    "             '20180108_FAH18647/barcode01': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Saccharomyces/cerevisiae',\n",
    "             '20180108_FAH18647/barcode02': 'Ascomycota/Eurotiomycetes/Chaetothyriales/Herpotrichiellaceae/Cladophialophora/sp.',\n",
    "             '20180108_FAH18647/barcode03': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/albicans',\n",
    "             '20180108_FAH18647/barcode04': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/metapsilosis',\n",
    "             '20180108_FAH18647/barcode05': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/orthopsilosis',\n",
    "             '20180108_FAH18647/barcode06': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/parapsilosis',\n",
    "                 '20180108_FAH18647/barcode07': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/sp. (Cryptococcus_gattii)',\n",
    "             '20180108_FAH18647/barcode08': 'Ascomycota/Saccharomycetes/Saccharomycetales/Dipodascaceae/Geotrichum/candidum',\n",
    "             '20180108_FAH18647/barcode09': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Kluyveromyces/lactis',\n",
    "             '20180108_FAH18647/barcode10': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Kluyveromyces/marxianus',\n",
    "             '20180108_FAH18647/barcode11': 'Ascomycota/Saccharomycetes/Saccharomycetales/Pichiaceae/Pichia/kudriavzevii',\n",
    "             '20180108_FAH18647/barcode12': 'Ascomycota/Saccharomycetes/Saccharomycetales/Pichiaceae/Pichia/membranifaciens'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../test_reads_vs_database.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "Enter the analysis directory to generate a dataframe that extracts the details\n",
    "for each sample\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "phylogeny = {'20171103_FAH15473/barcode01': 'Basidiomycota/Pucciniomycetes/Pucciniales/Pucciniaceae/Puccinia/striiformis_tritici',\n",
    "             '20171103_FAH15473/barcode02': 'Ascomycota/Dothideomycetes/Capnodiales/Mycosphaerellaceae/Zymoseptoria/tritici',\n",
    "             '20171103_FAH15473/barcode03': 'Ascomycota/Dothideomycetes/Pleosporales/Pleospraceae/Pyrenophora/tritici-repentis',\n",
    "             '20171103_FAH15473/barcode04': 'Ascomycota/Sordariomycetes/Hypocreales/Nectriaceae/Fusarium/oxysporum',\n",
    "             '20171103_FAH15473/barcode05': 'Ascomycota/Pezizomycetes/Pezizales/Tuberaceae/Tuber/brumale',\n",
    "             '20171103_FAH15473/barcode06': 'Basidiomycota/Agaricomycetes/Agaricales/Cortinariaceae/Cortinarius/globuliformis',\n",
    "             '20171103_FAH15473/barcode07': 'Ascomycota/Eurotiomycetes/Eurotiales/Trichocomaceae/Aspergillus/niger',\n",
    "             '20171103_FAH15473/barcode08': 'Ascomycota/Saccharomycetes/Saccharomycetales/Metschnikowiaceae/Clavispora/lusitaniae',\n",
    "                 '20171103_FAH15473/barcode09': 'Basidiomycota/Tremellomycetes/Tremellales/Tremellaceae/Cryptococcus/neoformans',\n",
    "             '20171103_FAH15473/barcode10': 'Ascomycota/Eurotiomycetes/Eurotiales/Trichocomaceae/Penicillium/chrysogenum',\n",
    "             '20171103_FAH15473/barcode11': 'Basidiomycota/Microbotryomycetes/Sporidiobolales/Sporidiobolaceae/Rhodotorula/mucilaginosa',\n",
    "             '20171103_FAH15473/barcode12': 'Ascomycota/Sordariomycetes/Microascales/Microascaceae/Scedosporium/boydii',\n",
    "             '20171207_FAH18654/barcode01': 'Ascomycota/Saccharomycetes/Saccharomycetales/Trichomonascaceae/Blastobotrys/proliferans',\n",
    "                 '20171207_FAH18654/barcode02': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/zeylanoides',\n",
    "                 '20171207_FAH18654/barcode03': 'Ascomycota/Saccharomycetes/Saccharomycetales/Dipodascaceae/Galactomyces/geotrichum',\n",
    "             '20171207_FAH18654/barcode04': 'Ascomycota/Saccharomycetes/Saccharomycetales/Metschnikowiaceae/Kodamaea/ohmeri',\n",
    "             '20171207_FAH18654/barcode05': 'Ascomycota/Saccharomycetes/Saccharomycetales/Debaryomycetaceae/Meyerozyma/guillermondii',\n",
    "             '20171207_FAH18654/barcode06': 'Ascomycota/Saccharomycetes/Saccharomycetales/Phaffomycetaceae/Wickerhamomyces/anomalus',\n",
    "             '20171207_FAH18654/barcode07': 'Ascomycota/Saccharomycetes/Saccharomycetales/Debaryomycetaceae/Yamadazyma/mexicana',\n",
    "             '20171207_FAH18654/barcode08': 'Ascomycota/Saccharomycetes/Saccharomycetales/Debaryomycetaceae/Yamadazyma/scolyti',\n",
    "             '20171207_FAH18654/barcode09': 'Ascomycota/Saccharomycetes/Saccharomycetales/Dipodascaceae/Yarrowia/lipolytica',\n",
    "             '20171207_FAH18654/barcode11': 'Ascomycota/Saccharomycetes/Saccharomycetales/Trichomonascaceae/Zygoascus/hellenicus',\n",
    "             '20171207_FAH18654/barcode12': 'Ascomycota/Eurotiomycetes/Eurotiales/Trichocomaceae/Aspergillus/flavus',\n",
    "             '20171212_FAH18688/barcode01': 'Basidiomycota/Tremellomycetes/Tremellales/Tremellaceae/Cryptococcus/zero',\n",
    "             '20171212_FAH18688/barcode02': 'Ascomycota/Eurotiomycetes/Eurotiales/Trichocomaceae/Aspergillus/sp.',\n",
    "                 '20171212_FAH18688/barcode03': 'Ascomycota/Sordariomycetes/Diaporthales/Diaporthaceae/Diaporthe/sp._CCL067',\n",
    "             '20171212_FAH18688/barcode04': 'Ascomycota/Sordariomycetes/Diaporthales/Diaporthaceae/Diaporthe/sp.',\n",
    "             '20171212_FAH18688/barcode05': 'Ascomycota/Leotiomycetes/Helotiales/Dermateaceae/Oculimacula/yallundae_CCL031',\n",
    "             '20171212_FAH18688/barcode06': 'Ascomycota/Leotiomycetes/Helotiales/Dermateaceae/Oculimacula/yallundae_CCL029',\n",
    "             '20171212_FAH18688/barcode07': 'Ascomycota/Dothideomycetes/Botryosphaeriales/Botryosphaeriaceae/Dothiorella/vidmadera',\n",
    "             '20171212_FAH18688/barcode08': 'Exobasidiomycetes/Microstromatales/Quambalariaceae/Quambalaria/cyanescens',\n",
    "             '20171212_FAH18688/barcode09': 'Ascomycota/Sordariomycetes/Xylariales.Xylariaceae/Entoleuca/sp.',\n",
    "                 '20171212_FAH18688/barcode11': 'Ascomycota/Sordariomycetes/Diaporthales/Diaporthaceae/Diaporthe/sp._CCL060',\n",
    "                 '20171212_FAH18688/barcode12': 'Ascomycota/Sordariomycetes/Diaporthales/Diaporthaceae/Diaporthe/sp._CCL068',\n",
    "             '20180108_FAH18647/barcode01': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Saccharomyces/cerevisiae',\n",
    "             '20180108_FAH18647/barcode02': 'Ascomycota/Eurotiomycetes/Chaetothyriales/Herpotrichiellaceae/Cladophialophora/sp.',\n",
    "             '20180108_FAH18647/barcode03': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/albicans',\n",
    "             '20180108_FAH18647/barcode04': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/metapsilosis',\n",
    "             '20180108_FAH18647/barcode05': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/orthopsilosis',\n",
    "             '20180108_FAH18647/barcode06': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/parapsilosis',\n",
    "                 '20180108_FAH18647/barcode07': 'Basidiomycota/Tremellomycetes/Tremellales/Tremellaceae/Cryptococcus/gattii',\n",
    "             '20180108_FAH18647/barcode08': 'Ascomycota/Saccharomycetes/Saccharomycetales/Dipodascaceae/Geotrichum/candidum',\n",
    "             '20180108_FAH18647/barcode09': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Kluyveromyces/lactis',\n",
    "             '20180108_FAH18647/barcode10': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Kluyveromyces/marxianus',\n",
    "             '20180108_FAH18647/barcode11': 'Ascomycota/Saccharomycetes/Saccharomycetales/Pichiaceae/Pichia/kudriavzevii',\n",
    "             '20180108_FAH18647/barcode12': 'Ascomycota/Saccharomycetes/Saccharomycetales/Pichiaceae/Pichia/membranifaciens'}\n",
    "input_file = args.input_file\n",
    "\n",
    "print(input_file+'\\n')\n",
    "\n",
    "f=open(input_file,\"r\")\n",
    "if f.mode == \"r\":\n",
    "    contents=f.read()  \n",
    "tmp=contents.replace(\"\\t\",\",\").split('\\n')\n",
    "tmp\n",
    "tmp_dict = {}\n",
    "counter=0\n",
    "\n",
    "for line in tmp[:-1]:\n",
    "    tmp_dict[counter] = str(line.split(\",\")[5].split('_')[0])\n",
    "    counter+=1\n",
    "tmp2 = pd.DataFrame.from_dict(data=tmp_dict, orient='index')\n",
    "counts = tmp2[0].value_counts()\n",
    "tmp3 = counts.to_frame().reset_index(drop=False)\n",
    "\n",
    "g=open(\"database/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt\", \"r\")\n",
    "if g.mode == \"r\":\n",
    "    contents2=g.read()\n",
    "tmp4=contents2.replace(\"\\t\",\",\").split(\"\\n\")\n",
    "\n",
    "tmp5 = []\n",
    "for i in range(0,len(tmp4)):\n",
    "    tmp5.append(tmp4[i].split(','))\n",
    "\n",
    "tmp6 = pd.DataFrame(tmp5)\n",
    "\n",
    "tmp_dict2 = {}\n",
    "for i in range(0,len(tmp6[0])):\n",
    "    tmp_dict2[tmp6[0][i].split(\"_\")[0]]=tmp6[1][i]\n",
    "\n",
    "    \n",
    "tobewritten = pd.DataFrame(data=[],columns=['Count','Species Match','Genus Match','Family Match','Order Match','Class Match','Phylum Match'])\n",
    "\n",
    "for a,b in tmp3[:7].itertuples(index=False):\n",
    "    # a == id, b == count\n",
    "    if a in tmp_dict2:\n",
    "        tmp_specs = []\n",
    "        tmp_specs.append(b)\n",
    "        if len(tmp_dict2[a].split('s__')[-1].split('_')) > 1:\n",
    "            tmp_specs.append(tmp_dict2[a].split('s__')[-1].split('_')[1])\n",
    "            tmp_specs.append(tmp_dict2[a].split('s__')[-1].split('_')[0])\n",
    "        else:\n",
    "            tmp_specs.append(tmp_dict2[a].split('s__')[-1].split('_')[0])\n",
    "            tmp_specs.append(tmp_dict2[a].split('s__')[-1].split('_')[0])\n",
    "        tmp_specs.append(tmp_dict2[a].split('s__')[-2].split('g__')[-2].split('f__')[-1].split(';')[0])\n",
    "        tmp_specs.append(tmp_dict2[a].split('s__')[-2].split('g__')[-2].split('f__')[-2].split('o__')[-1].split(';')[0])\n",
    "        tmp_specs.append(tmp_dict2[a].split('s__')[-2].split('g__')[-2].split('f__')[-2].split('o__')[-2].split('c__')[-1].split(';')[0])\n",
    "        tmp_specs.append(tmp_dict2[a].split('s__')[-2].split('g__')[-2].split('f__')[-2].split('o__')[-2].split('c__')[-2].split('p__')[-1].split(';')[0])\n",
    "        tmpf = pd.DataFrame([tmp_specs],columns=['Count','Species Match','Genus Match','Family Match','Order Match','Class Match','Phylum Match'],index=[a])\n",
    "        tobewritten = tobewritten.append(tmpf)\n",
    "\n",
    "correct_list = ['N/A', phylogeny[input_file[19:-19]].split('/')[-1].split('_')[0], phylogeny[input_file[19:-19]].split('/')[-2],phylogeny[input_file[19:-19]].split('/')[-3],phylogeny[input_file[19:-19]].split('/')[-4],phylogeny[input_file[19:-19]].split('/')[-5],phylogeny[input_file[19:-19]].split('/')[-6]]\n",
    "correct = pd.DataFrame([correct_list], columns=['Count','Species Match','Genus Match','Family Match','Order Match','Class Match','Phylum Match'],index=['Suggested'])\n",
    "tobewritten = tobewritten.append(correct)\n",
    "\n",
    "consensus = [0]*7\n",
    "count = 0\n",
    "for line in tobewritten:\n",
    "    if line != 'Count':\n",
    "        tmps = tobewritten[line].value_counts().to_frame()\n",
    "        consensus[count] = tmps.index[0]\n",
    "        count += 1\n",
    "    else:\n",
    "        consensus[count] = 'N/A'\n",
    "        count += 1\n",
    "consensus = pd.DataFrame([consensus], columns=['Count','Species Match','Genus Match','Family Match','Order Match','Class Match','Phylum Match'],index=['Consensus Match'])\n",
    "tobewritten = tobewritten.append([consensus])\n",
    "\n",
    "if (correct.values[0][1:2] == consensus.values[0][1:2]).all():\n",
    "    print('\\033[0;31m'+\"SPECIES MATCH\"+'\\033[1;37m')\n",
    "else:\n",
    "    print('\\033[0;34m'+\"NO SPECIES MATCH\"+'\\033[1;37m')\n",
    "    \n",
    "if (correct.values[0][2:3] == consensus.values[0][2:3]).all():\n",
    "    print('\\033[0;31m'+\"GENUS MATCH\"+'\\033[1;37m')\n",
    "else:\n",
    "    print('\\033[0;34m'+\"NO GENUS MATCH\"+'\\033[1;37m')\n",
    "    \n",
    "if (correct.values[0][3:4] == consensus.values[0][3:4]).all():\n",
    "    print('\\033[0;31m'+\"FAMILY MATCH\"+'\\033[1;37m')\n",
    "else:\n",
    "    print('\\033[0;34m'+\"NO FAMILY MATCH\"+'\\033[1;37m')\n",
    "    \n",
    "if (correct.values[0][4:5] == consensus.values[0][4:5]).all():\n",
    "    print('\\033[0;31m'+\"ORDER MATCH\"+'\\033[1;37m')\n",
    "else:\n",
    "    print('\\033[0;34m'+\"NO ORDER MATCH\"+'\\033[1;37m')\n",
    "    \n",
    "if (correct.values[0][5:6] == consensus.values[0][5:6]).all():\n",
    "    print('\\033[0;31m'+\"CLASS MATCH\"+'\\033[1;37m')\n",
    "else:\n",
    "    print('\\033[0;34m'+\"NO CLASS MATCH\"+'\\033[1;37m')\n",
    "    \n",
    "if (correct.values[0][6:] == consensus.values[0][6:]).all():\n",
    "    print('\\033[0;31m'+\"PHYLUM MATCH\"+'\\033[1;37m')\n",
    "else:\n",
    "    print('\\033[0;34m'+\"NO PHYLUM MATCH\"+'\\033[1;37m')\n",
    "    \n",
    "print('\\n\\n')\n",
    "\n",
    "\n",
    "tobewritten.to_csv(input_file[:-18]+'reads_matches_found.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../test_consensus_vs_database.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "Enter the analysis directory to generate a dataframe that extracts the details\n",
    "for each sample\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "phylogeny = {'20171103_FAH15473/barcode01': 'Basidiomycota/Pucciniomycetes/Pucciniales/Pucciniaceae/Puccinia/striiformis_tritici',\n",
    "             '20171103_FAH15473/barcode02': 'Ascomycota/Dothideomycetes/Capnodiales/Mycosphaerellaceae/Zymoseptoria/tritici',\n",
    "             '20171103_FAH15473/barcode03': 'Ascomycota/Dothideomycetes/Pleosporales/Pleospraceae/Pyrenophora/tritici-repentis',\n",
    "             '20171103_FAH15473/barcode04': 'Ascomycota/Sordariomycetes/Hypocreales/Nectriaceae/Fusarium/oxysporum',\n",
    "             '20171103_FAH15473/barcode05': 'Ascomycota/Pezizomycetes/Pezizales/Tuberaceae/Tuber/brumale',\n",
    "             '20171103_FAH15473/barcode06': 'Basidiomycota/Agaricomycetes/Agaricales/Cortinariaceae/Cortinarius/globuliformis',\n",
    "             '20171103_FAH15473/barcode07': 'Ascomycota/Eurotiomycetes/Eurotiales/Trichocomaceae/Aspergillus/niger',\n",
    "             '20171103_FAH15473/barcode08': 'Ascomycota/Saccharomycetes/Saccharomycetales/Metschnikowiaceae/Clavispora/lusitaniae',\n",
    "                 '20171103_FAH15473/barcode09': 'Basidiomycota/Tremellomycetes/Tremellales/Tremellaceae/Cryptococcus/neoformans',\n",
    "             '20171103_FAH15473/barcode10': 'Ascomycota/Eurotiomycetes/Eurotiales/Trichocomaceae/Penicillium/chrysogenum',\n",
    "             '20171103_FAH15473/barcode11': 'Basidiomycota/Microbotryomycetes/Sporidiobolales/Sporidiobolaceae/Rhodotorula/mucilaginosa',\n",
    "             '20171103_FAH15473/barcode12': 'Ascomycota/Sordariomycetes/Microascales/Microascaceae/Scedosporium/boydii',\n",
    "             '20171207_FAH18654/barcode01': 'Ascomycota/Saccharomycetes/Saccharomycetales/Trichomonascaceae/Blastobotrys/proliferans',\n",
    "                 '20171207_FAH18654/barcode02': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/zeylanoides',\n",
    "                 '20171207_FAH18654/barcode03': 'Ascomycota/Saccharomycetes/Saccharomycetales/Dipodascaceae/Galactomyces/geotrichum',\n",
    "             '20171207_FAH18654/barcode04': 'Ascomycota/Saccharomycetes/Saccharomycetales/Metschnikowiaceae/Kodamaea/ohmeri',\n",
    "             '20171207_FAH18654/barcode05': 'Ascomycota/Saccharomycetes/Saccharomycetales/Debaryomycetaceae/Meyerozyma/guillermondii',\n",
    "             '20171207_FAH18654/barcode06': 'Ascomycota/Saccharomycetes/Saccharomycetales/Phaffomycetaceae/Wickerhamomyces/anomalus',\n",
    "             '20171207_FAH18654/barcode07': 'Ascomycota/Saccharomycetes/Saccharomycetales/Debaryomycetaceae/Yamadazyma/mexicana',\n",
    "             '20171207_FAH18654/barcode08': 'Ascomycota/Saccharomycetes/Saccharomycetales/Debaryomycetaceae/Yamadazyma/scolyti',\n",
    "             '20171207_FAH18654/barcode09': 'Ascomycota/Saccharomycetes/Saccharomycetales/Dipodascaceae/Yarrowia/lipolytica',\n",
    "             '20171207_FAH18654/barcode11': 'Ascomycota/Saccharomycetes/Saccharomycetales/Trichomonascaceae/Zygoascus/hellenicus',\n",
    "             '20171207_FAH18654/barcode12': 'Ascomycota/Eurotiomycetes/Eurotiales/Trichocomaceae/Aspergillus/flavus',\n",
    "             '20171212_FAH18688/barcode01': 'Basidiomycota/Tremellomycetes/Tremellales/Tremellaceae/Cryptococcus/zero',\n",
    "             '20171212_FAH18688/barcode02': 'Ascomycota/Eurotiomycetes/Eurotiales/Trichocomaceae/Aspergillus/sp.',\n",
    "                 '20171212_FAH18688/barcode03': 'Ascomycota/Sordariomycetes/Diaporthales/Diaporthaceae/Diaporthe/sp._CCL067',\n",
    "             '20171212_FAH18688/barcode04': 'Ascomycota/Sordariomycetes/Diaporthales/Diaporthaceae/Diaporthe/sp.',\n",
    "             '20171212_FAH18688/barcode05': 'Ascomycota/Leotiomycetes/Helotiales/Dermateaceae/Oculimacula/yallundae_CCL031',\n",
    "             '20171212_FAH18688/barcode06': 'Ascomycota/Leotiomycetes/Helotiales/Dermateaceae/Oculimacula/yallundae_CCL029',\n",
    "             '20171212_FAH18688/barcode07': 'Ascomycota/Dothideomycetes/Botryosphaeriales/Botryosphaeriaceae/Dothiorella/vidmadera',\n",
    "             '20171212_FAH18688/barcode08': 'Exobasidiomycetes/Microstromatales/Quambalariaceae/Quambalaria/cyanescens',\n",
    "             '20171212_FAH18688/barcode09': 'Ascomycota/Sordariomycetes/Xylariales.Xylariaceae/Entoleuca/sp.',\n",
    "                 '20171212_FAH18688/barcode11': 'Ascomycota/Sordariomycetes/Diaporthales/Diaporthaceae/Diaporthe/sp._CCL060',\n",
    "                 '20171212_FAH18688/barcode12': 'Ascomycota/Sordariomycetes/Diaporthales/Diaporthaceae/Diaporthe/sp._CCL068',\n",
    "             '20180108_FAH18647/barcode01': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Saccharomyces/cerevisiae',\n",
    "             '20180108_FAH18647/barcode02': 'Ascomycota/Eurotiomycetes/Chaetothyriales/Herpotrichiellaceae/Cladophialophora/sp.',\n",
    "             '20180108_FAH18647/barcode03': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/albicans',\n",
    "             '20180108_FAH18647/barcode04': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/metapsilosis',\n",
    "             '20180108_FAH18647/barcode05': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/orthopsilosis',\n",
    "             '20180108_FAH18647/barcode06': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Candida/parapsilosis',\n",
    "                 '20180108_FAH18647/barcode07': 'Basidiomycota/Tremellomycetes/Tremellales/Tremellaceae/Cryptococcus/gattii',\n",
    "             '20180108_FAH18647/barcode08': 'Ascomycota/Saccharomycetes/Saccharomycetales/Dipodascaceae/Geotrichum/candidum',\n",
    "             '20180108_FAH18647/barcode09': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Kluyveromyces/lactis',\n",
    "             '20180108_FAH18647/barcode10': 'Ascomycota/Saccharomycetes/Saccharomycetales/Saccharomycetaceae/Kluyveromyces/marxianus',\n",
    "             '20180108_FAH18647/barcode11': 'Ascomycota/Saccharomycetes/Saccharomycetales/Pichiaceae/Pichia/kudriavzevii',\n",
    "             '20180108_FAH18647/barcode12': 'Ascomycota/Saccharomycetes/Saccharomycetales/Pichiaceae/Pichia/membranifaciens'}\n",
    "input_file = args.input_file\n",
    "\n",
    "print(input_file+'\\n')\n",
    "\n",
    "f=open(input_file,\"r\")\n",
    "if f.mode == \"r\":\n",
    "    contents=f.read()  \n",
    "tmp=contents.replace(\"\\t\",\",\").split('\\n')\n",
    "tmp\n",
    "tmp_dict = {}\n",
    "counter=0\n",
    "\n",
    "for line in tmp[:-1]:\n",
    "    tmp_dict[counter] = str(line.split(\",\")[5].split('_')[0])\n",
    "    counter+=1\n",
    "tmp2 = pd.DataFrame.from_dict(data=tmp_dict, orient='index')\n",
    "counts = tmp2[0].value_counts()\n",
    "tmp3 = counts.to_frame().reset_index(drop=False)\n",
    "\n",
    "g=open(\"database/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt\", \"r\")\n",
    "if g.mode == \"r\":\n",
    "    contents2=g.read()\n",
    "tmp4=contents2.replace(\"\\t\",\",\").split(\"\\n\")\n",
    "\n",
    "tmp5 = []\n",
    "for i in range(0,len(tmp4)):\n",
    "    tmp5.append(tmp4[i].split(','))\n",
    "\n",
    "tmp6 = pd.DataFrame(tmp5)\n",
    "\n",
    "tmp_dict2 = {}\n",
    "for i in range(0,len(tmp6[0])):\n",
    "    tmp_dict2[tmp6[0][i].split(\"_\")[0]]=tmp6[1][i]\n",
    "\n",
    "    \n",
    "tobewritten = pd.DataFrame(data=[],columns=['Count','Species Match','Genus Match','Family Match','Order Match','Class Match','Phylum Match'])\n",
    "\n",
    "for a,b in tmp3[:7].itertuples(index=False):\n",
    "    # a == id, b == count\n",
    "    if a in tmp_dict2:\n",
    "        tmp_specs = []\n",
    "        tmp_specs.append(b)\n",
    "        if len(tmp_dict2[a].split('s__')[-1].split('_')) > 1:\n",
    "            tmp_specs.append(tmp_dict2[a].split('s__')[-1].split('_')[1])\n",
    "            tmp_specs.append(tmp_dict2[a].split('s__')[-1].split('_')[0])\n",
    "        else:\n",
    "            tmp_specs.append(tmp_dict2[a].split('s__')[-1].split('_')[0])\n",
    "            tmp_specs.append(tmp_dict2[a].split('s__')[-1].split('_')[0])\n",
    "        tmp_specs.append(tmp_dict2[a].split('s__')[-2].split('g__')[-2].split('f__')[-1].split(';')[0])\n",
    "        tmp_specs.append(tmp_dict2[a].split('s__')[-2].split('g__')[-2].split('f__')[-2].split('o__')[-1].split(';')[0])\n",
    "        tmp_specs.append(tmp_dict2[a].split('s__')[-2].split('g__')[-2].split('f__')[-2].split('o__')[-2].split('c__')[-1].split(';')[0])\n",
    "        tmp_specs.append(tmp_dict2[a].split('s__')[-2].split('g__')[-2].split('f__')[-2].split('o__')[-2].split('c__')[-2].split('p__')[-1].split(';')[0])\n",
    "        tmpf = pd.DataFrame([tmp_specs],columns=['Count','Species Match','Genus Match','Family Match','Order Match','Class Match','Phylum Match'],index=[a])\n",
    "        tobewritten = tobewritten.append(tmpf)\n",
    "\n",
    "        \n",
    "pd.set_option('display.max_rows', None)\n",
    "print('\\033[0;34m')\n",
    "print(tobewritten)\n",
    "print('\\033[1;37m')       \n",
    "# correct_list = ['N/A', phylogeny[input_file[19:-33]].split('/')[-1].split('_')[0], phylogeny[input_file[19:-33]].split('/')[-2],phylogeny[input_file[19:-33]].split('/')[-3],phylogeny[input_file[19:-33]].split('/')[-4],phylogeny[input_file[19:-33]].split('/')[-5],phylogeny[input_file[19:-33]].split('/')[-6]]\n",
    "# correct = pd.DataFrame([correct_list], columns=['Count','Species Match','Genus Match','Family Match','Order Match','Class Match','Phylum Match'],index=['Suggested'])\n",
    "\n",
    "# tps1 = correct[['Species Match']]\n",
    "# tps2 = tobewritten[['Species Match']]\n",
    "# tpg1 = correct[['Genus Match']]\n",
    "# tpg2 = tobewritten[['Genus Match']]\n",
    "# print(pd.merge(tps1,tps2,on='Species Match'))\n",
    "# if pd.merge(tps1,tps2,on='Species Match').any().any():\n",
    "#     print('\\033[0;34m')\n",
    "#     print(pd.merge(tps1,tps2,on='Species Match').any())\n",
    "# else:\n",
    "#     print('\\033[0;31m')\n",
    "#     print(pd.merge(tps1,tps2,on='Species Match').any())\n",
    "# print('\\033[1;37m')\n",
    "\n",
    "# print(pd.merge(tpg1,tpg2,on='Genus Match'))\n",
    "# if pd.merge(tpg1,tpg2,on='Genus Match').any().any():\n",
    "#     print('\\033[0;34m')\n",
    "#     print(pd.merge(tpg1,tpg2,on='Genus Match').any())\n",
    "# else:\n",
    "#     print('\\033[0;31m')\n",
    "#     print(pd.merge(tpg1,tpg2,on='Genus Match').any())\n",
    "# print('\\033[1;37m')\n",
    "\n",
    "\n",
    "\n",
    "# consensus = [0]*7\n",
    "# count = 0\n",
    "# total = 0\n",
    "# match = 0\n",
    "# for line in tobewritten:\n",
    "#     if line != 'Count':\n",
    "#         tmps = tobewritten[line].value_counts().to_frame()\n",
    "#         if tmps.index[0] in correct_list:\n",
    "#             consensus[count] = tmps.index[0]\n",
    "#             print(consensus[count])\n",
    "#         else:\n",
    "#             print(\"NO\")\n",
    "#         count += 1\n",
    "#     else:\n",
    "#         consensus[count] = 'N/A'\n",
    "#         count += 1\n",
    "\n",
    "        \n",
    "# correct = pd.DataFrame([correct_list], columns=['Count','Species Match','Genus Match','Family Match','Order Match','Class Match','Phylum Match'],index=['Suggested'])\n",
    "# tobewritten = tobewritten.append(correct)\n",
    "\n",
    "# consensus = pd.DataFrame([consensus], columns=['Count','Species Match','Genus Match','Family Match','Order Match','Class Match','Phylum Match'],index=['Consensus Match'])\n",
    "# tobewritten = tobewritten.append([consensus])\n",
    "\n",
    "# if (correct.values[0][1:2] == consensus.values[0][1:2]).all():\n",
    "#     print('\\033[0;31m'+\"SPECIES MATCH\"+'\\033[1;37m')\n",
    "# else:\n",
    "#     print('\\033[0;34m'+\"NO SPECIES MATCH\"+'\\033[1;37m')\n",
    "    \n",
    "# if (correct.values[0][2:3] == consensus.values[0][2:3]).all():\n",
    "#     print('\\033[0;31m'+\"GENUS MATCH\"+'\\033[1;37m')\n",
    "# else:\n",
    "#     print('\\033[0;34m'+\"NO GENUS MATCH\"+'\\033[1;37m')\n",
    "    \n",
    "# if (correct.values[0][3:4] == consensus.values[0][3:4]).all():\n",
    "#     print('\\033[0;31m'+\"FAMILY MATCH\"+'\\033[1;37m')\n",
    "# else:\n",
    "#     print('\\033[0;34m'+\"NO FAMILY MATCH\"+'\\033[1;37m')\n",
    "    \n",
    "# if (correct.values[0][4:5] == consensus.values[0][4:5]).all():\n",
    "#     print('\\033[0;31m'+\"ORDER MATCH\"+'\\033[1;37m')\n",
    "# else:\n",
    "#     print('\\033[0;34m'+\"NO ORDER MATCH\"+'\\033[1;37m')\n",
    "    \n",
    "# if (correct.values[0][5:6] == consensus.values[0][5:6]).all():\n",
    "#     print('\\033[0;31m'+\"CLASS MATCH\"+'\\033[1;37m')\n",
    "# else:\n",
    "#     print('\\033[0;34m'+\"NO CLASS MATCH\"+'\\033[1;37m')\n",
    "    \n",
    "# if (correct.values[0][6:] == consensus.values[0][6:]).all():\n",
    "#     print('\\033[0;31m'+\"PHYLUM MATCH\"+'\\033[1;37m')\n",
    "# else:\n",
    "#     print('\\033[0;34m'+\"NO PHYLUM MATCH\"+'\\033[1;37m')\n",
    "    \n",
    "# print('\\n\\n')\n",
    "\n",
    "\n",
    "# tobewritten.to_csv(input_file[:-32]+'matches_found.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../overall_statistics.py\n",
    "\n",
    "\"\"\"\n",
    "The goal of this program is to examine the distribution of reads within\n",
    "each file, and for the files generated after homology analysis.\n",
    "The program will generate summary statistics for the result of the homology\n",
    "analysis and save figures illustrating the read distribution for the\n",
    "frDNA reads\n",
    "\"\"\"\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import argparse\n",
    "import csv\n",
    "\n",
    "\n",
    "WHITE='\\033[1;37m'\n",
    "BLUE='\\033[0;34m'\n",
    "RED='\\033[0;31m'\n",
    "NC='\\033[0m'\n",
    "GREEN='\\033[0;32m'\n",
    "PURPLE='\\033[1;35m'\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "The goal of this program is to examine the distribution of reads within\n",
    "each file, and for the files generated after homology analysis.\n",
    "The program will generate summary statistics for the result of the homology\n",
    "analysis and save figures illustrating the read distribution for the\n",
    "frDNA reads\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"full_file\")\n",
    "args = parser.parse_args()\n",
    "print('\\033[0;35m'+'START'+'\\033[1;37m')\n",
    "full_file = args.full_file\n",
    "print(full_file)\n",
    "homology_folder = args.full_file[:9]+'Python_Processing'+args.full_file[21:-12]+'combined_test.paf'\n",
    "print(homology_folder)\n",
    "length_folder = args.full_file[:9]+'Length_Filtered'+args.full_file[21:-12]+'length_restricted_reads.fasta'\n",
    "print(length_folder)\n",
    "\n",
    "# Load the full file containing all reads for this barcode\n",
    "full_file_dict = SeqIO.to_dict(SeqIO.parse(full_file, \"fastq\"))\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Loaded \" + full_file + '\\033[1;37m')\n",
    "\n",
    "    \n",
    "# Open the original merged.fastq file\n",
    "# Extract the number of reads contained within this file\n",
    "full_lengths = []\n",
    "for key in full_file_dict:\n",
    "    full_lengths.append(len(full_file_dict[key].seq))\n",
    "full_lengths_len = len(full_file_dict)\n",
    "print(full_lengths_len)\n",
    "\n",
    "\n",
    "# Import the PAF file resulting from the minimap2 homology filtering\n",
    "homology_paf = pd.read_csv(homology_folder, sep='\\t', header=None, engine='python')\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Loaded \" + homology_folder + '\\033[1;37m')\n",
    "          \n",
    "# Create new dictionary using only keys present in paf file\n",
    "# Extract the number of reads contained within this file\n",
    "homology_dict = {}\n",
    "for key in homology_paf[0].unique():\n",
    "    homology_dict[key] = full_file_dict[key]\n",
    "\n",
    "homology_lengths = []\n",
    "for key in homology_dict:\n",
    "          homology_lengths.append(len(homology_dict[key].seq))\n",
    "homology_lengths_len = len(homology_lengths)\n",
    "print(homology_lengths_len)\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Loaded \" + length_folder + '\\033[1;37m')\n",
    "    \n",
    "# Open length restricted fasta\n",
    "# Extract the number of reads contained within this file\n",
    "length_dict = SeqIO.to_dict(SeqIO.parse(length_folder, \"fasta\"))\n",
    "\n",
    "length_lengths = []\n",
    "for key in length_dict:\n",
    "    length_lengths.append(len(length_dict[key].seq))\n",
    "length_lengths_len = len(length_dict)\n",
    "print(length_lengths_len)\n",
    "\n",
    "number_of_reads = pd.DataFrame([[full_lengths_len, homology_lengths_len, length_lengths_len]], columns=['Original # reads','# reads after homology filtering','# reads after length filtering'])\n",
    "number_of_reads.to_csv(args.full_file[:9]+'Stats'+args.full_file[21:-12]+'read_numbers.csv', index=False)\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;32m' + \"Number of reads file saved to \" + args.full_file[:9]+'Stats'+args.full_file[21:-12]+'read_numbers.csv' + '\\033[1;37m')\n",
    "\n",
    "fields = [full_lengths_len, homology_lengths_len, length_lengths_len]\n",
    "if args.full_file[40:-13] != 'unclassified':\n",
    "    if '20171212_FAH18688/barcode10' not in args.full_file[22:-13] and '20171207_FAH18654/barcode10' not in args.full_file[22:-13]:\n",
    "        with open(args.full_file[:9]+'Stats/read_numbers.csv','a+') as fd:\n",
    "            writer = csv.writer(fd)\n",
    "            writer.writerow(fields)\n",
    "print('\\033[0;35m'+'END'+'\\033[1;37m')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "number_stats = pd.read_csv('../../analysis/Stats/read_number.csv', header=None, engine='python')\n",
    "print(number_stats)\n",
    "print(np.mean(number_stats[0]))\n",
    "print(np.mean(number_stats[1]))\n",
    "print(min(number_stats[2]))\n",
    "print(max(number_stats[2]))\n",
    "print(np.mean(number_stats[2]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(0,len(tmp['read_id'][1:])):\n",
    "    tmps = pd.DataFrame(data=[tmp['read_id'][i+1:],tmp['mean_qscore_template'][i+1:]])\n",
    "    id_scores.append(tmps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = pd.read_csv(\"../../analysis/Basecalled/20171103_FAH15473/barcode01/sequencing_summary.txt\",sep='\\t', header=None, names=['filename', 'read_id', 'run_id', 'batch_id', 'channel', 'mux', 'start_time', 'duration', 'num_events', 'passes_filtering', 'template_start', 'num_events_template', 'template_duration', 'sequence_length_template', 'mean_qscore_template', 'strand_score_template', 'median_template', 'mad_template'], engine='python')\n",
    "tmpd = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Length_Filtered/20171103_FAH15473/barcode01/length_restricted_reads.fasta\", \"fasta\"))\n",
    "keys = []\n",
    "for key in tmpd:\n",
    "    keys.append(key)\n",
    "id_score = pd.DataFrame(data=None,columns=['read_id','mean_qscore_template'])\n",
    "tmp_keys = {}\n",
    "for i in range(1,len(tmp)):\n",
    "    tmp_keys[tmp['read_id'][i]] = tmp['mean_qscore_template'][i]\n",
    "for key in keys:\n",
    "    if key in tmp_keys:\n",
    "        tmpf = pd.DataFrame(data=[[key,tmp_keys[key]]],columns=['read_id','mean_qscore_template'])\n",
    "    id_score = id_score.append(tmpf)\n",
    "id_score = id_score.sort_values(by=['mean_qscore_template'])\n",
    "new_ids_list = id_score['read_id'][:100]\n",
    "\n",
    "new_dict = {}\n",
    "for key in new_ids_list:\n",
    "    new_dict[key] = tmpd[key]\n",
    "SeqIO.write(new_dict.values(), \"../../analysis/Consensus/20171103_FAH15473/barcode01/high_qscore_100.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../for_consensus_high_qscore_100.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from mothur_py import Mothur\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "WHITE='\\033[1;37m'\n",
    "BLUE='\\033[0;34m'\n",
    "RED='\\033[0;31m'\n",
    "NC='\\033[0m'\n",
    "GREEN='\\033[0;32m'\n",
    "PURPLE='\\033[1;35m'\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "Input the sequencing_summary.txt file from Basecalled\n",
    "Read in the length_restricted_reads.fasta from Length_Filtered\n",
    "Save to Consensus\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;31m' + \"Input file is \" + args.input_file + '\\033[1;37m')\n",
    "\n",
    "    \n",
    "tmp = pd.read_csv(args.input_file,sep='\\t', header=None, names=['filename', 'read_id', 'run_id', 'batch_id', 'channel', 'mux', 'start_time', 'duration', 'num_events', 'passes_filtering', 'template_start', 'num_events_template', 'template_duration', 'sequence_length_template', 'mean_qscore_template', 'strand_score_template', 'median_template', 'mad_template'], engine='python')\n",
    "tmpd = SeqIO.to_dict(SeqIO.parse(\"Length_Filtered\"+args.input_file[10:-22]+\"length_restricted_reads.fasta\", \"fasta\"))\n",
    "keys = []\n",
    "for key in tmpd:\n",
    "    keys.append(key)\n",
    "id_score = pd.DataFrame(data=None,columns=['read_id','mean_qscore_template'])\n",
    "tmp_keys = {}\n",
    "for i in range(1,len(tmp)):\n",
    "    tmp_keys[tmp['read_id'][i]] = tmp['mean_qscore_template'][i]\n",
    "for key in keys:\n",
    "    if key in tmp_keys:\n",
    "        tmpf = pd.DataFrame(data=[[key,tmp_keys[key]]],columns=['read_id','mean_qscore_template'])\n",
    "    id_score = id_score.append(tmpf)\n",
    "id_score = id_score.sort_values(by=['mean_qscore_template'])\n",
    "new_ids_list = id_score['read_id'][:100]\n",
    "\n",
    "new_dict = {}\n",
    "for key in new_ids_list:\n",
    "    new_dict[key] = tmpd[key]\n",
    "SeqIO.write(new_dict.values(), \"Consensus\"+args.input_file[10:-22]+\"high_qscore_100.fasta\", \"fasta\")\n",
    "    \n",
    "\n",
    "if args.verbose:\n",
    "    print('\\033[0;34m' + \"Reads saved to \" + '\\033[0;35m' + \"Consensus\"+args.input_file[10:-22]+\"high_qscore_100.fasta\" + '\\033[1;37m')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map genetic distance on x vs crossmapping percentage on y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylogeny = {'20171103_FAH15473/barcode01': 'Puccinia_striiformis_tritici',\n",
    "             '20171103_FAH15473/barcode02': 'Zymoseptoria_tritici',\n",
    "             '20171103_FAH15473/barcode03': 'Pyrenophora_tritici-repentis',\n",
    "             '20171103_FAH15473/barcode04': 'Fusarium_oxysporum',\n",
    "             '20171103_FAH15473/barcode05': 'Tuber_brumale',\n",
    "             '20171103_FAH15473/barcode06': 'Cortinarius_globuliformis',\n",
    "             '20171103_FAH15473/barcode07': 'Aspergillus_niger',\n",
    "             '20171103_FAH15473/barcode08': 'Clavispora_lusitaniae',\n",
    "                 '20171103_FAH15473/barcode09': 'Cryptococcus_neoformans',\n",
    "             '20171103_FAH15473/barcode10': 'Penicillium_chrysogenum',\n",
    "             '20171103_FAH15473/barcode11': 'Rhodotorula_mucilaginosa',\n",
    "             '20171103_FAH15473/barcode12': 'Scedosporium_boydii',\n",
    "             '20171207_FAH18654/barcode01': 'Blastobotrys_proliferans',\n",
    "                 '20171207_FAH18654/barcode02': 'Candida_zeylanoides',\n",
    "                 '20171207_FAH18654/barcode03': 'Galactomyces_geotrichum',\n",
    "             '20171207_FAH18654/barcode04': 'Kodamaea_ohmeri',\n",
    "             '20171207_FAH18654/barcode05': 'Meyerozyma_guillermondii',\n",
    "             '20171207_FAH18654/barcode06': 'Wickerhamomyces_anomalus',\n",
    "             '20171207_FAH18654/barcode07': 'Yamadazyma_mexicana',\n",
    "             '20171207_FAH18654/barcode08': 'Yamadazyma_scolyti',\n",
    "             '20171207_FAH18654/barcode09': 'Yarrowia_lipolytica',\n",
    "             '20171207_FAH18654/barcode11': 'Zygoascus_hellenicus',\n",
    "             '20171207_FAH18654/barcode12': 'Aspergillus_flavus',\n",
    "             '20171212_FAH18688/barcode01': 'Cryptococcus_zero',\n",
    "             '20171212_FAH18688/barcode02': 'Aspergillus_sp.',\n",
    "                 '20171212_FAH18688/barcode03': 'CCL067',\n",
    "             '20171212_FAH18688/barcode04': 'Diaporthe_sp.',\n",
    "             '20171212_FAH18688/barcode05': 'Tapesia_yallundae_CCL031',\n",
    "             '20171212_FAH18688/barcode06': 'Tapesia_yallundae_CCL029',\n",
    "             '20171212_FAH18688/barcode07': 'Dothiorella_vidmadera',\n",
    "             '20171212_FAH18688/barcode08': 'Quambalaria_cyanescens',\n",
    "             '20171212_FAH18688/barcode09': 'Entoleuca_sp.',\n",
    "                 '20171212_FAH18688/barcode11': 'CCL060',\n",
    "                 '20171212_FAH18688/barcode12': 'CCL068',\n",
    "             '20180108_FAH18647/barcode01': 'Saccharomyces_cerevisiae',\n",
    "             '20180108_FAH18647/barcode02': 'Cladophialophora_sp.',\n",
    "             '20180108_FAH18647/barcode03': 'Candida_albicans',\n",
    "             '20180108_FAH18647/barcode04': 'Candida_metapsilosis',\n",
    "             '20180108_FAH18647/barcode05': 'Candida_orthopsilosis',\n",
    "             '20180108_FAH18647/barcode06': 'Candida_parapsilosis',\n",
    "                 '20180108_FAH18647/barcode07': 'Cryptococcus_gattii',\n",
    "             '20180108_FAH18647/barcode08': 'Geotrichum_candidum',\n",
    "             '20180108_FAH18647/barcode09': 'Kluyveromyces_lactis',\n",
    "             '20180108_FAH18647/barcode10': 'Kluyveromyces_marxianus',\n",
    "             '20180108_FAH18647/barcode11': 'Pichia_kudriavzevii',\n",
    "             '20180108_FAH18647/barcode12': 'Pichia_membranifaciens'}\n",
    "\n",
    "large_frame = pd.read_csv(\"../../analysis/Alignment/database.csv\", header=0, index_col=0, engine='python')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Puccinia_striiformis_tritici</th>\n",
       "      <th>Geotrichum_candidum</th>\n",
       "      <th>Rhodotorula_mucilaginosa</th>\n",
       "      <th>Cortinarius_globuliformis</th>\n",
       "      <th>Cryptococcus_zero</th>\n",
       "      <th>Quambalaria_cyanescens</th>\n",
       "      <th>Yarrowia_lipolytica</th>\n",
       "      <th>Tuber_brumale</th>\n",
       "      <th>Diaporthe_sp.</th>\n",
       "      <th>CCL060</th>\n",
       "      <th>...</th>\n",
       "      <th>Candida_albicans</th>\n",
       "      <th>Kluyveromyces_lactis</th>\n",
       "      <th>Cryptococcus_gattii</th>\n",
       "      <th>Candida_parapsilosis</th>\n",
       "      <th>Candida_metapsilosis</th>\n",
       "      <th>Candida_orthopsilosis</th>\n",
       "      <th>Yamadazyma_mexicana</th>\n",
       "      <th>Yamadazyma_scolyti</th>\n",
       "      <th>Candida_zeylanoides</th>\n",
       "      <th>Meyerozyma_guillermondii</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Puccinia_striiformis_tritici</td>\n",
       "      <td>NaN</td>\n",
       "      <td>63.080</td>\n",
       "      <td>79.119</td>\n",
       "      <td>77.724</td>\n",
       "      <td>78.490</td>\n",
       "      <td>78.029</td>\n",
       "      <td>67.484</td>\n",
       "      <td>74.459</td>\n",
       "      <td>72.815</td>\n",
       "      <td>72.503</td>\n",
       "      <td>...</td>\n",
       "      <td>74.129</td>\n",
       "      <td>74.597</td>\n",
       "      <td>72.674</td>\n",
       "      <td>75.689</td>\n",
       "      <td>75.569</td>\n",
       "      <td>75.603</td>\n",
       "      <td>74.991</td>\n",
       "      <td>74.607</td>\n",
       "      <td>76.313</td>\n",
       "      <td>75.231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Geotrichum_candidum</td>\n",
       "      <td>63.080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>64.659</td>\n",
       "      <td>65.081</td>\n",
       "      <td>65.344</td>\n",
       "      <td>64.476</td>\n",
       "      <td>67.220</td>\n",
       "      <td>65.202</td>\n",
       "      <td>64.461</td>\n",
       "      <td>64.048</td>\n",
       "      <td>...</td>\n",
       "      <td>67.026</td>\n",
       "      <td>67.211</td>\n",
       "      <td>66.369</td>\n",
       "      <td>67.898</td>\n",
       "      <td>67.352</td>\n",
       "      <td>67.294</td>\n",
       "      <td>68.555</td>\n",
       "      <td>68.902</td>\n",
       "      <td>67.022</td>\n",
       "      <td>66.816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Rhodotorula_mucilaginosa</td>\n",
       "      <td>79.119</td>\n",
       "      <td>64.659</td>\n",
       "      <td>NaN</td>\n",
       "      <td>79.211</td>\n",
       "      <td>79.896</td>\n",
       "      <td>80.212</td>\n",
       "      <td>68.029</td>\n",
       "      <td>75.752</td>\n",
       "      <td>74.743</td>\n",
       "      <td>76.012</td>\n",
       "      <td>...</td>\n",
       "      <td>76.862</td>\n",
       "      <td>75.984</td>\n",
       "      <td>73.566</td>\n",
       "      <td>76.588</td>\n",
       "      <td>76.513</td>\n",
       "      <td>76.778</td>\n",
       "      <td>77.567</td>\n",
       "      <td>76.897</td>\n",
       "      <td>76.831</td>\n",
       "      <td>77.115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cortinarius_globuliformis</td>\n",
       "      <td>77.724</td>\n",
       "      <td>65.081</td>\n",
       "      <td>79.211</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.495</td>\n",
       "      <td>81.594</td>\n",
       "      <td>68.569</td>\n",
       "      <td>77.239</td>\n",
       "      <td>75.299</td>\n",
       "      <td>75.094</td>\n",
       "      <td>...</td>\n",
       "      <td>75.974</td>\n",
       "      <td>76.324</td>\n",
       "      <td>73.562</td>\n",
       "      <td>76.863</td>\n",
       "      <td>75.658</td>\n",
       "      <td>76.291</td>\n",
       "      <td>76.726</td>\n",
       "      <td>76.553</td>\n",
       "      <td>77.186</td>\n",
       "      <td>77.125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cryptococcus_zero</td>\n",
       "      <td>78.490</td>\n",
       "      <td>65.344</td>\n",
       "      <td>79.896</td>\n",
       "      <td>80.495</td>\n",
       "      <td>NaN</td>\n",
       "      <td>81.750</td>\n",
       "      <td>68.776</td>\n",
       "      <td>76.334</td>\n",
       "      <td>75.360</td>\n",
       "      <td>75.351</td>\n",
       "      <td>...</td>\n",
       "      <td>77.435</td>\n",
       "      <td>77.648</td>\n",
       "      <td>74.179</td>\n",
       "      <td>76.547</td>\n",
       "      <td>76.704</td>\n",
       "      <td>76.610</td>\n",
       "      <td>77.659</td>\n",
       "      <td>77.304</td>\n",
       "      <td>77.622</td>\n",
       "      <td>76.724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Quambalaria_cyanescens</td>\n",
       "      <td>78.029</td>\n",
       "      <td>64.476</td>\n",
       "      <td>80.212</td>\n",
       "      <td>81.594</td>\n",
       "      <td>81.750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>68.664</td>\n",
       "      <td>77.296</td>\n",
       "      <td>76.607</td>\n",
       "      <td>75.974</td>\n",
       "      <td>...</td>\n",
       "      <td>76.863</td>\n",
       "      <td>76.878</td>\n",
       "      <td>73.283</td>\n",
       "      <td>77.178</td>\n",
       "      <td>76.939</td>\n",
       "      <td>76.920</td>\n",
       "      <td>77.117</td>\n",
       "      <td>78.259</td>\n",
       "      <td>77.945</td>\n",
       "      <td>77.801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Yarrowia_lipolytica</td>\n",
       "      <td>67.484</td>\n",
       "      <td>67.220</td>\n",
       "      <td>68.029</td>\n",
       "      <td>68.569</td>\n",
       "      <td>68.776</td>\n",
       "      <td>68.664</td>\n",
       "      <td>NaN</td>\n",
       "      <td>68.829</td>\n",
       "      <td>68.166</td>\n",
       "      <td>68.332</td>\n",
       "      <td>...</td>\n",
       "      <td>72.129</td>\n",
       "      <td>72.986</td>\n",
       "      <td>69.777</td>\n",
       "      <td>72.280</td>\n",
       "      <td>72.393</td>\n",
       "      <td>72.178</td>\n",
       "      <td>71.394</td>\n",
       "      <td>71.414</td>\n",
       "      <td>71.718</td>\n",
       "      <td>71.261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Tuber_brumale</td>\n",
       "      <td>74.459</td>\n",
       "      <td>65.202</td>\n",
       "      <td>75.752</td>\n",
       "      <td>77.239</td>\n",
       "      <td>76.334</td>\n",
       "      <td>77.296</td>\n",
       "      <td>68.829</td>\n",
       "      <td>NaN</td>\n",
       "      <td>78.803</td>\n",
       "      <td>78.469</td>\n",
       "      <td>...</td>\n",
       "      <td>79.386</td>\n",
       "      <td>78.696</td>\n",
       "      <td>75.352</td>\n",
       "      <td>79.780</td>\n",
       "      <td>79.808</td>\n",
       "      <td>81.128</td>\n",
       "      <td>80.711</td>\n",
       "      <td>79.352</td>\n",
       "      <td>81.909</td>\n",
       "      <td>82.085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Diaporthe_sp.</td>\n",
       "      <td>72.815</td>\n",
       "      <td>64.461</td>\n",
       "      <td>74.743</td>\n",
       "      <td>75.299</td>\n",
       "      <td>75.360</td>\n",
       "      <td>76.607</td>\n",
       "      <td>68.166</td>\n",
       "      <td>78.803</td>\n",
       "      <td>NaN</td>\n",
       "      <td>94.821</td>\n",
       "      <td>...</td>\n",
       "      <td>77.279</td>\n",
       "      <td>77.218</td>\n",
       "      <td>72.843</td>\n",
       "      <td>76.362</td>\n",
       "      <td>77.340</td>\n",
       "      <td>76.422</td>\n",
       "      <td>77.375</td>\n",
       "      <td>77.260</td>\n",
       "      <td>77.747</td>\n",
       "      <td>77.766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CCL060</td>\n",
       "      <td>72.503</td>\n",
       "      <td>64.048</td>\n",
       "      <td>76.012</td>\n",
       "      <td>75.094</td>\n",
       "      <td>75.351</td>\n",
       "      <td>75.974</td>\n",
       "      <td>68.332</td>\n",
       "      <td>78.469</td>\n",
       "      <td>94.821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>77.587</td>\n",
       "      <td>76.831</td>\n",
       "      <td>73.046</td>\n",
       "      <td>76.380</td>\n",
       "      <td>76.003</td>\n",
       "      <td>76.080</td>\n",
       "      <td>77.375</td>\n",
       "      <td>76.781</td>\n",
       "      <td>77.414</td>\n",
       "      <td>77.425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CCL067</td>\n",
       "      <td>72.908</td>\n",
       "      <td>63.892</td>\n",
       "      <td>74.776</td>\n",
       "      <td>75.531</td>\n",
       "      <td>75.840</td>\n",
       "      <td>76.197</td>\n",
       "      <td>68.056</td>\n",
       "      <td>78.481</td>\n",
       "      <td>88.404</td>\n",
       "      <td>87.764</td>\n",
       "      <td>...</td>\n",
       "      <td>78.258</td>\n",
       "      <td>77.097</td>\n",
       "      <td>72.546</td>\n",
       "      <td>76.406</td>\n",
       "      <td>76.424</td>\n",
       "      <td>76.604</td>\n",
       "      <td>77.026</td>\n",
       "      <td>76.442</td>\n",
       "      <td>78.013</td>\n",
       "      <td>77.409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CCL068</td>\n",
       "      <td>72.544</td>\n",
       "      <td>64.029</td>\n",
       "      <td>74.484</td>\n",
       "      <td>76.000</td>\n",
       "      <td>75.445</td>\n",
       "      <td>76.728</td>\n",
       "      <td>68.341</td>\n",
       "      <td>78.906</td>\n",
       "      <td>87.285</td>\n",
       "      <td>87.681</td>\n",
       "      <td>...</td>\n",
       "      <td>77.163</td>\n",
       "      <td>77.063</td>\n",
       "      <td>73.040</td>\n",
       "      <td>77.113</td>\n",
       "      <td>76.719</td>\n",
       "      <td>77.121</td>\n",
       "      <td>77.310</td>\n",
       "      <td>77.076</td>\n",
       "      <td>78.093</td>\n",
       "      <td>77.858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Scedosporium_boydii</td>\n",
       "      <td>70.945</td>\n",
       "      <td>63.356</td>\n",
       "      <td>73.597</td>\n",
       "      <td>73.913</td>\n",
       "      <td>74.434</td>\n",
       "      <td>75.411</td>\n",
       "      <td>68.485</td>\n",
       "      <td>78.386</td>\n",
       "      <td>82.697</td>\n",
       "      <td>82.715</td>\n",
       "      <td>...</td>\n",
       "      <td>77.015</td>\n",
       "      <td>76.652</td>\n",
       "      <td>72.299</td>\n",
       "      <td>75.914</td>\n",
       "      <td>75.461</td>\n",
       "      <td>75.846</td>\n",
       "      <td>76.119</td>\n",
       "      <td>76.621</td>\n",
       "      <td>76.374</td>\n",
       "      <td>77.523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Fusarium_oxysporum</td>\n",
       "      <td>72.677</td>\n",
       "      <td>64.263</td>\n",
       "      <td>74.332</td>\n",
       "      <td>74.710</td>\n",
       "      <td>75.154</td>\n",
       "      <td>76.145</td>\n",
       "      <td>68.648</td>\n",
       "      <td>79.192</td>\n",
       "      <td>86.701</td>\n",
       "      <td>84.374</td>\n",
       "      <td>...</td>\n",
       "      <td>78.473</td>\n",
       "      <td>78.404</td>\n",
       "      <td>73.535</td>\n",
       "      <td>77.568</td>\n",
       "      <td>78.633</td>\n",
       "      <td>77.610</td>\n",
       "      <td>77.793</td>\n",
       "      <td>78.025</td>\n",
       "      <td>78.449</td>\n",
       "      <td>79.252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Entoleuca_sp.</td>\n",
       "      <td>72.500</td>\n",
       "      <td>65.103</td>\n",
       "      <td>74.512</td>\n",
       "      <td>75.428</td>\n",
       "      <td>76.150</td>\n",
       "      <td>76.666</td>\n",
       "      <td>70.302</td>\n",
       "      <td>79.241</td>\n",
       "      <td>85.276</td>\n",
       "      <td>84.700</td>\n",
       "      <td>...</td>\n",
       "      <td>77.763</td>\n",
       "      <td>78.726</td>\n",
       "      <td>74.281</td>\n",
       "      <td>76.789</td>\n",
       "      <td>77.118</td>\n",
       "      <td>77.129</td>\n",
       "      <td>77.561</td>\n",
       "      <td>78.557</td>\n",
       "      <td>78.491</td>\n",
       "      <td>78.374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pyrenophora_tritici-repentis</td>\n",
       "      <td>73.861</td>\n",
       "      <td>64.382</td>\n",
       "      <td>76.762</td>\n",
       "      <td>76.968</td>\n",
       "      <td>76.712</td>\n",
       "      <td>77.015</td>\n",
       "      <td>69.081</td>\n",
       "      <td>80.643</td>\n",
       "      <td>80.247</td>\n",
       "      <td>80.192</td>\n",
       "      <td>...</td>\n",
       "      <td>80.316</td>\n",
       "      <td>78.462</td>\n",
       "      <td>74.570</td>\n",
       "      <td>78.547</td>\n",
       "      <td>78.853</td>\n",
       "      <td>78.814</td>\n",
       "      <td>79.275</td>\n",
       "      <td>79.486</td>\n",
       "      <td>79.980</td>\n",
       "      <td>79.952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cladophialophora_sp.</td>\n",
       "      <td>73.411</td>\n",
       "      <td>65.086</td>\n",
       "      <td>75.514</td>\n",
       "      <td>76.376</td>\n",
       "      <td>76.797</td>\n",
       "      <td>77.485</td>\n",
       "      <td>69.631</td>\n",
       "      <td>81.128</td>\n",
       "      <td>79.664</td>\n",
       "      <td>79.520</td>\n",
       "      <td>...</td>\n",
       "      <td>78.884</td>\n",
       "      <td>78.359</td>\n",
       "      <td>74.829</td>\n",
       "      <td>77.842</td>\n",
       "      <td>78.025</td>\n",
       "      <td>78.271</td>\n",
       "      <td>78.559</td>\n",
       "      <td>78.535</td>\n",
       "      <td>80.827</td>\n",
       "      <td>79.097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Penicillium_chrysogenum</td>\n",
       "      <td>72.631</td>\n",
       "      <td>64.829</td>\n",
       "      <td>75.694</td>\n",
       "      <td>76.342</td>\n",
       "      <td>77.013</td>\n",
       "      <td>77.523</td>\n",
       "      <td>69.675</td>\n",
       "      <td>80.636</td>\n",
       "      <td>80.323</td>\n",
       "      <td>79.684</td>\n",
       "      <td>...</td>\n",
       "      <td>79.541</td>\n",
       "      <td>79.239</td>\n",
       "      <td>73.807</td>\n",
       "      <td>78.044</td>\n",
       "      <td>78.383</td>\n",
       "      <td>78.249</td>\n",
       "      <td>78.481</td>\n",
       "      <td>79.444</td>\n",
       "      <td>79.343</td>\n",
       "      <td>79.829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Aspergillus_flavus</td>\n",
       "      <td>72.332</td>\n",
       "      <td>65.153</td>\n",
       "      <td>75.778</td>\n",
       "      <td>76.434</td>\n",
       "      <td>76.821</td>\n",
       "      <td>77.675</td>\n",
       "      <td>69.270</td>\n",
       "      <td>80.799</td>\n",
       "      <td>80.898</td>\n",
       "      <td>80.815</td>\n",
       "      <td>...</td>\n",
       "      <td>79.294</td>\n",
       "      <td>79.274</td>\n",
       "      <td>74.314</td>\n",
       "      <td>78.481</td>\n",
       "      <td>78.417</td>\n",
       "      <td>78.598</td>\n",
       "      <td>78.946</td>\n",
       "      <td>79.911</td>\n",
       "      <td>79.678</td>\n",
       "      <td>80.377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Aspergillus_niger</td>\n",
       "      <td>72.005</td>\n",
       "      <td>64.984</td>\n",
       "      <td>75.771</td>\n",
       "      <td>76.384</td>\n",
       "      <td>76.386</td>\n",
       "      <td>77.341</td>\n",
       "      <td>69.654</td>\n",
       "      <td>80.863</td>\n",
       "      <td>80.779</td>\n",
       "      <td>80.929</td>\n",
       "      <td>...</td>\n",
       "      <td>78.807</td>\n",
       "      <td>78.733</td>\n",
       "      <td>74.966</td>\n",
       "      <td>78.364</td>\n",
       "      <td>78.086</td>\n",
       "      <td>78.596</td>\n",
       "      <td>79.740</td>\n",
       "      <td>79.142</td>\n",
       "      <td>79.446</td>\n",
       "      <td>80.212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Aspergillus_sp.</td>\n",
       "      <td>71.844</td>\n",
       "      <td>64.345</td>\n",
       "      <td>75.317</td>\n",
       "      <td>75.718</td>\n",
       "      <td>76.318</td>\n",
       "      <td>77.273</td>\n",
       "      <td>69.699</td>\n",
       "      <td>80.150</td>\n",
       "      <td>80.523</td>\n",
       "      <td>80.744</td>\n",
       "      <td>...</td>\n",
       "      <td>78.897</td>\n",
       "      <td>78.863</td>\n",
       "      <td>74.278</td>\n",
       "      <td>77.770</td>\n",
       "      <td>78.197</td>\n",
       "      <td>78.154</td>\n",
       "      <td>78.679</td>\n",
       "      <td>79.170</td>\n",
       "      <td>80.171</td>\n",
       "      <td>79.576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Tapesia_yallundae_CCL031</td>\n",
       "      <td>74.141</td>\n",
       "      <td>65.229</td>\n",
       "      <td>77.185</td>\n",
       "      <td>78.320</td>\n",
       "      <td>78.158</td>\n",
       "      <td>78.694</td>\n",
       "      <td>69.460</td>\n",
       "      <td>84.418</td>\n",
       "      <td>82.472</td>\n",
       "      <td>81.856</td>\n",
       "      <td>...</td>\n",
       "      <td>80.027</td>\n",
       "      <td>79.767</td>\n",
       "      <td>75.567</td>\n",
       "      <td>79.602</td>\n",
       "      <td>79.280</td>\n",
       "      <td>80.756</td>\n",
       "      <td>80.212</td>\n",
       "      <td>80.301</td>\n",
       "      <td>81.231</td>\n",
       "      <td>82.557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Tapesia_yallundae_CCL029</td>\n",
       "      <td>73.956</td>\n",
       "      <td>65.371</td>\n",
       "      <td>77.095</td>\n",
       "      <td>79.720</td>\n",
       "      <td>78.112</td>\n",
       "      <td>79.118</td>\n",
       "      <td>70.010</td>\n",
       "      <td>83.123</td>\n",
       "      <td>82.296</td>\n",
       "      <td>82.215</td>\n",
       "      <td>...</td>\n",
       "      <td>80.110</td>\n",
       "      <td>80.096</td>\n",
       "      <td>75.592</td>\n",
       "      <td>80.520</td>\n",
       "      <td>79.267</td>\n",
       "      <td>79.657</td>\n",
       "      <td>80.102</td>\n",
       "      <td>80.794</td>\n",
       "      <td>81.052</td>\n",
       "      <td>81.355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Zymoseptoria_tritici</td>\n",
       "      <td>73.702</td>\n",
       "      <td>64.832</td>\n",
       "      <td>75.907</td>\n",
       "      <td>76.101</td>\n",
       "      <td>77.028</td>\n",
       "      <td>77.626</td>\n",
       "      <td>69.300</td>\n",
       "      <td>81.955</td>\n",
       "      <td>81.777</td>\n",
       "      <td>81.443</td>\n",
       "      <td>...</td>\n",
       "      <td>79.507</td>\n",
       "      <td>79.307</td>\n",
       "      <td>75.034</td>\n",
       "      <td>78.567</td>\n",
       "      <td>78.873</td>\n",
       "      <td>79.011</td>\n",
       "      <td>79.589</td>\n",
       "      <td>79.445</td>\n",
       "      <td>81.659</td>\n",
       "      <td>80.274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Dothiorella_vidmadera</td>\n",
       "      <td>74.615</td>\n",
       "      <td>65.732</td>\n",
       "      <td>77.573</td>\n",
       "      <td>78.708</td>\n",
       "      <td>77.823</td>\n",
       "      <td>78.830</td>\n",
       "      <td>69.470</td>\n",
       "      <td>83.276</td>\n",
       "      <td>82.856</td>\n",
       "      <td>82.776</td>\n",
       "      <td>...</td>\n",
       "      <td>80.740</td>\n",
       "      <td>80.597</td>\n",
       "      <td>77.050</td>\n",
       "      <td>80.350</td>\n",
       "      <td>79.897</td>\n",
       "      <td>80.240</td>\n",
       "      <td>81.579</td>\n",
       "      <td>80.946</td>\n",
       "      <td>81.436</td>\n",
       "      <td>81.573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pichia_kudriavzevii</td>\n",
       "      <td>70.411</td>\n",
       "      <td>66.809</td>\n",
       "      <td>73.021</td>\n",
       "      <td>72.734</td>\n",
       "      <td>73.494</td>\n",
       "      <td>72.562</td>\n",
       "      <td>70.944</td>\n",
       "      <td>74.596</td>\n",
       "      <td>73.953</td>\n",
       "      <td>74.502</td>\n",
       "      <td>...</td>\n",
       "      <td>79.219</td>\n",
       "      <td>78.742</td>\n",
       "      <td>75.369</td>\n",
       "      <td>78.991</td>\n",
       "      <td>79.142</td>\n",
       "      <td>79.046</td>\n",
       "      <td>78.559</td>\n",
       "      <td>77.176</td>\n",
       "      <td>78.788</td>\n",
       "      <td>79.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pichia_membranifaciens</td>\n",
       "      <td>70.372</td>\n",
       "      <td>66.292</td>\n",
       "      <td>73.386</td>\n",
       "      <td>73.910</td>\n",
       "      <td>74.164</td>\n",
       "      <td>73.648</td>\n",
       "      <td>70.592</td>\n",
       "      <td>75.137</td>\n",
       "      <td>74.443</td>\n",
       "      <td>74.134</td>\n",
       "      <td>...</td>\n",
       "      <td>79.358</td>\n",
       "      <td>79.316</td>\n",
       "      <td>76.067</td>\n",
       "      <td>79.585</td>\n",
       "      <td>79.347</td>\n",
       "      <td>79.778</td>\n",
       "      <td>79.394</td>\n",
       "      <td>78.356</td>\n",
       "      <td>79.048</td>\n",
       "      <td>79.691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Galactomyces_geotrichum</td>\n",
       "      <td>71.641</td>\n",
       "      <td>66.607</td>\n",
       "      <td>72.624</td>\n",
       "      <td>71.876</td>\n",
       "      <td>73.273</td>\n",
       "      <td>72.040</td>\n",
       "      <td>72.238</td>\n",
       "      <td>74.321</td>\n",
       "      <td>73.161</td>\n",
       "      <td>73.308</td>\n",
       "      <td>...</td>\n",
       "      <td>79.027</td>\n",
       "      <td>77.445</td>\n",
       "      <td>75.574</td>\n",
       "      <td>78.480</td>\n",
       "      <td>78.885</td>\n",
       "      <td>78.991</td>\n",
       "      <td>77.349</td>\n",
       "      <td>76.883</td>\n",
       "      <td>79.103</td>\n",
       "      <td>78.252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Zygoascus_hellenicus</td>\n",
       "      <td>75.627</td>\n",
       "      <td>67.583</td>\n",
       "      <td>76.799</td>\n",
       "      <td>77.541</td>\n",
       "      <td>78.187</td>\n",
       "      <td>77.249</td>\n",
       "      <td>72.389</td>\n",
       "      <td>79.822</td>\n",
       "      <td>76.849</td>\n",
       "      <td>77.489</td>\n",
       "      <td>...</td>\n",
       "      <td>81.709</td>\n",
       "      <td>81.220</td>\n",
       "      <td>78.440</td>\n",
       "      <td>81.818</td>\n",
       "      <td>81.415</td>\n",
       "      <td>81.662</td>\n",
       "      <td>81.812</td>\n",
       "      <td>80.902</td>\n",
       "      <td>83.977</td>\n",
       "      <td>82.536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Clavispora_lusitaniae</td>\n",
       "      <td>70.284</td>\n",
       "      <td>69.506</td>\n",
       "      <td>72.759</td>\n",
       "      <td>72.452</td>\n",
       "      <td>73.085</td>\n",
       "      <td>72.905</td>\n",
       "      <td>72.446</td>\n",
       "      <td>74.905</td>\n",
       "      <td>72.957</td>\n",
       "      <td>72.850</td>\n",
       "      <td>...</td>\n",
       "      <td>77.967</td>\n",
       "      <td>78.480</td>\n",
       "      <td>75.240</td>\n",
       "      <td>78.756</td>\n",
       "      <td>78.586</td>\n",
       "      <td>78.566</td>\n",
       "      <td>79.054</td>\n",
       "      <td>79.958</td>\n",
       "      <td>78.488</td>\n",
       "      <td>78.015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Blastobotrys_proliferans</td>\n",
       "      <td>77.021</td>\n",
       "      <td>67.062</td>\n",
       "      <td>76.287</td>\n",
       "      <td>77.324</td>\n",
       "      <td>78.210</td>\n",
       "      <td>76.495</td>\n",
       "      <td>72.673</td>\n",
       "      <td>80.563</td>\n",
       "      <td>76.805</td>\n",
       "      <td>76.676</td>\n",
       "      <td>...</td>\n",
       "      <td>82.640</td>\n",
       "      <td>83.001</td>\n",
       "      <td>78.861</td>\n",
       "      <td>82.751</td>\n",
       "      <td>82.987</td>\n",
       "      <td>82.768</td>\n",
       "      <td>81.784</td>\n",
       "      <td>81.884</td>\n",
       "      <td>84.504</td>\n",
       "      <td>83.162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cryptococcus_neoformans</td>\n",
       "      <td>74.691</td>\n",
       "      <td>67.510</td>\n",
       "      <td>76.564</td>\n",
       "      <td>77.158</td>\n",
       "      <td>78.345</td>\n",
       "      <td>77.398</td>\n",
       "      <td>71.979</td>\n",
       "      <td>79.515</td>\n",
       "      <td>77.717</td>\n",
       "      <td>77.573</td>\n",
       "      <td>...</td>\n",
       "      <td>84.586</td>\n",
       "      <td>85.425</td>\n",
       "      <td>79.931</td>\n",
       "      <td>84.451</td>\n",
       "      <td>84.528</td>\n",
       "      <td>84.597</td>\n",
       "      <td>85.606</td>\n",
       "      <td>85.448</td>\n",
       "      <td>85.861</td>\n",
       "      <td>85.484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Kluyveromyces_marxianus</td>\n",
       "      <td>74.297</td>\n",
       "      <td>67.407</td>\n",
       "      <td>76.979</td>\n",
       "      <td>76.652</td>\n",
       "      <td>77.966</td>\n",
       "      <td>77.296</td>\n",
       "      <td>70.922</td>\n",
       "      <td>80.048</td>\n",
       "      <td>77.614</td>\n",
       "      <td>77.538</td>\n",
       "      <td>...</td>\n",
       "      <td>84.993</td>\n",
       "      <td>84.425</td>\n",
       "      <td>80.208</td>\n",
       "      <td>84.354</td>\n",
       "      <td>84.960</td>\n",
       "      <td>85.069</td>\n",
       "      <td>85.330</td>\n",
       "      <td>85.133</td>\n",
       "      <td>85.656</td>\n",
       "      <td>86.032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>75.683</td>\n",
       "      <td>66.816</td>\n",
       "      <td>76.452</td>\n",
       "      <td>76.836</td>\n",
       "      <td>78.454</td>\n",
       "      <td>77.417</td>\n",
       "      <td>71.129</td>\n",
       "      <td>80.137</td>\n",
       "      <td>77.349</td>\n",
       "      <td>77.239</td>\n",
       "      <td>...</td>\n",
       "      <td>84.782</td>\n",
       "      <td>84.856</td>\n",
       "      <td>78.483</td>\n",
       "      <td>84.483</td>\n",
       "      <td>84.404</td>\n",
       "      <td>84.341</td>\n",
       "      <td>84.479</td>\n",
       "      <td>84.137</td>\n",
       "      <td>86.183</td>\n",
       "      <td>85.959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Kodamaea_ohmeri</td>\n",
       "      <td>71.033</td>\n",
       "      <td>69.648</td>\n",
       "      <td>72.064</td>\n",
       "      <td>71.556</td>\n",
       "      <td>73.602</td>\n",
       "      <td>72.108</td>\n",
       "      <td>74.256</td>\n",
       "      <td>75.034</td>\n",
       "      <td>73.566</td>\n",
       "      <td>72.184</td>\n",
       "      <td>...</td>\n",
       "      <td>79.659</td>\n",
       "      <td>79.090</td>\n",
       "      <td>76.578</td>\n",
       "      <td>79.447</td>\n",
       "      <td>80.864</td>\n",
       "      <td>79.909</td>\n",
       "      <td>79.643</td>\n",
       "      <td>79.213</td>\n",
       "      <td>79.227</td>\n",
       "      <td>79.234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Wickerhamomyces_anomalus</td>\n",
       "      <td>74.180</td>\n",
       "      <td>67.792</td>\n",
       "      <td>76.469</td>\n",
       "      <td>78.074</td>\n",
       "      <td>76.374</td>\n",
       "      <td>77.649</td>\n",
       "      <td>72.268</td>\n",
       "      <td>80.437</td>\n",
       "      <td>77.318</td>\n",
       "      <td>77.155</td>\n",
       "      <td>...</td>\n",
       "      <td>85.582</td>\n",
       "      <td>85.494</td>\n",
       "      <td>79.725</td>\n",
       "      <td>86.850</td>\n",
       "      <td>85.966</td>\n",
       "      <td>86.054</td>\n",
       "      <td>86.873</td>\n",
       "      <td>86.047</td>\n",
       "      <td>88.649</td>\n",
       "      <td>87.916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Candida_albicans</td>\n",
       "      <td>74.129</td>\n",
       "      <td>67.026</td>\n",
       "      <td>76.862</td>\n",
       "      <td>75.974</td>\n",
       "      <td>77.435</td>\n",
       "      <td>76.863</td>\n",
       "      <td>72.129</td>\n",
       "      <td>79.386</td>\n",
       "      <td>77.279</td>\n",
       "      <td>77.587</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>94.990</td>\n",
       "      <td>81.512</td>\n",
       "      <td>89.736</td>\n",
       "      <td>89.771</td>\n",
       "      <td>89.709</td>\n",
       "      <td>86.619</td>\n",
       "      <td>86.392</td>\n",
       "      <td>87.311</td>\n",
       "      <td>87.944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Kluyveromyces_lactis</td>\n",
       "      <td>74.597</td>\n",
       "      <td>67.211</td>\n",
       "      <td>75.984</td>\n",
       "      <td>76.324</td>\n",
       "      <td>77.648</td>\n",
       "      <td>76.878</td>\n",
       "      <td>72.986</td>\n",
       "      <td>78.696</td>\n",
       "      <td>77.218</td>\n",
       "      <td>76.831</td>\n",
       "      <td>...</td>\n",
       "      <td>94.990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>81.239</td>\n",
       "      <td>89.616</td>\n",
       "      <td>89.066</td>\n",
       "      <td>88.766</td>\n",
       "      <td>86.462</td>\n",
       "      <td>86.421</td>\n",
       "      <td>86.930</td>\n",
       "      <td>87.229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cryptococcus_gattii</td>\n",
       "      <td>72.674</td>\n",
       "      <td>66.369</td>\n",
       "      <td>73.566</td>\n",
       "      <td>73.562</td>\n",
       "      <td>74.179</td>\n",
       "      <td>73.283</td>\n",
       "      <td>69.777</td>\n",
       "      <td>75.352</td>\n",
       "      <td>72.843</td>\n",
       "      <td>73.046</td>\n",
       "      <td>...</td>\n",
       "      <td>81.512</td>\n",
       "      <td>81.239</td>\n",
       "      <td>NaN</td>\n",
       "      <td>85.054</td>\n",
       "      <td>85.499</td>\n",
       "      <td>85.855</td>\n",
       "      <td>82.306</td>\n",
       "      <td>80.262</td>\n",
       "      <td>81.959</td>\n",
       "      <td>80.949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Candida_parapsilosis</td>\n",
       "      <td>75.689</td>\n",
       "      <td>67.898</td>\n",
       "      <td>76.588</td>\n",
       "      <td>76.863</td>\n",
       "      <td>76.547</td>\n",
       "      <td>77.178</td>\n",
       "      <td>72.280</td>\n",
       "      <td>79.780</td>\n",
       "      <td>76.362</td>\n",
       "      <td>76.380</td>\n",
       "      <td>...</td>\n",
       "      <td>89.736</td>\n",
       "      <td>89.616</td>\n",
       "      <td>85.054</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95.306</td>\n",
       "      <td>95.273</td>\n",
       "      <td>87.267</td>\n",
       "      <td>86.244</td>\n",
       "      <td>88.505</td>\n",
       "      <td>88.162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Candida_metapsilosis</td>\n",
       "      <td>75.569</td>\n",
       "      <td>67.352</td>\n",
       "      <td>76.513</td>\n",
       "      <td>75.658</td>\n",
       "      <td>76.704</td>\n",
       "      <td>76.939</td>\n",
       "      <td>72.393</td>\n",
       "      <td>79.808</td>\n",
       "      <td>77.340</td>\n",
       "      <td>76.003</td>\n",
       "      <td>...</td>\n",
       "      <td>89.771</td>\n",
       "      <td>89.066</td>\n",
       "      <td>85.499</td>\n",
       "      <td>95.306</td>\n",
       "      <td>NaN</td>\n",
       "      <td>97.255</td>\n",
       "      <td>86.888</td>\n",
       "      <td>86.273</td>\n",
       "      <td>88.483</td>\n",
       "      <td>88.150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Candida_orthopsilosis</td>\n",
       "      <td>75.603</td>\n",
       "      <td>67.294</td>\n",
       "      <td>76.778</td>\n",
       "      <td>76.291</td>\n",
       "      <td>76.610</td>\n",
       "      <td>76.920</td>\n",
       "      <td>72.178</td>\n",
       "      <td>81.128</td>\n",
       "      <td>76.422</td>\n",
       "      <td>76.080</td>\n",
       "      <td>...</td>\n",
       "      <td>89.709</td>\n",
       "      <td>88.766</td>\n",
       "      <td>85.855</td>\n",
       "      <td>95.273</td>\n",
       "      <td>97.255</td>\n",
       "      <td>NaN</td>\n",
       "      <td>87.388</td>\n",
       "      <td>86.189</td>\n",
       "      <td>88.701</td>\n",
       "      <td>89.532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Yamadazyma_mexicana</td>\n",
       "      <td>74.991</td>\n",
       "      <td>68.555</td>\n",
       "      <td>77.567</td>\n",
       "      <td>76.726</td>\n",
       "      <td>77.659</td>\n",
       "      <td>77.117</td>\n",
       "      <td>71.394</td>\n",
       "      <td>80.711</td>\n",
       "      <td>77.375</td>\n",
       "      <td>77.375</td>\n",
       "      <td>...</td>\n",
       "      <td>86.619</td>\n",
       "      <td>86.462</td>\n",
       "      <td>82.306</td>\n",
       "      <td>87.267</td>\n",
       "      <td>86.888</td>\n",
       "      <td>87.388</td>\n",
       "      <td>NaN</td>\n",
       "      <td>92.182</td>\n",
       "      <td>89.450</td>\n",
       "      <td>89.924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Yamadazyma_scolyti</td>\n",
       "      <td>74.607</td>\n",
       "      <td>68.902</td>\n",
       "      <td>76.897</td>\n",
       "      <td>76.553</td>\n",
       "      <td>77.304</td>\n",
       "      <td>78.259</td>\n",
       "      <td>71.414</td>\n",
       "      <td>79.352</td>\n",
       "      <td>77.260</td>\n",
       "      <td>76.781</td>\n",
       "      <td>...</td>\n",
       "      <td>86.392</td>\n",
       "      <td>86.421</td>\n",
       "      <td>80.262</td>\n",
       "      <td>86.244</td>\n",
       "      <td>86.273</td>\n",
       "      <td>86.189</td>\n",
       "      <td>92.182</td>\n",
       "      <td>NaN</td>\n",
       "      <td>88.710</td>\n",
       "      <td>88.980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Candida_zeylanoides</td>\n",
       "      <td>76.313</td>\n",
       "      <td>67.022</td>\n",
       "      <td>76.831</td>\n",
       "      <td>77.186</td>\n",
       "      <td>77.622</td>\n",
       "      <td>77.945</td>\n",
       "      <td>71.718</td>\n",
       "      <td>81.909</td>\n",
       "      <td>77.747</td>\n",
       "      <td>77.414</td>\n",
       "      <td>...</td>\n",
       "      <td>87.311</td>\n",
       "      <td>86.930</td>\n",
       "      <td>81.959</td>\n",
       "      <td>88.505</td>\n",
       "      <td>88.483</td>\n",
       "      <td>88.701</td>\n",
       "      <td>89.450</td>\n",
       "      <td>88.710</td>\n",
       "      <td>NaN</td>\n",
       "      <td>91.750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Meyerozyma_guillermondii</td>\n",
       "      <td>75.231</td>\n",
       "      <td>66.816</td>\n",
       "      <td>77.115</td>\n",
       "      <td>77.125</td>\n",
       "      <td>76.724</td>\n",
       "      <td>77.801</td>\n",
       "      <td>71.261</td>\n",
       "      <td>82.085</td>\n",
       "      <td>77.766</td>\n",
       "      <td>77.425</td>\n",
       "      <td>...</td>\n",
       "      <td>87.944</td>\n",
       "      <td>87.229</td>\n",
       "      <td>80.949</td>\n",
       "      <td>88.162</td>\n",
       "      <td>88.150</td>\n",
       "      <td>89.532</td>\n",
       "      <td>89.924</td>\n",
       "      <td>88.980</td>\n",
       "      <td>91.750</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46 rows  46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Puccinia_striiformis_tritici  \\\n",
       "Puccinia_striiformis_tritici                           NaN   \n",
       "Geotrichum_candidum                                 63.080   \n",
       "Rhodotorula_mucilaginosa                            79.119   \n",
       "Cortinarius_globuliformis                           77.724   \n",
       "Cryptococcus_zero                                   78.490   \n",
       "Quambalaria_cyanescens                              78.029   \n",
       "Yarrowia_lipolytica                                 67.484   \n",
       "Tuber_brumale                                       74.459   \n",
       "Diaporthe_sp.                                       72.815   \n",
       "CCL060                                              72.503   \n",
       "CCL067                                              72.908   \n",
       "CCL068                                              72.544   \n",
       "Scedosporium_boydii                                 70.945   \n",
       "Fusarium_oxysporum                                  72.677   \n",
       "Entoleuca_sp.                                       72.500   \n",
       "Pyrenophora_tritici-repentis                        73.861   \n",
       "Cladophialophora_sp.                                73.411   \n",
       "Penicillium_chrysogenum                             72.631   \n",
       "Aspergillus_flavus                                  72.332   \n",
       "Aspergillus_niger                                   72.005   \n",
       "Aspergillus_sp.                                     71.844   \n",
       "Tapesia_yallundae_CCL031                            74.141   \n",
       "Tapesia_yallundae_CCL029                            73.956   \n",
       "Zymoseptoria_tritici                                73.702   \n",
       "Dothiorella_vidmadera                               74.615   \n",
       "Pichia_kudriavzevii                                 70.411   \n",
       "Pichia_membranifaciens                              70.372   \n",
       "Galactomyces_geotrichum                             71.641   \n",
       "Zygoascus_hellenicus                                75.627   \n",
       "Clavispora_lusitaniae                               70.284   \n",
       "Blastobotrys_proliferans                            77.021   \n",
       "Cryptococcus_neoformans                             74.691   \n",
       "Kluyveromyces_marxianus                             74.297   \n",
       "Saccharomyces_cerevisiae                            75.683   \n",
       "Kodamaea_ohmeri                                     71.033   \n",
       "Wickerhamomyces_anomalus                            74.180   \n",
       "Candida_albicans                                    74.129   \n",
       "Kluyveromyces_lactis                                74.597   \n",
       "Cryptococcus_gattii                                 72.674   \n",
       "Candida_parapsilosis                                75.689   \n",
       "Candida_metapsilosis                                75.569   \n",
       "Candida_orthopsilosis                               75.603   \n",
       "Yamadazyma_mexicana                                 74.991   \n",
       "Yamadazyma_scolyti                                  74.607   \n",
       "Candida_zeylanoides                                 76.313   \n",
       "Meyerozyma_guillermondii                            75.231   \n",
       "\n",
       "                              Geotrichum_candidum  Rhodotorula_mucilaginosa  \\\n",
       "Puccinia_striiformis_tritici               63.080                    79.119   \n",
       "Geotrichum_candidum                           NaN                    64.659   \n",
       "Rhodotorula_mucilaginosa                   64.659                       NaN   \n",
       "Cortinarius_globuliformis                  65.081                    79.211   \n",
       "Cryptococcus_zero                          65.344                    79.896   \n",
       "Quambalaria_cyanescens                     64.476                    80.212   \n",
       "Yarrowia_lipolytica                        67.220                    68.029   \n",
       "Tuber_brumale                              65.202                    75.752   \n",
       "Diaporthe_sp.                              64.461                    74.743   \n",
       "CCL060                                     64.048                    76.012   \n",
       "CCL067                                     63.892                    74.776   \n",
       "CCL068                                     64.029                    74.484   \n",
       "Scedosporium_boydii                        63.356                    73.597   \n",
       "Fusarium_oxysporum                         64.263                    74.332   \n",
       "Entoleuca_sp.                              65.103                    74.512   \n",
       "Pyrenophora_tritici-repentis               64.382                    76.762   \n",
       "Cladophialophora_sp.                       65.086                    75.514   \n",
       "Penicillium_chrysogenum                    64.829                    75.694   \n",
       "Aspergillus_flavus                         65.153                    75.778   \n",
       "Aspergillus_niger                          64.984                    75.771   \n",
       "Aspergillus_sp.                            64.345                    75.317   \n",
       "Tapesia_yallundae_CCL031                   65.229                    77.185   \n",
       "Tapesia_yallundae_CCL029                   65.371                    77.095   \n",
       "Zymoseptoria_tritici                       64.832                    75.907   \n",
       "Dothiorella_vidmadera                      65.732                    77.573   \n",
       "Pichia_kudriavzevii                        66.809                    73.021   \n",
       "Pichia_membranifaciens                     66.292                    73.386   \n",
       "Galactomyces_geotrichum                    66.607                    72.624   \n",
       "Zygoascus_hellenicus                       67.583                    76.799   \n",
       "Clavispora_lusitaniae                      69.506                    72.759   \n",
       "Blastobotrys_proliferans                   67.062                    76.287   \n",
       "Cryptococcus_neoformans                    67.510                    76.564   \n",
       "Kluyveromyces_marxianus                    67.407                    76.979   \n",
       "Saccharomyces_cerevisiae                   66.816                    76.452   \n",
       "Kodamaea_ohmeri                            69.648                    72.064   \n",
       "Wickerhamomyces_anomalus                   67.792                    76.469   \n",
       "Candida_albicans                           67.026                    76.862   \n",
       "Kluyveromyces_lactis                       67.211                    75.984   \n",
       "Cryptococcus_gattii                        66.369                    73.566   \n",
       "Candida_parapsilosis                       67.898                    76.588   \n",
       "Candida_metapsilosis                       67.352                    76.513   \n",
       "Candida_orthopsilosis                      67.294                    76.778   \n",
       "Yamadazyma_mexicana                        68.555                    77.567   \n",
       "Yamadazyma_scolyti                         68.902                    76.897   \n",
       "Candida_zeylanoides                        67.022                    76.831   \n",
       "Meyerozyma_guillermondii                   66.816                    77.115   \n",
       "\n",
       "                              Cortinarius_globuliformis  Cryptococcus_zero  \\\n",
       "Puccinia_striiformis_tritici                     77.724             78.490   \n",
       "Geotrichum_candidum                              65.081             65.344   \n",
       "Rhodotorula_mucilaginosa                         79.211             79.896   \n",
       "Cortinarius_globuliformis                           NaN             80.495   \n",
       "Cryptococcus_zero                                80.495                NaN   \n",
       "Quambalaria_cyanescens                           81.594             81.750   \n",
       "Yarrowia_lipolytica                              68.569             68.776   \n",
       "Tuber_brumale                                    77.239             76.334   \n",
       "Diaporthe_sp.                                    75.299             75.360   \n",
       "CCL060                                           75.094             75.351   \n",
       "CCL067                                           75.531             75.840   \n",
       "CCL068                                           76.000             75.445   \n",
       "Scedosporium_boydii                              73.913             74.434   \n",
       "Fusarium_oxysporum                               74.710             75.154   \n",
       "Entoleuca_sp.                                    75.428             76.150   \n",
       "Pyrenophora_tritici-repentis                     76.968             76.712   \n",
       "Cladophialophora_sp.                             76.376             76.797   \n",
       "Penicillium_chrysogenum                          76.342             77.013   \n",
       "Aspergillus_flavus                               76.434             76.821   \n",
       "Aspergillus_niger                                76.384             76.386   \n",
       "Aspergillus_sp.                                  75.718             76.318   \n",
       "Tapesia_yallundae_CCL031                         78.320             78.158   \n",
       "Tapesia_yallundae_CCL029                         79.720             78.112   \n",
       "Zymoseptoria_tritici                             76.101             77.028   \n",
       "Dothiorella_vidmadera                            78.708             77.823   \n",
       "Pichia_kudriavzevii                              72.734             73.494   \n",
       "Pichia_membranifaciens                           73.910             74.164   \n",
       "Galactomyces_geotrichum                          71.876             73.273   \n",
       "Zygoascus_hellenicus                             77.541             78.187   \n",
       "Clavispora_lusitaniae                            72.452             73.085   \n",
       "Blastobotrys_proliferans                         77.324             78.210   \n",
       "Cryptococcus_neoformans                          77.158             78.345   \n",
       "Kluyveromyces_marxianus                          76.652             77.966   \n",
       "Saccharomyces_cerevisiae                         76.836             78.454   \n",
       "Kodamaea_ohmeri                                  71.556             73.602   \n",
       "Wickerhamomyces_anomalus                         78.074             76.374   \n",
       "Candida_albicans                                 75.974             77.435   \n",
       "Kluyveromyces_lactis                             76.324             77.648   \n",
       "Cryptococcus_gattii                              73.562             74.179   \n",
       "Candida_parapsilosis                             76.863             76.547   \n",
       "Candida_metapsilosis                             75.658             76.704   \n",
       "Candida_orthopsilosis                            76.291             76.610   \n",
       "Yamadazyma_mexicana                              76.726             77.659   \n",
       "Yamadazyma_scolyti                               76.553             77.304   \n",
       "Candida_zeylanoides                              77.186             77.622   \n",
       "Meyerozyma_guillermondii                         77.125             76.724   \n",
       "\n",
       "                              Quambalaria_cyanescens  Yarrowia_lipolytica  \\\n",
       "Puccinia_striiformis_tritici                  78.029               67.484   \n",
       "Geotrichum_candidum                           64.476               67.220   \n",
       "Rhodotorula_mucilaginosa                      80.212               68.029   \n",
       "Cortinarius_globuliformis                     81.594               68.569   \n",
       "Cryptococcus_zero                             81.750               68.776   \n",
       "Quambalaria_cyanescens                           NaN               68.664   \n",
       "Yarrowia_lipolytica                           68.664                  NaN   \n",
       "Tuber_brumale                                 77.296               68.829   \n",
       "Diaporthe_sp.                                 76.607               68.166   \n",
       "CCL060                                        75.974               68.332   \n",
       "CCL067                                        76.197               68.056   \n",
       "CCL068                                        76.728               68.341   \n",
       "Scedosporium_boydii                           75.411               68.485   \n",
       "Fusarium_oxysporum                            76.145               68.648   \n",
       "Entoleuca_sp.                                 76.666               70.302   \n",
       "Pyrenophora_tritici-repentis                  77.015               69.081   \n",
       "Cladophialophora_sp.                          77.485               69.631   \n",
       "Penicillium_chrysogenum                       77.523               69.675   \n",
       "Aspergillus_flavus                            77.675               69.270   \n",
       "Aspergillus_niger                             77.341               69.654   \n",
       "Aspergillus_sp.                               77.273               69.699   \n",
       "Tapesia_yallundae_CCL031                      78.694               69.460   \n",
       "Tapesia_yallundae_CCL029                      79.118               70.010   \n",
       "Zymoseptoria_tritici                          77.626               69.300   \n",
       "Dothiorella_vidmadera                         78.830               69.470   \n",
       "Pichia_kudriavzevii                           72.562               70.944   \n",
       "Pichia_membranifaciens                        73.648               70.592   \n",
       "Galactomyces_geotrichum                       72.040               72.238   \n",
       "Zygoascus_hellenicus                          77.249               72.389   \n",
       "Clavispora_lusitaniae                         72.905               72.446   \n",
       "Blastobotrys_proliferans                      76.495               72.673   \n",
       "Cryptococcus_neoformans                       77.398               71.979   \n",
       "Kluyveromyces_marxianus                       77.296               70.922   \n",
       "Saccharomyces_cerevisiae                      77.417               71.129   \n",
       "Kodamaea_ohmeri                               72.108               74.256   \n",
       "Wickerhamomyces_anomalus                      77.649               72.268   \n",
       "Candida_albicans                              76.863               72.129   \n",
       "Kluyveromyces_lactis                          76.878               72.986   \n",
       "Cryptococcus_gattii                           73.283               69.777   \n",
       "Candida_parapsilosis                          77.178               72.280   \n",
       "Candida_metapsilosis                          76.939               72.393   \n",
       "Candida_orthopsilosis                         76.920               72.178   \n",
       "Yamadazyma_mexicana                           77.117               71.394   \n",
       "Yamadazyma_scolyti                            78.259               71.414   \n",
       "Candida_zeylanoides                           77.945               71.718   \n",
       "Meyerozyma_guillermondii                      77.801               71.261   \n",
       "\n",
       "                              Tuber_brumale  Diaporthe_sp.  CCL060  ...  \\\n",
       "Puccinia_striiformis_tritici         74.459         72.815  72.503  ...   \n",
       "Geotrichum_candidum                  65.202         64.461  64.048  ...   \n",
       "Rhodotorula_mucilaginosa             75.752         74.743  76.012  ...   \n",
       "Cortinarius_globuliformis            77.239         75.299  75.094  ...   \n",
       "Cryptococcus_zero                    76.334         75.360  75.351  ...   \n",
       "Quambalaria_cyanescens               77.296         76.607  75.974  ...   \n",
       "Yarrowia_lipolytica                  68.829         68.166  68.332  ...   \n",
       "Tuber_brumale                           NaN         78.803  78.469  ...   \n",
       "Diaporthe_sp.                        78.803            NaN  94.821  ...   \n",
       "CCL060                               78.469         94.821     NaN  ...   \n",
       "CCL067                               78.481         88.404  87.764  ...   \n",
       "CCL068                               78.906         87.285  87.681  ...   \n",
       "Scedosporium_boydii                  78.386         82.697  82.715  ...   \n",
       "Fusarium_oxysporum                   79.192         86.701  84.374  ...   \n",
       "Entoleuca_sp.                        79.241         85.276  84.700  ...   \n",
       "Pyrenophora_tritici-repentis         80.643         80.247  80.192  ...   \n",
       "Cladophialophora_sp.                 81.128         79.664  79.520  ...   \n",
       "Penicillium_chrysogenum              80.636         80.323  79.684  ...   \n",
       "Aspergillus_flavus                   80.799         80.898  80.815  ...   \n",
       "Aspergillus_niger                    80.863         80.779  80.929  ...   \n",
       "Aspergillus_sp.                      80.150         80.523  80.744  ...   \n",
       "Tapesia_yallundae_CCL031             84.418         82.472  81.856  ...   \n",
       "Tapesia_yallundae_CCL029             83.123         82.296  82.215  ...   \n",
       "Zymoseptoria_tritici                 81.955         81.777  81.443  ...   \n",
       "Dothiorella_vidmadera                83.276         82.856  82.776  ...   \n",
       "Pichia_kudriavzevii                  74.596         73.953  74.502  ...   \n",
       "Pichia_membranifaciens               75.137         74.443  74.134  ...   \n",
       "Galactomyces_geotrichum              74.321         73.161  73.308  ...   \n",
       "Zygoascus_hellenicus                 79.822         76.849  77.489  ...   \n",
       "Clavispora_lusitaniae                74.905         72.957  72.850  ...   \n",
       "Blastobotrys_proliferans             80.563         76.805  76.676  ...   \n",
       "Cryptococcus_neoformans              79.515         77.717  77.573  ...   \n",
       "Kluyveromyces_marxianus              80.048         77.614  77.538  ...   \n",
       "Saccharomyces_cerevisiae             80.137         77.349  77.239  ...   \n",
       "Kodamaea_ohmeri                      75.034         73.566  72.184  ...   \n",
       "Wickerhamomyces_anomalus             80.437         77.318  77.155  ...   \n",
       "Candida_albicans                     79.386         77.279  77.587  ...   \n",
       "Kluyveromyces_lactis                 78.696         77.218  76.831  ...   \n",
       "Cryptococcus_gattii                  75.352         72.843  73.046  ...   \n",
       "Candida_parapsilosis                 79.780         76.362  76.380  ...   \n",
       "Candida_metapsilosis                 79.808         77.340  76.003  ...   \n",
       "Candida_orthopsilosis                81.128         76.422  76.080  ...   \n",
       "Yamadazyma_mexicana                  80.711         77.375  77.375  ...   \n",
       "Yamadazyma_scolyti                   79.352         77.260  76.781  ...   \n",
       "Candida_zeylanoides                  81.909         77.747  77.414  ...   \n",
       "Meyerozyma_guillermondii             82.085         77.766  77.425  ...   \n",
       "\n",
       "                              Candida_albicans  Kluyveromyces_lactis  \\\n",
       "Puccinia_striiformis_tritici            74.129                74.597   \n",
       "Geotrichum_candidum                     67.026                67.211   \n",
       "Rhodotorula_mucilaginosa                76.862                75.984   \n",
       "Cortinarius_globuliformis               75.974                76.324   \n",
       "Cryptococcus_zero                       77.435                77.648   \n",
       "Quambalaria_cyanescens                  76.863                76.878   \n",
       "Yarrowia_lipolytica                     72.129                72.986   \n",
       "Tuber_brumale                           79.386                78.696   \n",
       "Diaporthe_sp.                           77.279                77.218   \n",
       "CCL060                                  77.587                76.831   \n",
       "CCL067                                  78.258                77.097   \n",
       "CCL068                                  77.163                77.063   \n",
       "Scedosporium_boydii                     77.015                76.652   \n",
       "Fusarium_oxysporum                      78.473                78.404   \n",
       "Entoleuca_sp.                           77.763                78.726   \n",
       "Pyrenophora_tritici-repentis            80.316                78.462   \n",
       "Cladophialophora_sp.                    78.884                78.359   \n",
       "Penicillium_chrysogenum                 79.541                79.239   \n",
       "Aspergillus_flavus                      79.294                79.274   \n",
       "Aspergillus_niger                       78.807                78.733   \n",
       "Aspergillus_sp.                         78.897                78.863   \n",
       "Tapesia_yallundae_CCL031                80.027                79.767   \n",
       "Tapesia_yallundae_CCL029                80.110                80.096   \n",
       "Zymoseptoria_tritici                    79.507                79.307   \n",
       "Dothiorella_vidmadera                   80.740                80.597   \n",
       "Pichia_kudriavzevii                     79.219                78.742   \n",
       "Pichia_membranifaciens                  79.358                79.316   \n",
       "Galactomyces_geotrichum                 79.027                77.445   \n",
       "Zygoascus_hellenicus                    81.709                81.220   \n",
       "Clavispora_lusitaniae                   77.967                78.480   \n",
       "Blastobotrys_proliferans                82.640                83.001   \n",
       "Cryptococcus_neoformans                 84.586                85.425   \n",
       "Kluyveromyces_marxianus                 84.993                84.425   \n",
       "Saccharomyces_cerevisiae                84.782                84.856   \n",
       "Kodamaea_ohmeri                         79.659                79.090   \n",
       "Wickerhamomyces_anomalus                85.582                85.494   \n",
       "Candida_albicans                           NaN                94.990   \n",
       "Kluyveromyces_lactis                    94.990                   NaN   \n",
       "Cryptococcus_gattii                     81.512                81.239   \n",
       "Candida_parapsilosis                    89.736                89.616   \n",
       "Candida_metapsilosis                    89.771                89.066   \n",
       "Candida_orthopsilosis                   89.709                88.766   \n",
       "Yamadazyma_mexicana                     86.619                86.462   \n",
       "Yamadazyma_scolyti                      86.392                86.421   \n",
       "Candida_zeylanoides                     87.311                86.930   \n",
       "Meyerozyma_guillermondii                87.944                87.229   \n",
       "\n",
       "                              Cryptococcus_gattii  Candida_parapsilosis  \\\n",
       "Puccinia_striiformis_tritici               72.674                75.689   \n",
       "Geotrichum_candidum                        66.369                67.898   \n",
       "Rhodotorula_mucilaginosa                   73.566                76.588   \n",
       "Cortinarius_globuliformis                  73.562                76.863   \n",
       "Cryptococcus_zero                          74.179                76.547   \n",
       "Quambalaria_cyanescens                     73.283                77.178   \n",
       "Yarrowia_lipolytica                        69.777                72.280   \n",
       "Tuber_brumale                              75.352                79.780   \n",
       "Diaporthe_sp.                              72.843                76.362   \n",
       "CCL060                                     73.046                76.380   \n",
       "CCL067                                     72.546                76.406   \n",
       "CCL068                                     73.040                77.113   \n",
       "Scedosporium_boydii                        72.299                75.914   \n",
       "Fusarium_oxysporum                         73.535                77.568   \n",
       "Entoleuca_sp.                              74.281                76.789   \n",
       "Pyrenophora_tritici-repentis               74.570                78.547   \n",
       "Cladophialophora_sp.                       74.829                77.842   \n",
       "Penicillium_chrysogenum                    73.807                78.044   \n",
       "Aspergillus_flavus                         74.314                78.481   \n",
       "Aspergillus_niger                          74.966                78.364   \n",
       "Aspergillus_sp.                            74.278                77.770   \n",
       "Tapesia_yallundae_CCL031                   75.567                79.602   \n",
       "Tapesia_yallundae_CCL029                   75.592                80.520   \n",
       "Zymoseptoria_tritici                       75.034                78.567   \n",
       "Dothiorella_vidmadera                      77.050                80.350   \n",
       "Pichia_kudriavzevii                        75.369                78.991   \n",
       "Pichia_membranifaciens                     76.067                79.585   \n",
       "Galactomyces_geotrichum                    75.574                78.480   \n",
       "Zygoascus_hellenicus                       78.440                81.818   \n",
       "Clavispora_lusitaniae                      75.240                78.756   \n",
       "Blastobotrys_proliferans                   78.861                82.751   \n",
       "Cryptococcus_neoformans                    79.931                84.451   \n",
       "Kluyveromyces_marxianus                    80.208                84.354   \n",
       "Saccharomyces_cerevisiae                   78.483                84.483   \n",
       "Kodamaea_ohmeri                            76.578                79.447   \n",
       "Wickerhamomyces_anomalus                   79.725                86.850   \n",
       "Candida_albicans                           81.512                89.736   \n",
       "Kluyveromyces_lactis                       81.239                89.616   \n",
       "Cryptococcus_gattii                           NaN                85.054   \n",
       "Candida_parapsilosis                       85.054                   NaN   \n",
       "Candida_metapsilosis                       85.499                95.306   \n",
       "Candida_orthopsilosis                      85.855                95.273   \n",
       "Yamadazyma_mexicana                        82.306                87.267   \n",
       "Yamadazyma_scolyti                         80.262                86.244   \n",
       "Candida_zeylanoides                        81.959                88.505   \n",
       "Meyerozyma_guillermondii                   80.949                88.162   \n",
       "\n",
       "                              Candida_metapsilosis  Candida_orthopsilosis  \\\n",
       "Puccinia_striiformis_tritici                75.569                 75.603   \n",
       "Geotrichum_candidum                         67.352                 67.294   \n",
       "Rhodotorula_mucilaginosa                    76.513                 76.778   \n",
       "Cortinarius_globuliformis                   75.658                 76.291   \n",
       "Cryptococcus_zero                           76.704                 76.610   \n",
       "Quambalaria_cyanescens                      76.939                 76.920   \n",
       "Yarrowia_lipolytica                         72.393                 72.178   \n",
       "Tuber_brumale                               79.808                 81.128   \n",
       "Diaporthe_sp.                               77.340                 76.422   \n",
       "CCL060                                      76.003                 76.080   \n",
       "CCL067                                      76.424                 76.604   \n",
       "CCL068                                      76.719                 77.121   \n",
       "Scedosporium_boydii                         75.461                 75.846   \n",
       "Fusarium_oxysporum                          78.633                 77.610   \n",
       "Entoleuca_sp.                               77.118                 77.129   \n",
       "Pyrenophora_tritici-repentis                78.853                 78.814   \n",
       "Cladophialophora_sp.                        78.025                 78.271   \n",
       "Penicillium_chrysogenum                     78.383                 78.249   \n",
       "Aspergillus_flavus                          78.417                 78.598   \n",
       "Aspergillus_niger                           78.086                 78.596   \n",
       "Aspergillus_sp.                             78.197                 78.154   \n",
       "Tapesia_yallundae_CCL031                    79.280                 80.756   \n",
       "Tapesia_yallundae_CCL029                    79.267                 79.657   \n",
       "Zymoseptoria_tritici                        78.873                 79.011   \n",
       "Dothiorella_vidmadera                       79.897                 80.240   \n",
       "Pichia_kudriavzevii                         79.142                 79.046   \n",
       "Pichia_membranifaciens                      79.347                 79.778   \n",
       "Galactomyces_geotrichum                     78.885                 78.991   \n",
       "Zygoascus_hellenicus                        81.415                 81.662   \n",
       "Clavispora_lusitaniae                       78.586                 78.566   \n",
       "Blastobotrys_proliferans                    82.987                 82.768   \n",
       "Cryptococcus_neoformans                     84.528                 84.597   \n",
       "Kluyveromyces_marxianus                     84.960                 85.069   \n",
       "Saccharomyces_cerevisiae                    84.404                 84.341   \n",
       "Kodamaea_ohmeri                             80.864                 79.909   \n",
       "Wickerhamomyces_anomalus                    85.966                 86.054   \n",
       "Candida_albicans                            89.771                 89.709   \n",
       "Kluyveromyces_lactis                        89.066                 88.766   \n",
       "Cryptococcus_gattii                         85.499                 85.855   \n",
       "Candida_parapsilosis                        95.306                 95.273   \n",
       "Candida_metapsilosis                           NaN                 97.255   \n",
       "Candida_orthopsilosis                       97.255                    NaN   \n",
       "Yamadazyma_mexicana                         86.888                 87.388   \n",
       "Yamadazyma_scolyti                          86.273                 86.189   \n",
       "Candida_zeylanoides                         88.483                 88.701   \n",
       "Meyerozyma_guillermondii                    88.150                 89.532   \n",
       "\n",
       "                              Yamadazyma_mexicana  Yamadazyma_scolyti  \\\n",
       "Puccinia_striiformis_tritici               74.991              74.607   \n",
       "Geotrichum_candidum                        68.555              68.902   \n",
       "Rhodotorula_mucilaginosa                   77.567              76.897   \n",
       "Cortinarius_globuliformis                  76.726              76.553   \n",
       "Cryptococcus_zero                          77.659              77.304   \n",
       "Quambalaria_cyanescens                     77.117              78.259   \n",
       "Yarrowia_lipolytica                        71.394              71.414   \n",
       "Tuber_brumale                              80.711              79.352   \n",
       "Diaporthe_sp.                              77.375              77.260   \n",
       "CCL060                                     77.375              76.781   \n",
       "CCL067                                     77.026              76.442   \n",
       "CCL068                                     77.310              77.076   \n",
       "Scedosporium_boydii                        76.119              76.621   \n",
       "Fusarium_oxysporum                         77.793              78.025   \n",
       "Entoleuca_sp.                              77.561              78.557   \n",
       "Pyrenophora_tritici-repentis               79.275              79.486   \n",
       "Cladophialophora_sp.                       78.559              78.535   \n",
       "Penicillium_chrysogenum                    78.481              79.444   \n",
       "Aspergillus_flavus                         78.946              79.911   \n",
       "Aspergillus_niger                          79.740              79.142   \n",
       "Aspergillus_sp.                            78.679              79.170   \n",
       "Tapesia_yallundae_CCL031                   80.212              80.301   \n",
       "Tapesia_yallundae_CCL029                   80.102              80.794   \n",
       "Zymoseptoria_tritici                       79.589              79.445   \n",
       "Dothiorella_vidmadera                      81.579              80.946   \n",
       "Pichia_kudriavzevii                        78.559              77.176   \n",
       "Pichia_membranifaciens                     79.394              78.356   \n",
       "Galactomyces_geotrichum                    77.349              76.883   \n",
       "Zygoascus_hellenicus                       81.812              80.902   \n",
       "Clavispora_lusitaniae                      79.054              79.958   \n",
       "Blastobotrys_proliferans                   81.784              81.884   \n",
       "Cryptococcus_neoformans                    85.606              85.448   \n",
       "Kluyveromyces_marxianus                    85.330              85.133   \n",
       "Saccharomyces_cerevisiae                   84.479              84.137   \n",
       "Kodamaea_ohmeri                            79.643              79.213   \n",
       "Wickerhamomyces_anomalus                   86.873              86.047   \n",
       "Candida_albicans                           86.619              86.392   \n",
       "Kluyveromyces_lactis                       86.462              86.421   \n",
       "Cryptococcus_gattii                        82.306              80.262   \n",
       "Candida_parapsilosis                       87.267              86.244   \n",
       "Candida_metapsilosis                       86.888              86.273   \n",
       "Candida_orthopsilosis                      87.388              86.189   \n",
       "Yamadazyma_mexicana                           NaN              92.182   \n",
       "Yamadazyma_scolyti                         92.182                 NaN   \n",
       "Candida_zeylanoides                        89.450              88.710   \n",
       "Meyerozyma_guillermondii                   89.924              88.980   \n",
       "\n",
       "                              Candida_zeylanoides  Meyerozyma_guillermondii  \n",
       "Puccinia_striiformis_tritici               76.313                    75.231  \n",
       "Geotrichum_candidum                        67.022                    66.816  \n",
       "Rhodotorula_mucilaginosa                   76.831                    77.115  \n",
       "Cortinarius_globuliformis                  77.186                    77.125  \n",
       "Cryptococcus_zero                          77.622                    76.724  \n",
       "Quambalaria_cyanescens                     77.945                    77.801  \n",
       "Yarrowia_lipolytica                        71.718                    71.261  \n",
       "Tuber_brumale                              81.909                    82.085  \n",
       "Diaporthe_sp.                              77.747                    77.766  \n",
       "CCL060                                     77.414                    77.425  \n",
       "CCL067                                     78.013                    77.409  \n",
       "CCL068                                     78.093                    77.858  \n",
       "Scedosporium_boydii                        76.374                    77.523  \n",
       "Fusarium_oxysporum                         78.449                    79.252  \n",
       "Entoleuca_sp.                              78.491                    78.374  \n",
       "Pyrenophora_tritici-repentis               79.980                    79.952  \n",
       "Cladophialophora_sp.                       80.827                    79.097  \n",
       "Penicillium_chrysogenum                    79.343                    79.829  \n",
       "Aspergillus_flavus                         79.678                    80.377  \n",
       "Aspergillus_niger                          79.446                    80.212  \n",
       "Aspergillus_sp.                            80.171                    79.576  \n",
       "Tapesia_yallundae_CCL031                   81.231                    82.557  \n",
       "Tapesia_yallundae_CCL029                   81.052                    81.355  \n",
       "Zymoseptoria_tritici                       81.659                    80.274  \n",
       "Dothiorella_vidmadera                      81.436                    81.573  \n",
       "Pichia_kudriavzevii                        78.788                    79.003  \n",
       "Pichia_membranifaciens                     79.048                    79.691  \n",
       "Galactomyces_geotrichum                    79.103                    78.252  \n",
       "Zygoascus_hellenicus                       83.977                    82.536  \n",
       "Clavispora_lusitaniae                      78.488                    78.015  \n",
       "Blastobotrys_proliferans                   84.504                    83.162  \n",
       "Cryptococcus_neoformans                    85.861                    85.484  \n",
       "Kluyveromyces_marxianus                    85.656                    86.032  \n",
       "Saccharomyces_cerevisiae                   86.183                    85.959  \n",
       "Kodamaea_ohmeri                            79.227                    79.234  \n",
       "Wickerhamomyces_anomalus                   88.649                    87.916  \n",
       "Candida_albicans                           87.311                    87.944  \n",
       "Kluyveromyces_lactis                       86.930                    87.229  \n",
       "Cryptococcus_gattii                        81.959                    80.949  \n",
       "Candida_parapsilosis                       88.505                    88.162  \n",
       "Candida_metapsilosis                       88.483                    88.150  \n",
       "Candida_orthopsilosis                      88.701                    89.532  \n",
       "Yamadazyma_mexicana                        89.450                    89.924  \n",
       "Yamadazyma_scolyti                         88.710                    88.980  \n",
       "Candida_zeylanoides                           NaN                    91.750  \n",
       "Meyerozyma_guillermondii                   91.750                       NaN  \n",
       "\n",
       "[46 rows x 46 columns]"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "large_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZ4AAAEWCAYAAABWn/G6AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nOzdd3wb5f3A8c/3NCzLS54BZxBGgFJm8a+0pIUUKFA2pZAwWvYKNFAoLbRQSqEFymgII6yyA2EVCGVvOmlDaUmZoYwMJ/GKvG1Zuuf3x50U2ZEdxbYkj+/79dLL1s3nTnf3vWfcc2KMQSmllMoWK9cJUEopNb5o4FFKKZVVGniUUkpllQYepZRSWaWBRymlVFZp4FFKKZVVGnhyTERuFZGbcp2OTBKRGSLSkOt0jAUicqeIzB3t61Ajm4hcIiJPZGz5mX6OR0RqgIuB6UAesBp4FrjaGLMqoytXiMgM4DWgvc+oHY0xn2ZgfVcANcaY/Yd72dkiIpsClwAHAFVAGPgIuM8Yc3cW07EC+LExZuEwLe8BYCbQDdg42/UP4GZjzGuDWN5WwFJgU2PM6uFIo3KIiAfnuvl9YAIQAT4Afm6MeSOXaRsOGc3xiMi3gT/jnLQ7G2OKgT2BRvdvqnl8mUzTOBUzxhT2+awXdMThzUUCRwoRmQy8DUwEDgJKgM2AS4FDc5i04fJ79/cvBr4KvAU8LyKzc5yuES/L16afAUcCBxljioCpwK+BziymIXOMMRn74NwN3bWBaV4H5gJPAi3Ahe7wI4D/AM3u38OT5pkKvIBzx7YW50KxjTtuH+Add1kNwMtJ832OcxfxGtAGLAF2BI4GPnHXdSfgTZrnbmA50Aq8DxyTNG4GEAWOB74AmoB7gMKkaQxwLvBvdxmvAVsljb8HuLPP9LOBf7rT/x3YNml8EXCfu64vgB+4aZjRz/6dAUT7Ged11/dD4F84B3UN8G2cO+EwUAc8BFQkzed39+PHbho/AQ4DjsW5M4u6+7cNmOL+Jl191nsZ8Jm7HS8D2yWNf8Dd779307ACOGWAY+gd4Kw+w34NvOj+vyvwV/f3bQT+ApT0s6x7gf8mHwMDrPcId7+F3WPj6KRxpwAfAj8CVrrbeQtg9TmO/4BTClALzAcK3HHP4eRKutz9+FzSvrk1aRlVrDtGW4DFycdXn/T2mjdp+GXuOor6TgcIcJWbvlb3NzvTHdfuHj/t7vw/c4df7U7X5h4bP0xa11buPMfi3MG3AM8DE/oc49e7y2h1f4/d3XG+pGMvDPwJ+MoAv9EuwJvu794EPANsnjRegDPcdbQAy4Az3HFXAC8Cv8M5D552h++Mcx6vBf6HEyQ87rg8nGtIvbu8j3GvXcAW7vKSr1v9/VbPAb8ZYLvi+/FknOtsGHgCqEyapiBpPza5y9xiQ+dx0rY/vxHLOhYng9EKrMG5wen/3NnQyTXYD7C1u2P22cB0r7s/0F7uQRAEvo5zwn0H5yJ1oPt9N3eeB4E73B/ZgxM8JrjjaoET3WXlAd9KWtfn7o/0JfcAfsA9cG53d+wU9wBLDi4nA+XuembhXFi3c8fNcLfxMZw74wk4F7jbkuY3OBelrYB84Cb3e/xAvYf1A88/3LTkAY8CLyWNvwsnF1kFFAMPu/MMJfC8g3NSeNx17oETgLzApu767k+a7zqcoL29u58nAzukOmDdYX0DzyU4B/vW7vouxwkuhUkXvg73d7eAo9z9Pqmf7ZgDLE76buFcQI52v/8D+Lm7fT6c4yu/n2XVAZemcXx/B+fiMt1d39dwTv7p7vhTgB7gV+42bu2On+mOD+KcxJe6x0UZzgX49qR1rABm9VlvclDw4ORYHnGPBwvYCdiknzT3F3i+RNK52mcdB7j7cpL7fQKwS5+L3yZ9lvd997iR+G8P7N1nnqfcbS7BubmanzT/4zjXhanuMqYBW7rjfgv8Ddjc3f7T3d+svxuJnXFKV/xACCfQ/ylp/A/d/by7u/8qgf9LOpajwDnu/EGg1P3df+YO2w7nuvIjd57ZOMG/zE37ZsCX3HGP4NxcxK9bO5EUKPqk+yKcG+cLgW/g3pAkjY/vx9fc374EWAQ8kzTNI+5+rnLTegXwHu5NFRtxHg+0LJwbhSiwhzttIfDNAc+fDZ1gg/3gnJAmvtMHmO51+uSKcALBgj7DHsK9oONcrJ9OtWz3ILiCFCefO+6CpO8HuGlMvkt4BPjdAOldDMx2/5/hzr9l0vj4iWa53w1wctL4IE4Z++5J29I38ByZ9P1AYK37v+XOu1fS+C3ZcOAxOBe9+OdJd1w88BzT3/a60x0G1CaloRPYr59p0wk8nwInJn33AKvi241z4XuqzzLWAgf2s85yd7/ET5p93ekD7vc/A7cBm6Vx3Nok5a5YV8cTdn/X+O/2PO4dftK081l3wT7FTUNyDucJ4Br3/1nAx33m383dt/G61w0Fnt3d7S5K85zsL/AUucfBzBTr2Afnwr4vkNdnvpSBJ8Xyn8S9e0+aZ5ek8ecA/3T/r3bHb5NiORbODcnufYZ/0Hc/DZCWnd3fOH5sfAycPsCx3Pc3+gHOdUSShp0FvJf0u3+Ec/3z9pn3AXdfbJtGOi13Xc+5x1G3e/zEbwDi+3HPpHm2dYdVAZu4/0/ss8w2nJuktM/jNJZV5C7rdKA0nd8hk3U89e7fiWlM+3mf75NxLk7J/ucOB7gA527xaRFZJSI3ikihO+5QnDukJSLyvoic22c5yQ0aOnDqP+r7DCsCEBFLRH4lIh+JSLOIhHHvUvos84s+25IHVKTaPmNMB86+mUT/ktPYHk+Pu15/n/Ul/9+fmDEmlPQ5rM/4z5O/iMj/iciLIrJaRFpwTpj4Nm8CBHBO2MGaRNLva4yJ4WzH5KRp+jY8Sd4PvRhjGnFuRE50B50IPGiM6XK//wDnN/mriHwqIpe5lbepNJL02xhj6uL7zV1G/JzZHPi5iITjH5w7/eqkZa0xxtj9bMPmwOZ95n8R586zqp+09TUVWG2MaU1z+v7Et7ex7whjzMvAL3ByZvUi8ryIfGWghYnIuSKyJGm7DmD9c6a/Y3wqzkVuaYpFT8DJHT7XZ79tRj/nk4hME5EnRGSleyy/ibOPy91JNmPgY/nzPt8nA58Z9+rrSr423eN+5gENIvKYiGzhjjsP52biGRGpFZEbRKQg1UqNMbYx5j5jzHdwck/fwAk29w2Qvvj/k3COL4D3kvZTE85N3mQ27jwecFnu8XcgTp3oZyKyWERmDbTAjAUeY8zHOGWGR6cxud3n+3LWbWzcFu5wjDH1xpg5xpitcO4sZgA/ccf9xxgzE+fkPR24UkT2GuRmHI1zB3METiQP4dQ3SZ/pNkv6fyrO3UlDn2EAiEgQ5yRcMYj01OMUOSWvb8ogltNX3/3/CE4RzjTjVEIfx7ptXo1z5z8tzWWlsoKk31dELJxtWr4Rae7rbuBYEanAyaHdHR9hjPnUGHOCMWYicDhOmf6x/SznOeB7aTSy+AK4uE9ALzTGHJJmer8A3u8zf4kxJmCMWeNOs6F9+TmwSdJN12DNxLn4v5VqpDHmVmPMdJzis/dxipZTpk9E9sSpXzsVKHfPmWdZ/5zpz+futFulGFeHc2c9o89+Cxpjru1nebfj5Bh2cI/lPeJJdf9+Qf/HMqS+Nk0VkeTtSb42RY0xVxpjdsU5xntw6nziNzFnG2O2BL6Jk5s8f4B1485njDH/xClm37nP6Kkp/l/BuhvSLfrsq3xjzKNs+DxOtqFlYYx51RhzME5AvxpYICJTUy8u88/xzMa5GPxGRKoBRKRKRC4SkZkDzHcPcISI7CciHhH5DvBd3IuJiMwUkc3dH78Zt0JbRPwicryIVLh3JGtxDpzoINNf7M5bD1gichJOjqevK0WkWESqgF/i1IckH7A/EpEtRSSAU1H7Kf2c5ANxl/kg8EsRqRSRIpyTfLgV49S7tYnIZsBP+6ThVuAaEdnObQk3SUS2dydZDWy2gRZA9wAXunejeTh30wbnoj9YzwMxnDvCpcaYxfERInKi20QanCKzKP0fE5fg5FYfF5HtRcTrfr7ZZ7q5wHkiMt09Rv0iUrOh3ECSRUChiPxURAqT9mNybnQ1A18Y3sIpo7/DPR4sEdlJRDZJJwEiMkFEfoRTj3BRqpyTiOzmbmMezoWqlXX7rg7nd0tOYzHO71APGBE5GKeYLi3GmFqc4qhbRWSKu1+miciWbs74RuA6EdnSTV+hiOw/wDYX4xQJNYtIJU5DimQ3AxeLyNfc/VcpziMg/XkaJ3f2E/c3/xJOCczv3fTsIyK7uMd/h/uJuuNmiUg8aDXjBKWUx6GI/NjdrhL3+9Y4Oeo/9Zn0F26aS3CuLc+7Aa4W5wbylqRrb0hEjhCRYBrnccKGliUi1SLyXREpdn+jtTiBvd/rbkYDjzHmJZws4nY4RV+tOC2KqoB+26IbY/6K01LsWpyN+C1wnDHm7+4ku7jzt+FUcP3LnRacu7cPRaQN5+S+1Bjz5iA34V6ck/sTnJZJ27H+Dx/DaSmzBKds91OcLHWyO3EqNetxAteh7g80GOfgVPZ+jNMS5yWck797kMtL5VScXEErzt3tw33G/xSnvPlp1rXU29Id9zDOBXONONnyVDmyK3EbTbjTfgPY1xjTNtgEu/vzfpxK/7v6jN4H+Jd7TPwZJzilfDbGGPMFTiu41Ti/awvOHd/lOCf+W+50z+KU7V+Pk7utxTkG08p9uNv6LZzj4SOcC9FLwA5Jk10OnCAia0Xkj/1s80E4J/h/cILqnRtIw8ki0iZOsdNinH1/kDHmxn6mL8ZpENPgfr4FHJO0Db8EHnN/6wtxcjcP4bTYqsfJfT414M5Y3/E4x/afcfb/H1hX/PhznBuUP7rb8DFwGv1fy87FabjUglOf/HSf8TcC1+Dc1Dbj7JNd+0uYMWYtTiA9ACfwPosTdG5wJ5kALMC5bq3CySWe6Y7bFef60eZu31s4LeZSacW5IfvMPW5fcqc/sc90D+I0aFqGc7E/IWncSTjXozfda++7ODfw8WLCgc7jvgZaloXTSGOZO+4G4PvGmH5LdTL+AOlYJs7DmS8bY/otlhERg9PC488ZSsM2OM12J7p3JkqpMU5G+cO72mXOKOMWMe7uFu9MwLljelODjlJqtNDAM/rk41SYNuMU73XgFn0opdRooEVtSimlskpzPEoppbJqVHcIWVFRYaZOnZrrZCil1Kjy9ttvNxhj+j7UmzWjOvBMnTqVxYsXb3hCpZRSCSKSTo8nGaNFbUoppbJKA49SSqms0sCjlFIqqzTwKKWUyioNPEoppbJKA49SSqms0sCjlFIqqzTwKKWUyqqMBR4RuUtE6kTkv0nDykTkJRFZ6v4tTRp3kYh8Is5rpvfLVLr6so1NQ2cDtW21NHQ2YJt0XqCplFJqsDKZ47kH2L/PsAuBV4wx04BX3O+IyHbALODL7jy3iIgng2kDnKDT2F5PXqePws4i8jp9NLbXa/BRSqkMyliXOcaYN2X9d24fCsxw/78X542AP3WHLzTGdOO8ce8T4KvA3zKVPoCO7jakKY8nbnuP1sYuisoDHHD6dnR42ygMFGdy1UopNW5lu45ngjFmFYD7N/4624nA8qTpVrjD1iMip4nIYhFZXF9fP6TESIfFs7e9T2tjFwCtjV08e9v7SIdWfSmlVKaMlE5CJcWwlC8KMsbcjvMiNGpqaob0MqFYFPKLfEw/opq8oEV3h82/XqwnFh3KUpVSSg0k24FnjYhsaoxZJSKbAnXu8BXA5KTpJgEZf5WzxwtfO6SE52/5FS31dRRXVrH/7AvxjJRwrJRSY1C2y5QWAce7/x8PPJU0fJaI5InI5sA04B+ZTowd6+T5W66ipd6Jfy31dTx/y1XYsc5Mr1oppcatjN3bi8hDOA0JKkRkBXApcBXwiIicDCwDjgQwxrwnIo8A7wNR4CxjTCxTaYuL2nYi6MS11NcR1VZtSimVMZls1XZ0P6P27mf6XwO/zlR6UvF4vRRXVvUKPsWVVXi0rE0ppTJmXDffMhLkgB9eRHGl07iuuLKKA354EUaCOU6ZUkqNXeP61t6I8O9XO9jrpF8kWrX9+9Vmps9K1chOKaXUcBjXgcfvsdn1wM158bb/Jh4g3ff07fF7tI5HKaUyZXwHHruZfJ+HA+bshIhgjMFnt+O3Y0B2ituMbdPR0kyspwePz0ewuASxxnUJqFJqjBvXgceIUPz6eXTseDp2oAqrq47gu7cRPeDa7KzftmlY/gVPXnN54jmiwy64hIrJm2nwUUqNWeP66uYprCS2508peGk2Rfd/lYKXZhPb86d4Ciuzsv6OluZE0AGnKfeT11xOR0tzVtavlFK5MK5zPJbHAxO2o+fEl5BYBOPx4ymsdIZnQaynJ+VzRLGenqysXymlcmFcBx5wgo9VsklO1u3x+VI/R+Tz5SQ9SimVDeO6qC3XgsUlHHbBJb2eIzrsgksIFpfkOGVKKZU54z7Hk0tiWVRM3oxjrrhOW7UppcYNDTw5JpZFQah0wxMqpdQYobfWSimlskpzPKOEbWyaupqIxCL4PX7KAmVYovcNSqnRZ9wHHtu26ejoIBqN4vV6CQaDWCOsjsU2NkvXLmXOq3Ooba+luqCaeXvNY1rpNA0+SqlRZ1wHHtu2qaurY+HChYTDYUKhELNmzaKqqmrQwcc2hoaeKBHb4LeECp8XS4bW6WhTV1Mi6ADUttcy59U5LDhwARX5FUNatlJKZdu4vl3u6OhIBB2AcDjMwoUL6ejoGNTybGP4oL2LA99eSs3f3ufAt5fyQXsXtjFDSmckFkkEnbja9loisciQlquUUrkwrgNPNBpNBJ24cDhMNBod1PIaeqKcsOQzlnc5AWF5V4QTlnxGQ8/glhfn9/ipLqgGYMeKHbl193k8PuNBghEvxtaetJVSo8u4Lmrzer2EQqFewScUCuH1Dm63RGyTCDpxy7siROyh5XjKAmXM22set7xzC2dNOp43brhJOxVVSo1a4/pqFQwGmTVrFqFQCCBRxxMMDu6VCH5LmBzw9xo2OeDHbw2tjscSi2ml07hsl5/zxo03aaeiSqlRbVzneCzLoqqqilNOOWVYWrVV+Lzcs8PmieK2yQE/9+ywORW+oe9mSyysGNqpqFJq1BvXgQec4FNYWDg8yxLhSwUBntl12rC2aovTTkWVUmPBuC5qywRLhCq/j0kBP1V+37AFHdBORZVSY8O4z/GMJtqpqFJqLNDAM4IZYxOJNGLbESzLT3u0iM6Ijd8bpLzYjzXERgtKKZULGnhGKGNs2to+5t0lp9HVtZJAYCKbbXUTP3y0mfq2Hu74QQ3bTCjS4KOUGnW0jGaEikQaE0EHoKtrJV98cjY/2XcCK9Z2cup9i2ls154LlFKjj+Z4RijbjtDVtZKiop0pn3w64g1homEmFBbyyJFbc+XfVxOJxnKdTKWU2mgaeEYoy/JTXr4P3k1OYPaffpnolfqGPa5jk7vnM//kM/D7NMOqlBp9xv2Vy9g20YYGempriTY0jJi+z/z+cjbd8mLOd4MOOB2DnvPm+cgxh9P243Mp6mzNcSqVUmrj5STwiMiPROQ9EfmviDwkIgERKRORl0Rkqfs34++DNrZN98dL+XzmTD7Za28+nzmT7o+XjojgI2IRw5OyV2q7vISelbWgPRYopUahrAceEZkIzAFqjDHbAx5gFnAh8IoxZhrwivs9o2JNTaw4a7ZzEQd6Vtay4qzZxJqaMr3qtCT3Sh1XXVCN1diMb2I14vf3M6dSSo1cuSpq8wL5IuIFgkAtcChwrzv+XuCwTCfCRCKJoBPXs7IWExkZrcXivVLHg091QTW/2+UyzINPMOnGm8CyRkTuTCmlNkbWGxcYY1aKyLXAMqATeNEY86KITDDGrHKnWSUiVZlOi/j9+CZW9wo+IyknEe+VesGBC4jEIvhsobjd0HPkUay67DJiDfVMuvkW8raepr0XKKVGjVwUtZXi5G42B6qBAhE5biPmP01EFovI4vr6+iGlxVNWxqSbb8E30clR+CZWM+nmW/CUlQ1pucPJEouK/AqqC6spjfhYftz3WX76GXT95z8jrmhQKaXSkVaOx62X2Sx5emPMm4Nc5z7AZ8aYenfZfwB2B9aIyKZubmdToC7VzMaY24HbAWpqaob0hjWxLPK2nsbUhx/GRCKI34+nrGzE5h5GetGgUkqlY4OBR0SuBmYC7wPxJxYNMNjAswz4mogEcYra9gYWA+3A8cBV7t+nBrn8jSKWhbeiIhurGrKRXjSolFLpSCfHcxiwjTGmezhWaIx5S0QeA/4FRIF3cHIwhcAjInIyTnA6cjjWN5bEiwbjLfFGYtGgUkptiBgzcGmViDwHHGmMactOktJXU1NjFi9enOtkZJWxbWJNTaOiaFApNTKJyNvGmJpcrT+dHE8H8G8ReQVI5HqMMXMylirVr9FUNKiUUqmkE3gWuR+llFJqyDYYeIwx94qIH9jaHfSRMUb7alFKKTUo6bRqm4HTk8DngACTReT4ITSnVumybeioh2gEvH4IVoLW5yilRrl0itquA/Y1xnwEICJbAw8Bu2YyYeOebUPd+7DwaAgvg9AUmPUQVG2nwUcpNaqlcwXzxYMOgDHmY8CXuSQpwMnpxIMOOH8XHu0MV0qpUSydHM9iEfk9cL/7/Vjg7cwlSQFO8Vo86MSFlznDlVJqFEsnx3Mm8B7OqwzOwenB4IxMJkrh1OmEpvQeFpriDFdKqVEsnVZt3cD17kdlS7DSqdPpW8cTrMx1ypRSakj6DTwi8ogx5igRWYLTN1svxpgdM5qy8c6ynIYEp7ysrdqUUmPKQDmec9y/B2UjISoFy4LCCblOhVJKDat+b5/jL2UDZhtjvkj+ALOzkzyllFJjTTrlNt9OMew7w50QpZRS48NAdTxn4uRsthSRd5NGFQF/zXTClFJKjU0D1fE8CDwHXAlcmDS81Rij71pWSik1KAPV8TQbYz4HbgCakup3ekRkt2wlUCml1NiSTh3PfCD5JXDt7jCllFJqo6UTeMQkvabUGGOTXlc7Siml1HrSCTyfisgcEfG5n3OATzOdMKWUUmNTOoHnDGB3YCWwAtgNOC2TiVJKKTV2pdNXWx0wKwtpyQlj23S0NBPr6cHj8xEsLkG0WxqllMqYdN5AGgBOBr4MBOLDjTEnZTBdWWFsm4blX/DkNZfTUl9HcWUVh11wCRWTN9Pgo5RSGZJOI4H7gQ+B/YBf4byP54NMJipbOlqa+cujC9jzpDMJFBXT1drCXx5dwLdPOYuCUGmuk6eUUmNSOoFnK2PMkSJyqDHmXhF5EHgh0wnLBmMbdjr0KJ565lnC4TChUIhDDz0KY6/XGbdSSqlhkk55Uo/7Nywi2wMlwNSMpSiLYpaVCDoA4XCYp555lpgWsymlVMakk+O5XURKgYuBRUAhcElGU5UlBhJBJy4cDq//8qFxwBibSKQR245gWX78/nJENAArpYbfgIFHnCtPizFmLfAmsEVWUpUlXq+XUCjUK/iEQiG83vH1fKwxNm1tH/PuktPo6lpJIDCRHXe4ncLCrTX4KKWG3YBXFbeXgrOzlJasCwaDzJo1i1AoBDhBZ9asWQSDwRynLLsikcZE0AHo6lrJu0tOIxJpzHHKlFJjUTq39i+JyI+Bh3H6aQNgLPRQbVkWVVVVnHLKKUSjUbxeL8FgEGuc1fHYdiQRdOK6ulZim0iOUqSUGsvSCTzx53XOShpmGCPFbpZlUVhYmOtk5JRl+QkEJvYKPoHARCzx5zBVSqmxaoO39saYzVN8hhR0RCQkIo+JyIci8oGIfF1EykTkJRFZ6v7VB2myxO8vZ8cdbicQmAiQqOPx+8tznDKl1FjUb+ARkd8k/Z/q9ddDcQPwvDFmW2AnnAdSLwReMcZMA16h98vnVAaJWBQWbk3Nro+z++5vUrPr49qwQCmVMQNdWfZP+v/q4VqhiBQDewC/BzDGRIwxYeBQ4F53snuBw4ZrnWOBbWwaOhuobaulobMB29jDunwRi7y8SvIDE8nLq9Sgo5TKmFxcXbYA6oG7ReQdEblTRAqACcaYVQDu36pUM4vIaSKyWEQW19fXZy/VOWQbm6Vrl3LsM8ey3+P7cewzx7J07dJhDz5KKZUNAwWeKhE5T0TOT/o/8RnCOr3AV4D5xphdcFrKpV2sZoy53RhTY4ypqaysHEIyRo+mribmvDqH2vZaAGrba5nz6hyaukZ9w0Kl1Dg0UOC5AyjC6akg/n/yZ7BWACuMMW+53x/DCURrRGRTAPdv3RDWMWoY26Y9vJaW+jraw2sx9vq5mEgskgg6cbXttURi2txZKTX69Nuc2hhzWSZWaIxZLSLLRWQbY8xHwN7A++7neOAq9+9TmVj/SJLuaxn8Hj/VBdW9gk91QTV+jzZ3VkqNPrmqQf4hsEBE3gV2Bn6DE3C+LSJLgW+738e0jpbmRNABaKmv48lrLqejpbnXdGWBMubtNY/qgmrACTrz9ppHWaAs62lWSqmhykmnZMaYfwM1KUbtne20DLeN6Wwz1tOTCDpxLfV1xHp6eg2zxGJa6TQWHLiASCyC3+OnLFCGpS3PlFKj0PjqDTPDNrazTY/PR3FlVa/gU1xZhcfnW29aSywq8isymn6llMqGDd4y923N5n5OFpGds5HA0WRjO9sMFpdw2AWXUFzptByP1/EEi0uylmallMq2dHI8Ne7naff7gcA/gTNE5FFjzG8zlbjRZmM72xTLomLyZhxzxXXEenrw+HwEi0t6NSxQSqmxJp3AUw58xRjTBiAil+I0gd4DeBvQwOMaTGebYlkUhLRbOqXU+JHOrfUUIPmWvQfYzBjTCXRnJFWjlHa2qZRSG5ZOjudB4O8iEn+u5mDgIbebm/czlrJRKLmzTdtEsERfIa2UUn1tMPAYYy4XkeeA6YAAZxhjFrujj81k4kajeGebQ2Hbhsb2CJFoDL/XQ3mBH8uSYUqhUkrlVrrNqd8BauPTi8gUY8yyjKVqHLNtw0drWjn1vsWsWNvJpNJ87vhBDdtMKNLgo5QaE9JpTv1DYA3wEvBH4Bn3r8qAxvZIIugArFjbyan3LaaxXftlU0qNDcg/R1MAACAASURBVOnkeM4BtjHGpH4YRQ2rSDSWCDpxK9Z2EonGcpQipZQaXunUei8Hmjc4lRoWfq+HSaX5vYZNKs3H7/XkKEVKKTW80snxfAq8LiLPkNR82hhzfcZSNY6VF/i54wc169XxlBdoT9RKqbEhncCzzP343Y/KIMsStplQxBOzp290qzbb2DR1NWlHokqpES2d5tQZeS+P6p9lCZVFeRs1T/z12PE3lcZfnTCtdJoGH6XUiNLvFUlE5rp/nxaRRX0/2UuiSoe+HlspNVoMlOO53/17bTYSooZGX4+tlBotBnr19dvu3zdExA9sCxjgI2P66W5Z5Yy+HlspNVqk8wDpgcD/gHnATcAnIvKdTCdMbRx9PfbGs21DfWs3K9d2UN/ajW2bXCdJqXEhnVZt1wHfMsZ8AiAiW+L0XvBcJhOmNo6+HnvjaNdESuVOOlelunjQcX0K1PU3scqd+Ouxqwurqciv0KAzAO2aSKncSSfH856IPAs8glPHcyTwTxH5LoAx5g8ZTJ9SGaFdEymVO+kEngBOJ6F7ut/rgTKc9/IYQAPPMNMHQTMv3jVRcvDRromUyo50HiA9MRsJUQ59EDQ7tGsipXJHjBm4JY+IbAHcAHwNJ4fzN+BcY8xnmU/ewGpqaszixYs3POEAjLGJRBqx7QiWlfs3hjZ0NnDsM8eu1yx6wYELqMivyFm6xiJ94Z4ar0TkbWNMTa7Wn+6rr28GDne/zwIWArtlKlHZYoxNW9vHvLvkNLq6VhIITGTHHW6nsHDrnAUffRA0ewbTNZFSaujSubqKMeZ+Y0zU/TyAk/MZ9SKRxkTQAejqWsm7S04jEsndq4fiD4Im0wdBlVJjSTqB5zURuVBEporIZiLyE+AZESkTkVH9dKJtRxJBJ66rayV2Djtm0AdBlVJjXTpFbTPdv6f3GX4STs5ni2FNURZZlp9AYGKv4BMITMSS3OUu9EFQpdRYl06rts2zkZBc8PvL2XGH29er4/H7y3OarviDoIOlleZKqZEsnRwPIrI9sB3OMz0AGGPuG8qKRcQDLAZWGmMOcovtHgamAp8DRxlj1g5lHRtOg0Vh4dbU7Po4tolgSe5btQ2VdgWjlBrp0ukk9FLgRvfzLeC3wCHDsO5zgA+Svl8IvGKMmQa84n7POBGLvLxK8gMTycurHNVBB7QrGKXUyJfOVfZ7wN7Aavdh0p2AIbVBFZFJwIHAnUmDDwXudf+/FzhsKOsYr7QrGKXUSJdO4Ok0xthAVESKcToIHWqDgrnATwA7adgEY8wqAPdvVaoZReQ0EVksIovr6+uHmIyxJ94VTDLtCkYpNZKkE3gWi0gIuAN4G/gX8I/BrlBEDsLp8frtwcxvjLndGFNjjKmprKwcbDLGrHhXMPHgo13BKKVGmnRatc12/71VRJ4Hio0x7w5hndOBQ0TkAJzGCsUi8gCwRkQ2NcasEpFN0VcvDIplCdtMKOKJ2dO1VZtSakRKqyZdRHYUkUOArwBbxV+JMBjGmIuMMZOMMVNxut951RhzHLAION6d7HjgqcGuY7yLdwUzsTRIZVGeBh2l1IiywRyPiNwF7Ai8x7o6mUy8DuEq4BERORlYhvPeH6WUUmNMOs/xfM0Ys10mVm6MeR143f2/Eaf1nFJKqTEsnaK2v4lIRgKPUkqp8SedHM+9OMFnNdANCGCMMTtmNGVKKaXGpHQCz13A94El9H7uZkwYaS+CU0qpsS6dwLPMGLMo4ynJgZH4IjillBrr0gk8H4rIg8DTOEVtABhjhrtVW9ZFIo18+ulcNp94KR6rlJi9lk8/ncu2215OXp4+nKqUUpmQTuDJxwk4+yYNy0Rz6qwztk1F8FSeueZmWurrKK6s4oBzzsLpIWh8MbZNR0szsZ4ePD4fweISxNJcn1Jq+KXTc8GJ2UhILvR0Cs/e4AQdgJb6Op694WZm/eoqAoENzDyGGNumYfkXPHnN5YkAfNgFl1AxeTMNPkqpYZfOaxF+KyLFIuITkVdEpEFEjstG4jLNxEgEnbiW+jrMOOvIuaOlORF0wNkHT15zOR0tzTlOmVJqLErndnZfY0wLcBCwAtgauCCjqcoSj89HcWXvTrCLK6vw+Hw5SlFuxHp6UgbgWE9PjlKklBrL0gk88avwAcBDxpimDKYnq4LFJRx2wSWJ4BMvYgoWl+Q4ZdmlAVgplU1ijBl4ApGrcF7K1gl8FQgBfzTG7Jb55A2spqbGLF68eEjL0Ep1reNRarwRkbeNMTU5W/+GAg+AiJQCLcaYmIgEcV6NsDrjqduA4Qg8yqEBWKnxI9eBJ53GBUcCUTfoXAw8AFRnPGUqq8SyKAiVUlxZRUGoVIOOUipj0rm6XGKMaRWRbwD74fTdNj+zyVJKKTVWpRN44o2LDwTmG2OeAvQ9ykoppQYlncCzUkRuA44CnhWRvDTnU0oppdaTTgA5CngB2N8YEwbKGCPP8SillMq+DQYeY0wH8D9gPxE5G6gyxryY8ZQppZQak9Jp1XYOsACocj8PiMgPM50wpZRSY1M6vVOfDOxmjGkHEJGrgb8BN2YyYUoppcamdOp4hHUt23D/l8wkRyml1FiXTo7nbuAtEXnC/X4Y8PvMJUkppdRYls77eK4XkdeBb+DkdE40xryT6YQppZQamwYMPCJiAe8aY7YH/pWdJCmlVPbZtk1HRwfRaBSv10swGMTSrqMyYsDAY4yxReQ/IjLFGLMsW4lSSqlssm2buro6Fi5cSDgcJhQKMWvWLKqqqjT4ZEA6e3RT4D337aOL4p9MJ0wppbKlo6MjEXQAwuEwCxcupKOjI8cpG5vSaVxwWcZToZRSORSNRhNBJy4cDhONRnOUorGt38AjIlsBE4wxb/QZvgewMtMJU0qpbPF6vYRCoV7BJxQK4fWmc2+uNtZARW1zgdYUwzvccUopNSYEg0FmzZpFKBQCSNTxBIPBHKdsbBoonE81xrzbd6AxZrGITM1YipRSKsssy6KqqopTTjlFW7VlwUB7NTDAuPzBrlBEJovIayLygYi85/YFh4iUichLIrLU/Vs62HUopdTGsiyLwsJCQqEQhYWFGnQyaKA9+08RObXvQBE5GXh7COuMAucbY74EfA04S0S2Ay4EXjHGTANecb8rpZQaYwYqajsXeEJEjmVdoKnBefvo4YNdoTFmFbDK/b9VRD4AJgKHAjPcye4FXgd+Otj1KKWUGpn6DTzGmDXA7iLyLWB7d/AzxphXh2vlbl3RLsBbOC3o4gFplYhU9TPPacBpAFOmTBmupCillMqSdPpqew14bbhXLCKFwOPAucaYFpH0Orw2xtwO3A5QU1NjhjtdSimlMisntWci4sMJOguMMX9wB68RkU3d8ZsCdblIm1JKqczKeuARJ2vze+ADY8z1SaMWAce7/x8PPJXttI1Itg1tayC83Plr27lOkVJKDUkuHsudDnwfWCIi/3aH/Qy4CnjEbTW3DDgyB2kbWWwb6t6HhUdDeBmEpsCsh6BqOxgFTT1tY9PU1UQkFsHv8VMWKMOSkZ9upVRmZT3wGGP+TP9vMN07m2kZ8Trq1wUdcP4uPBpOeRkKJ+Q2bRtgG5ula5cy59U51LbXUl1Qzby95jGtdJoGH6XGOb0CjGTRyLqgExde5gwf4Zq6mhJBB6C2vZY5r86hqaspxylTSuWaBp6RzOt3iteShaY4w0e4SCySCDpxte21RGIjP2gqpTJLA89IFqx06nTiwSdexxOszG260uD3+KkuqO41rLqgGr9n5AdNpVRmaZ/fI5llOQ0JTnnZKV7z+p2gMwoaFpQFypi317z16njKAmW5TppSKsc08Ix0ljXiGxKkYonFtNJpLDhwgbZqU0r1ooFHZYwlFhX5FblOhlJqhBn3gcfYNrGmJkwkgvj9eMrKkFFQlKWUUqPVuL7CGtum++OP+XzmTD7Za28+nzmT7o8/xmjvAEoplTHjOvBEGxtZcdZZ9Kx0mv32rKxlxVlnEW1szHHKBsfYNtGGBnpqa4k2NGgAVUqNSOO6qC3WHcGqrCT4m6uwyyuwGhuI/e46Yt0RfLlO3EZycm9LWXHWbHpW1uKbWM2km28hb+tpWnSolBpRxvUVKRYI0H7t7/ieXcDXlzfzPbuA9mt/Ryww0Fu/R6ZYU1Mi6EA89zabWJP2FKCUGlnGdeBpLijkpOVrWd7lPE2/vCvCScvX0lxQmOOUbTwTiSSCTlzPylpMRHsKUEqNLOM68NhIIujELe+KYPfbh+nIJX4/vom9ewrwTaxG/NpTgFJqZBnXgcfvESYHel+YJwf8+D2jL/B4ysqYdPMtieATr+PxlGlPAUqpkWVcNy6o8Hm5Z4fNOWHJZyzvijA54OeeHTanwjf6dotYFnlbT2Pqww/rM0lqRLJtQ2N7hEg0ht/robzAj2WNvps8NXSj7wo7jCwRtgn6WbRjJVE7gtfyU5nvx5LReTKIZeGtGBk9BehL4FQy2zZ8tKaVU+9bzIq1nUwqzeeOH9SwzYQiDT7j0LgOPLax+V/4E31Z2TAb6CVwAkQijdh2BLHyaYvZRGwNTmNdY3skEXQAVqzt5NT7FrPo7OnEbMZMLsi2bTo6OohGo3i9XoLBIJaWOqxnXAee/l5WtuDABeOqj7HhPln6268LD3oIb08d7y45DZ+vkuBmP+X8Ny/RoD8ORKKxRNCJqyzMY1W4i9MfeHtM5IJs26auro6FCxcSDocJhULMmjWLqqoqDT59jOu9oS8rW3ey3HnnncydO5c777yTuro67CH0etDffvWZTt5dchpdXSspn3x6IujEx4+3N5TatqG+tZuVazuob+3Gtk2uk5Qxfq+HSaX5vYbN2XtaIujAulxQY/voPP86OjoSQQcgHA6zcOFCOjo6cpyykWdcB56x+rKyjek6JxMnS3/7VYxNV9dKAMQbGtdBP17ncfgtf2H61a9x+C1/4aM1rWM2+JQX+LnjBzWJ4DOpNJ/NKwrWywWtWNtJJBrLRRKHLBqNJs6juHA4TDQazVGKRq5xXdQWygsx91tzOfe1cxPFPXO/NZdQXijXSRu0je06Z6CTZbANBPp7CZzXEyAQmEhX10pMNEx1QXWv4FNdUI3fylzQH0mtqvqr83hi9nQqi/JykqZMsixhmwlFPDF7emL/Gwz7blfFEbtOJpTvI9zZw+NvL8fv9eQ6uYPi9XoJhUK9zqdQKITXO64vsymJMaP3DqumpsYsXrx40POv7Wyita6OtvZmrPw87M5uCgtKKKqqojR/dD7/Em1o4POZM3v1YuCbWM3Uhx9O2eKtra2NO++8c72T5dRTT2VF94pBN7ywjU24O0xXtAvb2AQ8AUL+EjrbPuLd92enrOP53a6XMy20Fb7Sdft+uF5bMdJaVa1c28H0q18DYJfJIc6YsSWhfB+TSvPZtCR/VNZxbKxo1ObDNa2ckVTHc+txu7LthCK83tFXGDOa6nhE5G1jTE3O1j+uA0/jGh679CJa6usSw4orq/jeZVdSWj763voJ0FNbyyd77b3e8K1efQVfdfV6w/s7WbzFXo599tj1ciR373c/Pko2mFuIt2x75pM/csS0Q7EsLwHLjyx6Be9Wm2CVl2C6bNotD6ayHHvpp/TccCcTr72GaFEhweISgGHr+LS+tZvDb/lLr6KdSaX5OcthxNNTWZjHj/fbhp8+/u6ICIjZNNJ+k+EwWlq15TrwjO88YMzuFXQA53tseF4nkItnWeJd5/TN8fTXdY5lWVRVVXHKKaf0OllWd6xOWQezItzKeQ/+d4MXx6auJp755I8csMX+nPbKWetyTfteT8END9L62B8S0065/z6WnfRDfBOrWbP8c1548G4Ou+ASSotKqL9xHhMuvAgrVIIdbia86CkKTzwBG/D4fASLS9IKQqlaVeWyPiFe57G6uSsRdOJpGstFbslG2m8yHCzLorBw9PX1mG3jOvD4/HnsevB32W7GvliWhW3bvP/6i/j8Qz/hB3qWJZPBJ951Tt9cwkBd56Q6WeINBPrmeMLtdloXx0gswlFbf5eTXz6jd8u1N87j/nNuBTfw+CZWY4eb8U2spvyqK3nuyYW01Nfx5DWXc/SlV1F23HGsuvhielbWUrjP3uSdO4eHLruQlvo6iiurOOyCS6iYvNkGg0+8VVXfu+tM1SdsqD4pXudRkOcZcxffdGX7N1Ejx7gOPIGCYraZPoOnrvpF4kJ20PkXEygoHvKyc/WM0GC6zklVj1LmDzHvWzcw57VzEoHzit2v4zdPrAY2fHHM9wbotKMpc01Rv3Nh8U2sZtJNNyGhEP7Lf8mHyz7j0LMvwevxETMxIvkxWirzyLv8Iqwb7iT/iCN4Yu5VBEOl7HnSmQSKimlpbqagtDVRNNefeA6jbx1PecHGN2YwtqGjNYIdtbG8FsEiP5IUVPrWJ+27XRUXH7gdHkt6BSHLEvJ93jF18d2YBhzD+Zuo0WVcB562lmbeenwBu516Ev7iAiIt7bz1+AJmnDibUHn5kJady2eENqbrnFSt4CbfeSe+ohhl3X6u3eVaPHkeLGNRFZzA1YdUsKqlhwffXt3vxdEYG2/PGvKsYm7Z+xbyvfk0R5q5a8ldNHQ24BcvW736SiLItTc389n/PmG3GUfQ+vIyAl/LR4qEPBPkhbbFPFb7OL/7zWWURkNM3n4nvnHgd+lqb6e1rYN/fvghe1ZUESgsGrAsPVWrqsG0ajO2obG2jWfnL6G1sYui8gAHnLkD5dWFieCT3GJtl8khjt99c465862UdTgbe/Hd2JZ5xrbpaGkm1tOzUUWTg5nXtg2fN7bzRWMHQb+HjkiMzcqDTC0vSJnG4fpN1OgzrgMPGLY65mDO/+tPE3f11x1zNTD0BhdekZRFVd4R1g9cqhfI0bqGjmiUBc8sJhwOM2nSJPbaay/uu+feRAOEXxw1k7Jg6ve0RiKN1K56EsoO4Yq/X7EuxzT9CioCpZSYIOGifCLRGPmdESK+HrY6dF/WhsMUH1xErPN/eMmnp62TY7bamzdWvs6P3rmUh/a6hz32/Da1J5yQCJL7XnMNL77xBgcdfHDKsvV4PZtt2/i7IS8GQZ+PYEFJr1xKujpaI4mgA9Da2MWz85dwxE92paDEKXZMrrs4Y8aW69Xh/O6lj/jlIdtjjMHv9TCtsjCti+/Gtswztk3D8i948prLN1g02TfI5BcW0bhyeVrzxoU7I6xp6eKSp/6bSN8139uRUNBHWUHqIlnLkkRx7Uhq7q4ya+Q1t8iiSF4sEXTAyZGc/9efEskbevl6kcfium/+MvEgZXVBNdd985cUeUbWLk/1AjlPVQVRX3GiifX06dNZtGhRr4dMH33kYTo713/I1BibWKydkglHMuf183rt24v/cjGF3gJWdFscfstfuOGVD1nd8SnHP/999n9if074x6msiHVQ+O5r5N+6P8UPn4M/vIord/8V9+5xF0QtuoqLKTrySCrn30LZjfPwVpRz6Df3wNvW1utBWWPbtLc083HTx1z+18sJr1jB67+/jTWf/Y/w6lU0168h2uPkPm1j09DZQG1bLQ2dDdhm3XJs26atrY1wOExbWxsYkwg6ca2NXdhRk1gOEks8KBnK9/UqRovngI667W+JB0eX1rdRXuBnYmmQyqK8xMU2uWeDpvZuVnf30GYZLv7u9uw8OdTrSf/1tiEWpWNtQyJwAIm6s46W5t6/mRugHrz4fO44+yQevPh8GlYs4y+PPDDgvH17XuiJ2lzw2LtUFuZx2/d35bojd6Krx6Ynun5jnb7zRqP2iH+gtu+xYNv2esNisdiA34fSI8hYMuKaU4vI/sANgAe40xhzVX/TDrU5dW3LcvZ74oD1hr9w+LNUF08e9HIBurvr+eDDiymacATiDWGiYVrXPM6Xtr2CvLzKIS17OKV67mfzN18g2ryMO90czwknnMA999yz3rznnnsuoVDvh227u+tp7/iUTquK/Z88ZL15Xjj8GS57cjV1rd3MP2YKx79y2nq5wgV7z6fi8VNhxWLMNgdR//Vf8dS1v6alvo4ta3bj60cczaLrf5O4Ez/0/J/jLa2gI+bBb0Gpv5NWbLpMDK/xkNcl9HR2kFdQiB2LYdsxjDFYYhEpgDzLj+k02LEolteHx7KwYzHAxrZjiFhEe6LYloeYvxCfWHi72+ls66KjJcaHbzWz56wtsQMxOrq89NgGnyW8/UUTu26Sh9fYeD0WYnkwJoYds/F5vPh9QSIixGJgYgbxQCDQQ7TbRyzq5AYssfHkRVmBj6ZIjCmxbiQWxev1EfA7r2iP9QhgMHYnth0Fr4+uQAFiWcS6Y0RjNn4Bb6SVjkABPp+PUGsLVk+EmMdHV7CIoLfDyee3xZCeHozPB4Ww5tOP+dvDT7Bq6ccAnDTv9/TkFVLg8SBWGEMP4KObEO0IMQxeEe5//X/847MmLpwxiS3LA+T5vPT4CzDRCB6BvK5Owi2dNEYM3QUllBXlsaalm6uf+xBwcorlBX6mlOaDCD0xG7/XotSyIGKDT+jq6cCO9l8MmJyLw/LgsYJgpaqT678JdPI427Z54YUX+Oijj9Y9duD18tJLL7HzzjsTCoUoLAiSF23FQxRj+WmXILZt8Hg8iUDl8/tp7xF6bIPXEorzhPxAoFdRcfJ6RQSPx0Ms5hy3w9FMW5tTJxERD3Az8G1gBfBPEVlkjHk/E+vziSdlcZhPhl6x6/eXs+UWP0r0TRYITGTHHW7H7x9a3dFwS9UKrsNbgjdYwaxD92PhUy/Q2dmZ9hPZtjE0dk2ipJDUPRNEI/x2Dx+rYpXY0fbU9WCdTTD9XHj4ODq2/0Ei6AB8ec99EkEHnDvxp677NTue+Qtm3v+e+xDizoi/jhc+e4FjKg7ltccfZLfDj6KztYXn589NBKxDzvsZMV8e0eYOnrn+14nh3zn7fDxeL3+ce3Vi2L6nn8M7zy9ih4Nn4vP7eeH6XyUt5+fE/BFWNFmc/oBTl7PfdlX88pvlPH3FpYnpDv7RRfz9Dwv53+K3nPnO+Rnewgk8c5NTdDd1p3L+74DNef72fyfqj/b+wZfIL/bgKzSEGlfz9Nwr1y3v/J/jD1Tytyc/Y4c98nlh/lWJcfuc9zNa8yuZfd+6hzPnf39XnlwT5rRIC7XnzEn83oXXzqVh0kTKGupYdfZZieGb3HQjPf6/ss/smbx8y8O0h8PUd0QpikWJFnzOf5ec6fS7V74v3s1/y4nvLU+81+qub0zluC8V8Px1l/Hn+jq2/L+v8X9Hn8if3nyTfbfZhpUXXEDPyloCE6upuHYuZz4bpr6th5uO2YXuHpvzH/0PlYV5/GT/bbjgsXXPON32vZ2Y8K96YttbLLr5qn6LAVMVM+535oUsebOT3Q7eIlEnN9BDn8B64w455BDa29tZsWIFCxcu5Lvf/S677bYbb731Fnt88xt4Gj/G/9RJEF4GoSn4D7ubp/62lD323JMlS5bQ1dXFlrvuwRkL3ln30Oyxu1Bd0E2opCTRujZ5vdtssw177LEHjz766Ih/MDVdIy3VXwU+McZ8aoyJAAuBQzO1MsvycsX0K3oVh10x/Qosa+jxWMSisHBranZ9nN13f5OaXR+nsHBrZIT1vJzcCm6rV19h6sMPU1JcSKN/Im1SwPePO5YJEzZh5syZidxN/MAPBoPrLa+lO8iJ933Eo2+1MG/G9b327bw9r6PsT7+j5MkfsHmJhT+vKHVfeW11kF8KQCxY1etZq0BhYcpnrwp9zh3sirWdnPHAv2loj3DMlCN5+vpf8+U99+kVdOLzLLr+N+Sb/ETQiQ9/7qbr6Gxt6TXsxdtu4Mt77sOrN11FtLm+z3J+jd1pc/oD/0oUqx23Uzl/vO6KXtM9/bsr+fKe+6yb74bf0Fq/NlF096WvV/P87f/tVX/0yn0f0NoYoaSrkzfcoJNY3nW/pqU+zLa7lSSCTnzcy9f/hs6mtb3qls68/23OKPDR6gYdcOr02n58LsUdHYmgEx+++uwfsmnxTJZ+fgHfOO4oZpx1IaHCEAXBdv773pmJfvcKqk9MBB1wXh9/0nuf09zdkUjTdnvtx6OPPspu225Lkxt0ktd/0dc2YcXaTta293D+o/9xfscZWyaCTnwbTn/sP3imlyWCTnx7+xYDdrQ0r1fM+ML8q9h2txKenb+EjlYnrQP1VZhq3KJFi5g+fXrie0FBAYsWLWLnnXdGOhsJxoMOQHgZ+U+eyPSdt+aRRx5hl112YZevrws6ieN1wTu0RSXRP2Lf9e68886JoNM3jaPViMrxABOB5UnfVwC7JU8gIqcBpwFMmTJlSCvrjPUw919z+clXf0KJv4TmSDNz/zWXq/f47ZCWGydijahitf6kagW3WXkhjQE/XW5Fb1VJyXoPmaa624raFivWdnLNy5/yRXhT7t/3LqJ2BL8dpexPc7HeecBZp4lRFhPmfWsuc5L6ypv3tUsp+/PNsNMsCE3BFFZSXLku+HS1tfX6Dk5vE02d6+rlVqztJN8zBZ+xaKmvI+A2OkgVsEwsmnK4Ly+w3rB40Es1zsR6PwxZlu9JudxAUgOIlvo6vEmN1/KC3pT1R16/BxPrTrk8rx+8fmvAYJy8XwJ2lIY+dXo9K2vxmdh6dX09K2vxxCwnVzOpmuPuXcYd35uI5Ykmgg4A3gqWd7X0mnd5VwQJrsvdB4qcOsOCvDwaU6ynIuCkNehf91xT3/qx+DYYSf3gd6ynJ/E91tOTcpq8oJWok4MNd+yZalx+vlt/Fwph23ZiWD7RdUEnMcMyCvP9hMNhLMvCmNTPbRmsxDr7pik/P3/MdT46sm6/IVUTll6VUMaY240xNcaYmsrKoV3UA948GjobOPe1cznxhRM597VzaehsIOAZ20+MpyPe2ihe4e3xOA+ZhkIhCgsL+83iJ3d//8jiVSxf3kr1PYdQMa8mEXQITQFfABHYyl/Ggm/fzgvfvpsFO57DtD/fjOz5PodnagAADQ1JREFUE6LVNbw782kua+zhm+deRHGlU/Tx3hsvc9B5674XV1Yx46wLuf7P6y5mk0rz6Yy10iM2xZVVdLW10dPdlZgnrriyCvF4Uw7v6e5ab1g86KUaJ57e3f43dcZSLrerra3X92hS6/rujihF5b2DWlF5gGgk1m86oxHo7rBTjmvr6V1/O6k0ny7Li29i71ymb2I1PeJJOTzmsZ2OXSNCfVuEiDHYMS+BwMR1E0YbmBzo3fx7csCP6Vi3rV2tLYRCIdq7u1Oup6HLSWtHZF3DjHBnz3qvUphUmo8YK+X2enzrWll6fL6U03R32BSVB7C8zqUm3rFnsngxcn/j4kXP8Tqe+LDOnphzbPeaYQptnZFEkBITS71N2Imi677rja8vVRpHq5EWeFYAybX6k4DafqYdsrJ8pxflXsVBe82jbJR2EDoS9O3+/va3W+g5csG6EzI0BTPrQWxPIe12AKung/KYzSZ5xZSVbw3fvhSTV0x9rBDJr+BPzZ2cF47x1Z/9muPn3cbE7+7FrbX3s/OPTuaEG+Zz9KVX4C0qor7NuYLH63gqCvw8uOxRDj7v57z3xsvkFxWz/5nn9gpYh5z3M15v/BMHnvfzXsO/c/b55BcV9xq27+nn8N4bL7PX2RfiLansNe6g887HBAy3HfeVxHY/8J9GDjr/4l7THfyji3jvjZfXrf+cn1FUWZoINh/8rZb9T9s+8T1ex1NU5qcrmM+e5/YOuAf+6EKKK0N8+FYz+515Ya9x+5z3M/LLSnu9hmD+93fl1vYeim6Yl7j4x+t4WoJBNr3p5l7DN7npRla1PMz2X7qFO/7ezPxjv8K9b39Be0cB2395fiL4tNfezd1fnpwIPpMDfu768lQ2KSpKpOn9V1/gyCOP5K0PP6TsmmvWW/+Vf1/NpNJ8NisPcvuxuzi/4+v/45rv7dhrG2773k7E/tLEIWf13t7DLrik10PEweISDrvgkl7T7Hfmhfx/e/ceI1dZxnH8+zszs9vd7ZZyaRssINVAFJC0CoTKxapgENGC2m7BBIg2AgGKRgSiIVaSJsYLoITQKFZRQUTAipGkIAIFxFLoLpdyk3DRFkKL2NIte5uZxz/OO2W6zOyt23POsM8naXbmzJw5v3k7O8+e97znPc+u2crJ532E1vY4a2trK4sWLarZjVzrsY6ODmbOnMnixYuZPn06U6ZMoaOjg66uLqxlb96ev2Knz3rPqb/ioa7nWbhwIZ2dnXQ+vJrl4f3t+Lx+ZQ6T87aj63rwdru6uliwYMGIurobRaZGtUnKA88DnwY2AmuBM8xsfa3n7+qoNkhnPrX3uurzMQr5iOYcTOr/H3kbIMo3o7ZpEEXxyatvd9NMLxR7oVyCXIEtub3pLRpNOdEXQZG4T7g9L3qKWyiWBigoYk8KRJQol8u8ae30lSOatfOotmYVKPTEf91GUUSpNICVy0S5PNaSY3tpO2251jCqrUSUz79rVFsU5YgQJURfroV8JJqKPRRLRZTPM6m5BIJ+Ne80qm1ycw7r7aY0UCKKIqJcjnK5iJWMXC5PU76VYgT9xeFGtfVhUZnttBD19KJikVwuF7dlFL1rVJvCqDZ2jGozmiOjrQneUh6ZhVFtA5RyeXpb22krvA1RRHnrACoWsXwetQvKRbrLUxkwMalZFHuLgHYa1SYV6LepbEOUMQoSU0pGFJUo9/dR6u8jyhd2GtXW0teLiiXUVGBbSzs9A+Ud5+5Yucx/3+qhv2y0FCJKyo3TqLY2iDTmUW31upgrzymXy0SCSaVuovIA5ajAdrViBk1NTfT39/uotsr2s1R4ACSdDFxNPJx6hZktq/fc8Sg8zjk30aRdeDLXSWhmdwJ3pp3DOefc7uF9Ss455xLlhcc551yivPA455xLlBce55xzifLC45xzLlFeeJxzziXKC49zzrlEZe4E0tGQtBl4ZZxebh/gjXF6rd2pUXJC42T1nOOvUbI2Sk4Y36zvN7PUZjBu6MIzniQ9muaZvCPVKDmhcbJ6zvHXKFkbJSc0VtbheFebc865RHnhcc45lygvPO/4edoBRqhRckLjZPWc469RsjZKTmisrEPyYzzOOecS5Xs8zjnnEuWFxznnXKImfOGRdJKk5yS9IOmytPMMRdLLkp6U1CUpM1fAk7RC0iZJT1Ut20vS3ZL+FX7umWbGijpZl0raGNq1K1yMMFWS9pd0r6RnJK2XdFFYnql2HSJnFtt0kqRHJD0esn4/LM9am9bLmbk2HasJfYxHUo74UtsnAhuIL7V9upk9nWqwOiS9DBxhZpk64U3S8UA38BszOyws+yHwppn9IBT0Pc3s0jRzhly1si4Fus3sx2lmqyZpX2BfM1snqR14DDgVOJsMtesQOReSvTYV0GZm3ZIKwIPARcAXyVab1st5Ehlr07Ga6Hs8RwEvmNmLZtYP3AzMTzlTwzGz1cCbgxbPB24It28g/jJKXZ2smWNmr5nZunB7G/AMMJOMtesQOTPHYt3hbiH8M7LXpvVyvmdM9MIzE/hP1f0NZPSXJjDgLkmPSfp62mGGMcPMXoP4ywmYnnKe4Vwg6YnQFZeJbsEKSQcCc4A1ZLhdB+WEDLappJykLmATcLeZZbJN6+SEDLbpWEz0wqMay7L8l8UxZvZR4LPA+aHbyO2664APArOB14CfpBvnHZImA7cB3zCzt9LOU0+NnJlsUzMrmdlsYD/gKEmHpZ2pljo5M9mmYzHRC88GYP+q+/sBr6aUZVhm9mr4uQn4E3FXYVa9Hvr/K8cBNqWcpy4zez38opeBX5CRdg39+7cBN5rZ7WFx5tq1Vs6stmmFmW0B7iM+bpK5Nq2ozpn1Nh2NiV541gIHSZolqQlYBNyRcqaaJLWFg7dIagM+Azw19FqpugM4K9w+C/hzilmGVPnSCU4jA+0aDjD/EnjGzK6seihT7VovZ0bbdJqkqeF2C3AC8CzZa9OaObPYpmM1oUe1AYQhiVcDOWCFmS1LOVJNkj5AvJcDkAduykpWSb8H5hFP2/468D1gJXALcADwb2CBmaV+UL9O1nnE3RcGvAycU+nzT4ukY4EHgCeBclj8HeLjJ5lp1yFynk722vRw4sEDOeI/um8xsysk7U222rRezt+SsTYdqwlfeJxzziVrone1OeecS5gXHuecc4nywuOccy5RXnicc84lyguPc865RHnhcQ1H0gxJN0l6MUwf9LCk03bDds6W9L6q+9dLOmSE686TtFVSp+LZz1dLOqXq8XMlnTnM+h/ftXfgXDbl0w7g3GiEExZXAjeY2Rlh2fuBL+yGzZ1NfJJeZcaIxaNc/wEzOwVA0mxgpaQeM7vHzJYPs+484lm0/zHKbTqXeb7H4xrNp4D+6i9uM3vFzK6BHZMr/kjS2jCZ4jlh+TxJ90m6VdKzkm4MRQxJH5N0f9h7WiVpX0lfBo4AbgzXPmkJ6x8R1jlJ0jrF10y5Z7jQZtYFXAFcENZfKunicHuJpKdD3pvDZJvnAt8M2z5O0uclrQl7UH+TNKPqdVaEbC9KWlLZpqQzw2s+Hk4+rJwVf1ton7WSjtnF/w/nRs33eFyjORRYN8TjXwO2mtmRkpqBhyTdFR6bE9Z/FXgIOEbSGuAaYL6ZbZbUASwzs69KugC42MweBQh1CknTiOfKOt7MXpK01wizrwO+XWP5ZcAsM+uTNNXMtkhaTtW1VxTPRHy0mZmkxcAlwLfC+h8CPgm0A89Jug44GPgu8cSyb1Rl/ClwlZk9KOkAYBXw4RHmd25ceOFxDU3StcCxxHtBRxLPYXd42GMB2AM4COgHHjGzDWG9LuBAYAtwGHB3KCw54pl/h3I0sNrMXgIYxfQqtWZDB3iCeM9qJXE3Yi37AX8I83U1AS9VPfZXM+sD+iRtAmYQ7xneWrloYFXGE4BDKkUUmCKpPVxLx7lEeOFxjWY98KXKHTM7X9I+QOVS4AIuNLNV1StJmgf0VS0qEX/+Baw3s7mjyCDGdvmMOcQXShvsc8DxxMepLpd0aI3nXANcaWZ3hPeytOqxeu+rVsYImGtmPaNO79w48WM8rtH8HZgk6byqZa1Vt1cB5ymeqh9JByuezbue54BpkuaG5xeqvvi3EXdfDfYw8AlJs8I6w3a1hYkfLweuHbQ8AvY3s3uJu8+mApNrbHsPYGO4fRbDuwdYGCbArM54F+E4U1g+ewSv5dy48j0e11DCMY5TgaskXQJsBrYDl4anXE/chbYuDB7YzBCXMjaz/tAt9zNJexD/TlxNvGf1a2C5pB5gbtU6mxVfAfb2UDg2ASfWePnjJHUSF8ZNwBIzGzwQIQf8LmxbxMdftkj6C3CrpPnAhcR7OH+UtBH4JzBrmHZaL2kZcL+kEtBJPEpvCXCtpCfCe11NPJDBucT47NTOOecS5V1tzjnnEuWFxznnXKK88DjnnEuUFx7nnHOJ8sLjnHMuUV54nHPOJcoLj3POuUT9H4eUN/M8dxVEAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import glob\n",
    "path = \"/media/MassStorage/tmp/TE/honours/analysis/Alignment/*/*/match_distribution.csv\"\n",
    "path_names = glob.glob(path)\n",
    "for path in path_names:\n",
    "    template_frame = pd.DataFrame(data=None, columns = ['Species Name','Genetic Distance','Crossmapping Fraction'])\n",
    "    if path[71:-23] != 'unclassified':\n",
    "        if '20171212_FAH18688/barcode10' not in path and '20171207_FAH18654/barcode10' not in path:\n",
    "            match_dist = pd.read_csv(path, header=0)\n",
    "            for element in match_dist.index:\n",
    "#                 print(match_dist.at[element,'analysis/Consensus/'+path[53:-22]][19:-1])\n",
    "#                 print(match_dist.at[element,'Percentage Match'])\n",
    "                tmp = pd.DataFrame(data=[[phylogeny[match_dist.at[element,'analysis/Consensus/'+path[53:-22]][19:-1]],None,match_dist.at[element,'Percentage Match']]],columns=['Species Name','Genetic Distance','Crossmapping Fraction'])\n",
    "                template_frame = template_frame.append(tmp,ignore_index=True)\n",
    "            template_frame = template_frame.set_index('Species Name')\n",
    "            for element in template_frame.index:\n",
    "                if np.isnan(large_frame.at[element,phylogeny[path[53:-23]]]):\n",
    "                    template_frame.at[element,'Genetic Distance'] = 0\n",
    "                else:\n",
    "                    template_frame.at[element,'Genetic Distance'] = 100-large_frame.at[element,phylogeny[path[53:-23]]]\n",
    "            figures = sns.scatterplot(x=template_frame['Genetic Distance'],y=template_frame['Crossmapping Fraction'])\n",
    "            figures.set_title(\"Crossmapping Fraction vs Genetic Distance across Species\", fontsize=13)\n",
    "            ax = figures.get_figure()\n",
    "            ax.savefig(\"../../analysis/Stats/crossmapvsgendist.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "path = \"/media/MassStorage/tmp/TE/honours/analysis/Alignment/*/*/match_distribution.csv\"\n",
    "path_names = glob.glob(path)\n",
    "for path in path_names:\n",
    "    template_frame = pd.DataFrame(data=None, columns = ['Species Name','Genetic Distance','Crossmapping Fraction'])\n",
    "    if path[71:-23] != 'unclassified':\n",
    "        if '20171212_FAH18688/barcode10' not in path and '20171207_FAH18654/barcode10' not in path:\n",
    "            match_dist = pd.read_csv(path, header=0)\n",
    "            for element in match_dist.index:\n",
    "#                 print(match_dist.at[element,'analysis/Consensus/'+path[53:-22]][19:-1])\n",
    "#                 print(match_dist.at[element,'Percentage Match'])\n",
    "                tmp = pd.DataFrame(data=[[phylogeny[match_dist.at[element,'analysis/Consensus/'+path[53:-22]][19:-1]],None,match_dist.at[element,'Percentage Match']]],columns=['Species Name','Genetic Distance','Crossmapping Fraction'])\n",
    "                template_frame = template_frame.append(tmp,ignore_index=True)\n",
    "            template_frame = template_frame.set_index('Species Name')\n",
    "            for element in template_frame.index:\n",
    "                if np.isnan(large_frame.at[element,phylogeny[path[53:-23]]]):\n",
    "                    template_frame.at[element,'Genetic Distance'] = 0\n",
    "                else:\n",
    "                    template_frame.at[element,'Genetic Distance'] = 100-large_frame.at[element,phylogeny[path[53:-23]]]\n",
    "            plt.figure()\n",
    "            figures = sns.scatterplot(x=template_frame['Genetic Distance'],y=template_frame['Crossmapping Fraction'])\n",
    "            figures.set_title(\"Crossmapping Fraction vs Genetic Distance for \" + phylogeny[path[53:-23]], fontsize=15)\n",
    "            plt.show()\n",
    "            plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Draw line with nearest neighbour\n",
    "\n",
    "\n",
    "Draw in facet plot based on nearest neighbour?\n",
    "\n",
    "\n",
    "Color-code dots by genus?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting ../generate_summary_dataframe.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile ../generate_summary_dataframe.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "Enter the analysis directory to generate a dataframe that extracts the details\n",
    "for each sample\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_directory\", help=\"Run from honours folder and input analysis/\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "phylogeny = {'20171103_FAH15473/barcode01': 'k__Fungi;p__Basidiomycota;c__Pucciniomycetes;o__Pucciniales;f__Pucciniaceae;g__Puccinia;s__striiformis-tritici',\n",
    "             '20171103_FAH15473/barcode02': 'k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Capnodiales;f__Mycosphaerellaceae;g__Zymoseptoria;s__tritici',\n",
    "             '20171103_FAH15473/barcode03': 'k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Pleosporales;f__Pleosporaceae;g__Pyrenophora;s__tritici-repentis',\n",
    "             '20171103_FAH15473/barcode04': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Hypocreales;f__Nectriaceae;g__Fusarium;s__oxysporum',\n",
    "             '20171103_FAH15473/barcode05': 'k__Fungi;p__Ascomycota;c__Pezizomycetes;o__Pezizales;f__Tuberaceae;g__Tuber;s__brumale',\n",
    "             '20171103_FAH15473/barcode06': 'k__Fungi;p__Basidiomycota;c__Agaricomyectes;o__Agaricales;f__Cortinariaceae;g__Cortinarius;s__globuliformis',\n",
    "             '20171103_FAH15473/barcode07': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__niger',\n",
    "             '20171103_FAH15473/barcode08': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Metschnikowiaceae;g__Clavispora;s__lusitaniae',\n",
    "        '20171103_FAH15473/barcode09': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Kluyveromyces;s__unidentified', # formerly Cryptococcus neoformans, matched in both UNITE and tree to Kluyveromyces\n",
    "             '20171103_FAH15473/barcode10': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__chrysogenum',\n",
    "             '20171103_FAH15473/barcode11': 'k__Fungi;p__Basidiomycota;c__Microbotryomycetes;o__Sporidiobolales;f__Sporidiobolaceae;g__Rhodotorula;s__mucilaginosa',\n",
    "             '20171103_FAH15473/barcode12': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Microascales;f__Microascaceae;g__Scedosporium;s__boydii',\n",
    "             '20171207_FAH18654/barcode01': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Trichomonascaceae;g__Blastobotrys;s__proliferans', # matches Sugiyamella novakii in UNITE\n",
    "             '20171207_FAH18654/barcode02': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Debaryomyces;s__unidentified', # formerly Candida zeylanoides, matched in both UNITE and tree as Debaryomces/Meyerozyma\n",
    "        '20171207_FAH18654/barcode03': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Dipodascaceae;g__Galactomyces;s__geotrichum', # tree puts it further away from other samples of its type, UNITE calls it Dipodascus\n",
    "                                     # May have larger issues with Geotrichum candidum - maybe leave out?\n",
    "             '20171207_FAH18654/barcode04': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Metschnikowiaceae;g__Kodamaea;s__ohmeri',\n",
    "             '20171207_FAH18654/barcode05': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Meyerozyma;s__guilliermondii',\n",
    "             '20171207_FAH18654/barcode06': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Phaffomycetaceae;g__Wickerhamomyces;s__anomalus',\n",
    "             '20171207_FAH18654/barcode07': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Yamadazyma;s__mexicana',\n",
    "             '20171207_FAH18654/barcode08': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Yamadazyma;s__scolyti',\n",
    "             '20171207_FAH18654/barcode09': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Dipodascaceae;g__Yarrowia;s__lipolytica',\n",
    "             '20171207_FAH18654/barcode11': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Trichomonascaceae;g__Zygoascus;s__hellenicus', # matches Sugiyamella novakii in UNITE\n",
    "             '20171207_FAH18654/barcode12': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__flavus',\n",
    "             \n",
    "             '20171212_FAH18688/barcode01': 'k__Fungi;p__Basidiomycota;c__Tremellomycetes;o__Tremellales;f__Tremellaceae;g__Cryptococcus;s__zero',\n",
    "             '20171212_FAH18688/barcode02': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__unidentified',\n",
    "         '20171212_FAH18688/barcode03': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Diaporthaceae;g__Diaporthe;s__CCL067',\n",
    "             '20171212_FAH18688/barcode04': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Diaporthaceae;g__Diaporthe;s__unidentified',\n",
    "             '20171212_FAH18688/barcode05': 'k__Fungi;p__Ascomycota;c__Leotiomycetes;o__Helotiales;f__Dermateaceae;g__Oculimacula;s__yallundae-CCL031',\n",
    "             '20171212_FAH18688/barcode06': 'k__Fungi;p__Ascomycota;c__Leotiomycetes;o__Helotiales;f__Dermateaceae;g__Oculimacula;s__yallundae-CCL029',\n",
    "             '20171212_FAH18688/barcode07': 'k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Botryosphaeriales;f__Botryosphaeriaceae;g__Dothiorella;s__vidmadera',\n",
    "             '20171212_FAH18688/barcode08': 'k__Fungi;p__Basidiomycota;c__Exobasidiomycetes;o__Microstromatales;f__Quambalariaceae;g__Quambalaria;s__cyanescens',\n",
    "             '20171212_FAH18688/barcode09': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Xylariales;f__Xylariaceae;g__Entoleuca;s__unidentified',\n",
    "                 '20171212_FAH18688/barcode11': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Gnomoniaceae;g__Asteroma;s__CCL060', # may be asteroma or Diaporthe\n",
    "                 '20171212_FAH18688/barcode12': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Gnomoniaceae;g__Asteroma;s__CCL068',\n",
    "             '20180108_FAH18647/barcode01': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Saccharomyces;s__cerevisiae',\n",
    "             '20180108_FAH18647/barcode02': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Chaetothyriales;f__Herpotrichiellaceae;g__Cladophialophora;s__unidentified',\n",
    "             '20180108_FAH18647/barcode03': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__albicans',\n",
    "             '20180108_FAH18647/barcode04': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__metapsilosis',\n",
    "             '20180108_FAH18647/barcode05': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__orthopsilosis',\n",
    "             '20180108_FAH18647/barcode06': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__parapsilosis',\n",
    "#          '20180108_FAH18647/barcode07': 'k__Fungi;p__Basidiomycota;c__Tremellomycetes;o__Tremellales;f__Tremellaceae;g__Cryptococcus;s__gattii', # not including\n",
    "#              '20180108_FAH18647/barcode08': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Dipodascaceae;g__Geotrichum;s__candidum', # not including\n",
    "             '20180108_FAH18647/barcode09': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__unidentified', # formerly Kluyveromyces lactis\n",
    "             '20180108_FAH18647/barcode10': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Kluyveromyces;s__marxianus',\n",
    "             '20180108_FAH18647/barcode11': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Pichiaceae;g__Pichia;s__kudriavzevii',\n",
    "             '20180108_FAH18647/barcode12': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Pichiaceae;g__Pichia;s__membranifaciens'}\n",
    "\n",
    "summary = pd.DataFrame(data=None, columns = ['species','genus','family','order','class','phylum','kingdom','# raw reads','# reads after homology filtering','# reads after length filtering','path to raw reads','path to homology filtering','path to length filtering'])\n",
    "import glob\n",
    "path = \"/media/MassStorage/tmp/TE/honours/analysis/Consensus/*/*/\"\n",
    "path_names = glob.glob(path)\n",
    "for path in path_names:\n",
    "    if path[-13:-1] != 'unclassified' and path[-28:-1] != '20171207_FAH18654/barcode10' and path[-28:-1] != '20171212_FAH18688/barcode10' and path[-28:-1] != '20180108_FAH18647/barcode07' and path[-28:-1] != '20180108_FAH18647/barcode08':\n",
    "        key = path[-28:-1]\n",
    "        if args.verbose:\n",
    "            print('\\033[0;34m' + \"Opened barcode \" + '\\033[0;35m' + key + '\\033[1;37m')\n",
    "        species_ = phylogeny[key].split(';')[6].split('__')[1].lower()\n",
    "        genus_ = phylogeny[key].split(';')[5].split('__')[1].lower()\n",
    "        family_ = phylogeny[key].split(';')[4].split('__')[1].lower()\n",
    "        order_ = phylogeny[key].split(';')[3].split('__')[1].lower()\n",
    "        class_ = phylogeny[key].split(';')[2].split('__')[1].lower()\n",
    "        phylum_ = phylogeny[key].split(';')[1].split('__')[1].lower()\n",
    "        kingdom_ = phylogeny[key].split(';')[0].split('__')[1].lower()\n",
    "        \n",
    "        if args.verbose:\n",
    "            print('\\033[1;36m' + \"BEGIN RAW\"'\\033[1;37m')\n",
    "        raw_path = \"analysis/Concatenated/\"+key+\"/merged.fastq\"\n",
    "        raw_reads = SeqIO.to_dict(SeqIO.parse(raw_path, \"fastq\"))\n",
    "        raw_count = 0\n",
    "        for entry in raw_reads:\n",
    "            raw_count += 1\n",
    "            \n",
    "        if args.verbose:\n",
    "            print('\\033[1;36m' + \"BEGIN HOMOLOGY\"+ '\\033[1;37m')\n",
    "        homology_path = \"analysis/Python_Processing/\"+key+\"/combined_test.paf\"\n",
    "        homology_paf = pd.read_csv(homology_path, sep='\\t', header=None, engine='python')\n",
    "        homology_reads = {}\n",
    "        homology_count = 0\n",
    "        for entry in homology_paf[0].unique():\n",
    "            homology_count += 1\n",
    "            \n",
    "        if args.verbose:\n",
    "            print('\\033[1;36m' + \"BEGIN LENGTH\" + '\\033[1;37m')\n",
    "        length_path = \"analysis/Length_Filtered/\"+key+\"/length_restricted_reads.fasta\"\n",
    "        length_reads = SeqIO.to_dict(SeqIO.parse(length_path, \"fasta\"))\n",
    "        length_count = 0\n",
    "        for entry in length_reads:\n",
    "            length_count += 1\n",
    "            \n",
    "        if args.verbose:\n",
    "            print('\\033[1;36m' + \"BEGIN USE\" + '\\033[1;37m')\n",
    "        use_path = \"analysis/Length_Filtered/\"+key+\"/length_restricted_for_use.fasta\"\n",
    "        use_reads = SeqIO.to_dict(SeqIO.parse(use_path, \"fasta\"))\n",
    "        use_count = 0\n",
    "        for entry in use_reads:\n",
    "            use_count += 1\n",
    "\n",
    "\n",
    "        add = pd.DataFrame([[species_, genus_, family_, order_, class_, phylum_, kingdom_, raw_count, homology_count, length_count, use_count, raw_path, homology_path, length_path, use_path]], columns = ['species','genus','family','order','class','phylum','kingdom','# raw reads','# reads after homology filtering','# reads after length filtering','# for use', 'path to raw reads','path to homology filtering','path to length filtering', 'path for use'], index=[key])\n",
    "        summary = summary.append([add])\n",
    "summary = summary.sort_index(axis=0)\n",
    "summary = summary[['species','genus','family','order','class','phylum','kingdom','# raw reads','# reads after homology filtering','# reads after length filtering','# for use', 'path to raw reads','path to homology filtering','path to length filtering', 'path for use']]\n",
    "summary.to_csv(\"analysis/Stats/reference_dataframe.csv\")\n",
    "if args.verbose:\n",
    "            print('\\033[0;34m' + \"Reference Dataframe saved to \" + '\\033[0;35m' + \"analysis/Stats/reference_dataframe.csv\" + '\\033[1;37m')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k__Fungi;p__Basidiomycota;c__Pucciniomycetes;o__Pucciniales;f__Pucciniaceae;g__Puccinia;s__striiformis-tritici\n",
      "k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Capnodiales;f__Mycosphaerellaceae;g__Zymoseptoria;s__tritici\n",
      "k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Pleosporales;f__Pleosporaceae;g__Pyrenophora;s__tritici-repentis\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Hypocreales;f__Nectriaceae;g__Fusarium;s__oxysporum\n",
      "k__Fungi;p__Ascomycota;c__Pezizomycetes;o__Pezizales;f__Tuberaceae;g__Tuber;s__brumale\n",
      "k__Fungi;p__Basidiomycota;c__Agaricomycetes;o__Agaricales;f__Cortinariaceae;g__Cortinarius;s__globuliformis\n",
      "k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__niger\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Metschnikowiaceae;g__Clavispora;s__lusitaniae\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Kluyveromyces;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__chrysogenum\n",
      "k__Fungi;p__Basidiomycota;c__Microbotryomycetes;o__Sporidiobolales;f__Sporidiobolaceae;g__Rhodotorula;s__mucilaginosa\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Microascales;f__Microascaceae;g__Scedosporium;s__boydii\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Trichomonascaceae;g__Blastobotrys;s__proliferans\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Debaryomyces;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Dipodascaceae;g__Galactomyces;s__geotrichum\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Metschnikowiaceae;g__Kodamaea;s__ohmeri\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Meyerozyma;s__guilliermondii\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Phaffomycetaceae;g__Wickerhamomyces;s__anomalus\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Yamadazyma;s__mexicana\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Yamadazyma;s__scolyti\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Dipodascaceae;g__Yarrowia;s__lipolytica\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Trichomonascaceae;g__Zygoascus;s__hellenicus\n",
      "k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__flavus\n",
      "k__Fungi;p__Basidiomycota;c__Tremellomycetes;o__Tremellales;f__Tremellaceae;g__Cryptococcus;s__zero\n",
      "k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Diaporthaceae;g__Diaporthe;s__CCL067\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Diaporthaceae;g__Diaporthe;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Leotiomycetes;o__Helotiales;f__Dermateaceae;g__Oculimacula;s__yallundae-CCL031\n",
      "k__Fungi;p__Ascomycota;c__Leotiomycetes;o__Helotiales;f__Dermateaceae;g__Oculimacula;s__yallundae-CCL029\n",
      "k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Botryosphaeriales;f__Botryosphaeriaceae;g__Dothiorella;s__vidmadera\n",
      "k__Fungi;p__Basidiomycota;c__Exobasidiomycetes;o__Microstromatales;f__Quambalariaceae;g__Quambalaria;s__cyanescens\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Xylariales;f__Xylariaceae;g__Entoleuca;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Gnomoniaceae;g__Asteroma;s__CCL060\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Gnomoniaceae;g__Asteroma;s__CCL068\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Saccharomyces;s__cerevisiae\n",
      "k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Chaetothyriales;f__Herpotrichiellaceae;g__Cladophialophora;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__albicans\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__metapsilosis\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__orthopsilosis\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__parapsilosis\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Kluyveromyces;s__marxianus\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Pichiaceae;g__Pichia;s__kudriavzevii\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Pichiaceae;g__Pichia;s__membranifaciens\n"
     ]
    }
   ],
   "source": [
    "phylogeny = {'20171103_FAH15473/barcode01': 'k__Fungi;p__Basidiomycota;c__Pucciniomycetes;o__Pucciniales;f__Pucciniaceae;g__Puccinia;s__striiformis-tritici',\n",
    "             '20171103_FAH15473/barcode02': 'k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Capnodiales;f__Mycosphaerellaceae;g__Zymoseptoria;s__tritici',\n",
    "             '20171103_FAH15473/barcode03': 'k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Pleosporales;f__Pleosporaceae;g__Pyrenophora;s__tritici-repentis',\n",
    "             '20171103_FAH15473/barcode04': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Hypocreales;f__Nectriaceae;g__Fusarium;s__oxysporum',\n",
    "             '20171103_FAH15473/barcode05': 'k__Fungi;p__Ascomycota;c__Pezizomycetes;o__Pezizales;f__Tuberaceae;g__Tuber;s__brumale',\n",
    "             '20171103_FAH15473/barcode06': 'k__Fungi;p__Basidiomycota;c__Agaricomycetes;o__Agaricales;f__Cortinariaceae;g__Cortinarius;s__globuliformis',\n",
    "             '20171103_FAH15473/barcode07': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__niger',\n",
    "             '20171103_FAH15473/barcode08': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Metschnikowiaceae;g__Clavispora;s__lusitaniae',\n",
    "        '20171103_FAH15473/barcode09': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Kluyveromyces;s__unidentified', # formerly Cryptococcus neoformans, matched in both UNITE and tree to Kluyveromyces\n",
    "             '20171103_FAH15473/barcode10': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__chrysogenum',\n",
    "             '20171103_FAH15473/barcode11': 'k__Fungi;p__Basidiomycota;c__Microbotryomycetes;o__Sporidiobolales;f__Sporidiobolaceae;g__Rhodotorula;s__mucilaginosa',\n",
    "             '20171103_FAH15473/barcode12': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Microascales;f__Microascaceae;g__Scedosporium;s__boydii',\n",
    "             '20171207_FAH18654/barcode01': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Trichomonascaceae;g__Blastobotrys;s__proliferans', # matches Sugiyamella novakii in UNITE\n",
    "             '20171207_FAH18654/barcode02': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Debaryomyces;s__unidentified', # formerly Candida zeylanoides, matched in both UNITE and tree as Debaryomces/Meyerozyma\n",
    "        '20171207_FAH18654/barcode03': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Dipodascaceae;g__Galactomyces;s__geotrichum', # tree puts it further away from other samples of its type, UNITE calls it Dipodascus\n",
    "                                     # May have larger issues with Geotrichum candidum - maybe leave out?\n",
    "             '20171207_FAH18654/barcode04': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Metschnikowiaceae;g__Kodamaea;s__ohmeri',\n",
    "             '20171207_FAH18654/barcode05': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Meyerozyma;s__guilliermondii',\n",
    "             '20171207_FAH18654/barcode06': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Phaffomycetaceae;g__Wickerhamomyces;s__anomalus',\n",
    "             '20171207_FAH18654/barcode07': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Yamadazyma;s__mexicana',\n",
    "             '20171207_FAH18654/barcode08': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Yamadazyma;s__scolyti',\n",
    "             '20171207_FAH18654/barcode09': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Dipodascaceae;g__Yarrowia;s__lipolytica',\n",
    "             '20171207_FAH18654/barcode11': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Trichomonascaceae;g__Zygoascus;s__hellenicus', # matches Sugiyamella novakii in UNITE\n",
    "             '20171207_FAH18654/barcode12': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__flavus',\n",
    "             \n",
    "             '20171212_FAH18688/barcode01': 'k__Fungi;p__Basidiomycota;c__Tremellomycetes;o__Tremellales;f__Tremellaceae;g__Cryptococcus;s__zero',\n",
    "             '20171212_FAH18688/barcode02': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__unidentified',\n",
    "         '20171212_FAH18688/barcode03': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Diaporthaceae;g__Diaporthe;s__CCL067',\n",
    "             '20171212_FAH18688/barcode04': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Diaporthaceae;g__Diaporthe;s__unidentified',\n",
    "             '20171212_FAH18688/barcode05': 'k__Fungi;p__Ascomycota;c__Leotiomycetes;o__Helotiales;f__Dermateaceae;g__Oculimacula;s__yallundae-CCL031',\n",
    "             '20171212_FAH18688/barcode06': 'k__Fungi;p__Ascomycota;c__Leotiomycetes;o__Helotiales;f__Dermateaceae;g__Oculimacula;s__yallundae-CCL029',\n",
    "             '20171212_FAH18688/barcode07': 'k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Botryosphaeriales;f__Botryosphaeriaceae;g__Dothiorella;s__vidmadera',\n",
    "             '20171212_FAH18688/barcode08': 'k__Fungi;p__Basidiomycota;c__Exobasidiomycetes;o__Microstromatales;f__Quambalariaceae;g__Quambalaria;s__cyanescens',\n",
    "             '20171212_FAH18688/barcode09': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Xylariales;f__Xylariaceae;g__Entoleuca;s__unidentified',\n",
    "                 '20171212_FAH18688/barcode11': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Gnomoniaceae;g__Asteroma;s__CCL060', # may be asteroma or Diaporthe\n",
    "                 '20171212_FAH18688/barcode12': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Gnomoniaceae;g__Asteroma;s__CCL068',\n",
    "             '20180108_FAH18647/barcode01': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Saccharomyces;s__cerevisiae',\n",
    "             '20180108_FAH18647/barcode02': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Chaetothyriales;f__Herpotrichiellaceae;g__Cladophialophora;s__unidentified',\n",
    "             '20180108_FAH18647/barcode03': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__albicans',\n",
    "             '20180108_FAH18647/barcode04': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__metapsilosis',\n",
    "             '20180108_FAH18647/barcode05': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__orthopsilosis',\n",
    "             '20180108_FAH18647/barcode06': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__parapsilosis',\n",
    "#          '20180108_FAH18647/barcode07': 'k__Fungi;p__Basidiomycota;c__Tremellomycetes;o__Tremellales;f__Tremellaceae;g__Cryptococcus;s__gattii', # not including\n",
    "#              '20180108_FAH18647/barcode08': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Dipodascaceae;g__Geotrichum;s__candidum', # not including\n",
    "             '20180108_FAH18647/barcode09': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Candida;s__unidentified', # formerly Kluyveromyces lactis\n",
    "             '20180108_FAH18647/barcode10': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Kluyveromyces;s__marxianus',\n",
    "             '20180108_FAH18647/barcode11': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Pichiaceae;g__Pichia;s__kudriavzevii',\n",
    "             '20180108_FAH18647/barcode12': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Pichiaceae;g__Pichia;s__membranifaciens'}\n",
    "with open('../../analysis/Stats/taxonomy_file.csv', 'w+') as f:\n",
    "    for key in phylogeny.keys():\n",
    "        print(phylogeny[key])\n",
    "        f.write(\"%s,%s\\n\"%(key,phylogeny[key]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k__Fungi;p__Basidiomycota;c__Pucciniomycetes;o__Pucciniales;f__Pucciniaceae;g__Puccinia;s__Puccinia_striiformis\n",
      "k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Capnodiales;f__Mycosphaerellaceae;g__Zymoseptoria;s__Zymoseptoria_tritici\n",
      "k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Pleosporales;f__Pleosporaceae;g__Pyrenophora;s__Pyrenophora_tritici-repentis\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Hypocreales;f__Nectriaceae;g__Fusarium;s__Fusarium_oxysporum\n",
      "k__Fungi;p__Ascomycota;c__Pezizomycetes;o__Pezizales;f__Tuberaceae;g__Tuber;s__Tuber_brumale\n",
      "k__Fungi;p__Basidiomycota;c__Agaricomyctes;o__Agaricales;f__Cortinariaceae;g__Cortinarius;s__Cortinarius_globuliformis\n",
      "k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__Aspergillus_niger\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Metschnikowiaceae;g__Clavispora;s__Clavispora_lusitaniae\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Kluyveromyces;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum\n",
      "k__Fungi;p__Basidiomycota;c__Microbotryomycetes;o__Sporidiobolales;f__Sporidiobolaceae;g__Rhodotorula;s__Rhodotorula_mucilaginosa\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Microascales;f__Microascaceae;g__Scedosporium;s__Scedosporium_boydii\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Trichomonascaceae;g__Blastobotrys;s__Blastobotrys_proliferans\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Debaryomyces;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Dipodascaceae;g__Galactomyces;s__Galactomyces_geotrichum\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Metschnikowiaceae;g__Kodamaea;s__Kodamaea_ohmeri\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Meyerozyma;s__Meyerozyma_guilliermondii\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Phaffomycetaceae;g__Wickerhamomyces;s__Wickerhamomyces_anomalus\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Yamadazyma;s__Yamadazyma_mexicana\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Yamadazyma;s__Yamadazyma_scolyti\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Dipodascaceae;g__Yarrowia;s__Yarrowia_lipolytica\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Trichomonascaceae;g__Zygoascus;s__Zygoascus_hellenicus\n",
      "k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__Aspergillus_flavus\n",
      "k__Fungi;p__Basidiomycota;c__Tremellomycetes;o__Tremellales;f__Tremellaceae;g__Cryptococcus;s__Cryptococcus_zero\n",
      "k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Diaporthaceae;g__Diaporthe;s__Diaporthe_CCL067\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Diaporthaceae;g__Diaporthe;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Leotiomycetes;o__Helotiales;f__Dermateaceae;g__Oculimacula;s__Oculimacula_yallundae\n",
      "k__Fungi;p__Ascomycota;c__Leotiomycetes;o__Helotiales;f__Dermateaceae;g__Oculimacula;s__Oculimacula_yallundae\n",
      "k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Botryosphaeriales;f__Botryosphaeriaceae;g__Dothiorella;s__Dothiorella_vidmadera\n",
      "k__Fungi;p__Basidiomycota;c__Exobasidiomycetes;o__Microstromatales;f__Quambalariaceae;g__Quambalaria;s__Quambalaria_cyanescens\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Xylariales;f__Xylariaceae;g__Entoleuca;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Gnomoniaceae;g__Asteroma;s__Asteroma_CCL060\n",
      "k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Gnomoniaceae;g__Asteroma;s__Asteroma_CCL068\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Saccharomyces;s__Saccharomyces_cerevisiae\n",
      "k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Chaetothyriales;f__Herpotrichiellaceae;g__Cladophialophora;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetales_fam_Incertae_sedis;g__Candida;s__Candida_albicans\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetales_fam_Incertae_sedis;g__Candida;s__Candida_metapsilosis\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetales_fam_Incertae_sedis;g__Candida;s__Candida_orthopsilosis\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetales_fam_Incertae_sedis;g__Candida;s__Candida_parapsilosis\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetales_fam_Incertae_sedis;g__Candida;s__unidentified\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Kluyveromyces;s__Kluyveromyces_marxianus\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Pichiaceae;g__Pichia;s__Pichia_kudriavzevii\n",
      "k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Pichiaceae;g__Pichia;s__Pichia_membranifaciens\n"
     ]
    }
   ],
   "source": [
    "phylogeny = {'20171103_FAH15473/barcode01': 'k__Fungi;p__Basidiomycota;c__Pucciniomycetes;o__Pucciniales;f__Pucciniaceae;g__Puccinia;s__Puccinia_striiformis',\n",
    "             '20171103_FAH15473/barcode02': 'k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Capnodiales;f__Mycosphaerellaceae;g__Zymoseptoria;s__Zymoseptoria_tritici',\n",
    "             '20171103_FAH15473/barcode03': 'k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Pleosporales;f__Pleosporaceae;g__Pyrenophora;s__Pyrenophora_tritici-repentis',\n",
    "             '20171103_FAH15473/barcode04': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Hypocreales;f__Nectriaceae;g__Fusarium;s__Fusarium_oxysporum',\n",
    "             '20171103_FAH15473/barcode05': 'k__Fungi;p__Ascomycota;c__Pezizomycetes;o__Pezizales;f__Tuberaceae;g__Tuber;s__Tuber_brumale',\n",
    "             '20171103_FAH15473/barcode06': 'k__Fungi;p__Basidiomycota;c__Agaricomycetes;o__Agaricales;f__Cortinariaceae;g__Cortinarius;s__Cortinarius_globuliformis',\n",
    "             '20171103_FAH15473/barcode07': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__Aspergillus_niger',\n",
    "             '20171103_FAH15473/barcode08': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Metschnikowiaceae;g__Clavispora;s__Clavispora_lusitaniae',\n",
    "        '20171103_FAH15473/barcode09': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Kluyveromyces;s__unidentified', # formerly Cryptococcus neoformans, matched in both UNITE and tree to Kluyveromyces\n",
    "             '20171103_FAH15473/barcode10': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum',\n",
    "             '20171103_FAH15473/barcode11': 'k__Fungi;p__Basidiomycota;c__Microbotryomycetes;o__Sporidiobolales;f__Sporidiobolaceae;g__Rhodotorula;s__Rhodotorula_mucilaginosa',\n",
    "             '20171103_FAH15473/barcode12': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Microascales;f__Microascaceae;g__Scedosporium;s__Scedosporium_boydii',\n",
    "             '20171207_FAH18654/barcode01': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Trichomonascaceae;g__Blastobotrys;s__Blastobotrys_proliferans', # matches Sugiyamella novakii in UNITE\n",
    "             '20171207_FAH18654/barcode02': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Debaryomyces;s__unidentified', # formerly Candida zeylanoides, matched in both UNITE and tree as Debaryomces/Meyerozyma\n",
    "        '20171207_FAH18654/barcode03': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Dipodascaceae;g__Galactomyces;s__Galactomyces_geotrichum', # tree puts it further away from other samples of its type, UNITE calls it Dipodascus\n",
    "                                     # May have larger issues with Geotrichum candidum - maybe leave out?\n",
    "             '20171207_FAH18654/barcode04': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Metschnikowiaceae;g__Kodamaea;s__Kodamaea_ohmeri',\n",
    "             '20171207_FAH18654/barcode05': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Meyerozyma;s__Meyerozyma_guilliermondii',\n",
    "             '20171207_FAH18654/barcode06': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Phaffomycetaceae;g__Wickerhamomyces;s__Wickerhamomyces_anomalus',\n",
    "             '20171207_FAH18654/barcode07': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Yamadazyma;s__Yamadazyma_mexicana',\n",
    "             '20171207_FAH18654/barcode08': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Debaryomycetaceae;g__Yamadazyma;s__Yamadazyma_scolyti',\n",
    "             '20171207_FAH18654/barcode09': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Dipodascaceae;g__Yarrowia;s__Yarrowia_lipolytica',\n",
    "             '20171207_FAH18654/barcode11': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Trichomonascaceae;g__Zygoascus;s__Zygoascus_hellenicus', # matches Sugiyamella novakii in UNITE\n",
    "             '20171207_FAH18654/barcode12': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__Aspergillus_flavus',\n",
    "             \n",
    "             '20171212_FAH18688/barcode01': 'k__Fungi;p__Basidiomycota;c__Tremellomycetes;o__Tremellales;f__Tremellaceae;g__Cryptococcus;s__Cryptococcus_zero',\n",
    "             '20171212_FAH18688/barcode02': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Aspergillus;s__unidentified',\n",
    "         '20171212_FAH18688/barcode03': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Diaporthaceae;g__Diaporthe;s__Diaporthe_CCL067',\n",
    "             '20171212_FAH18688/barcode04': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Diaporthaceae;g__Diaporthe;s__unidentified',\n",
    "             '20171212_FAH18688/barcode05': 'k__Fungi;p__Ascomycota;c__Leotiomycetes;o__Helotiales;f__Dermateaceae;g__Oculimacula;s__Oculimacula_yallundae_CCL031',\n",
    "             '20171212_FAH18688/barcode06': 'k__Fungi;p__Ascomycota;c__Leotiomycetes;o__Helotiales;f__Dermateaceae;g__Oculimacula;s__Oculimacula_yallundae_CCL029',\n",
    "             '20171212_FAH18688/barcode07': 'k__Fungi;p__Ascomycota;c__Dothideomycetes;o__Botryosphaeriales;f__Botryosphaeriaceae;g__Dothiorella;s__Dothiorella_vidmadera',\n",
    "             '20171212_FAH18688/barcode08': 'k__Fungi;p__Basidiomycota;c__Exobasidiomycetes;o__Microstromatales;f__Quambalariaceae;g__Quambalaria;s__Quambalaria_cyanescens',\n",
    "             '20171212_FAH18688/barcode09': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Xylariales;f__Xylariaceae;g__Entoleuca;s__unidentified',\n",
    "                 '20171212_FAH18688/barcode11': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Gnomoniaceae;g__Asteroma;s__Asteroma_CCL060', # may be asteroma or Diaporthe\n",
    "                 '20171212_FAH18688/barcode12': 'k__Fungi;p__Ascomycota;c__Sordariomycetes;o__Diaporthales;f__Gnomoniaceae;g__Asteroma;s__Asteroma_CCL068',\n",
    "             '20180108_FAH18647/barcode01': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Saccharomyces;s__Saccharomyces_cerevisiae',\n",
    "             '20180108_FAH18647/barcode02': 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Chaetothyriales;f__Herpotrichiellaceae;g__Cladophialophora;s__unidentified',\n",
    "             '20180108_FAH18647/barcode03': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetales_fam_Incertae_sedis;g__Candida;s__Candida_albicans',\n",
    "             '20180108_FAH18647/barcode04': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetales_fam_Incertae_sedis;g__Candida;s__Candida_metapsilosis',\n",
    "             '20180108_FAH18647/barcode05': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetales_fam_Incertae_sedis;g__Candida;s__Candida_orthopsilosis',\n",
    "             '20180108_FAH18647/barcode06': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetales_fam_Incertae_sedis;g__Candida;s__Candida_parapsilosis',\n",
    "#          '20180108_FAH18647/barcode07': 'k__Fungi;p__Basidiomycota;c__Tremellomycetes;o__Tremellales;f__Tremellaceae;g__Cryptococcus;s__gattii', # not including\n",
    "#              '20180108_FAH18647/barcode08': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Dipodascaceae;g__Geotrichum;s__candidum', # not including\n",
    "             '20180108_FAH18647/barcode09': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetales_fam_Incertae_sedis;g__Candida;s__unidentified', # formerly Kluyveromyces lactis\n",
    "             '20180108_FAH18647/barcode10': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Saccharomycetaceae;g__Kluyveromyces;s__Kluyveromyces_marxianus',\n",
    "             '20180108_FAH18647/barcode11': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Pichiaceae;g__Pichia;s__Pichia_kudriavzevii',\n",
    "             '20180108_FAH18647/barcode12': 'k__Fungi;p__Ascomycota;c__Saccharomycetes;o__Saccharomycetales;f__Pichiaceae;g__Pichia;s__Pichia_membranifaciens'}\n",
    "with open('../../analysis/Stats/taxonomy_file_qiime.csv', 'w+') as f:\n",
    "    for key in phylogeny.keys():\n",
    "        print(phylogeny[key])\n",
    "        f.write(\"%s\\t%s\\n\"%(key,phylogeny[key]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%writefile ../numberfy_fasta.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "This creates a machine-learning friendly version of each length restricted fasta file\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "if args.verbose:\n",
    "    print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "fasta = SeqIO.to_dict(SeqIO.parse(args.input_file, \"fasta\"))\n",
    "\n",
    "fasta_numbers = fasta.copy()\n",
    "\n",
    "for key in fasta:\n",
    "    seq = str(fasta[key].seq[30:-30]).replace(\"A\",'0').replace(\"C\",'1').replace(\"G\",'2').replace(\"T\",'3')\n",
    "#     if len(seq) < max(total_lens):\n",
    "#         seq = seq + '4'*(max(total_lens)-len(seq))\n",
    "    fasta_numbers[key].seq = Seq(seq)\n",
    "\n",
    "with open(args.input_file[:-6]+'_numbers.fasta', \"w\") as output_handle:\n",
    "    SeqIO.write(fasta_numbers.values(), output_handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# WILL NEED TO APPLY FILTERING TO BADREAD GENERATED READS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import argparse\n",
    "\n",
    "# Load the full file containing all reads for this barcode\n",
    "full_file_dict = SeqIO.to_dict(SeqIO.parse('../../analysis/Length_Filtered/20171207_FAH18654/barcode03/length_restricted_extended.fastq', \"fastq\"))\n",
    "\n",
    "# Extract the information about the lengths of the sequence for each read in this barcode\n",
    "full_lengths = []\n",
    "for key in full_file_dict:\n",
    "    if 1500 < len(full_file_dict[key].seq) < 5000:\n",
    "        full_lengths.append(len(full_file_dict[key].seq))\n",
    "full_lengths_len = len(full_lengths)\n",
    "\n",
    "ax = sns.distplot(full_lengths, color=\"k\", kde=False, bins=5000)\n",
    "ax.set(xlim=(250, 3500))\n",
    "ax.set_title(\"Read spread for extended\", fontsize=15)\n",
    "ax.set_xlabel(\"Length of read\", fontsize=13)\n",
    "ax.set_ylabel(\"Number of reads\", fontsize=13)\n",
    "figure1 = ax.get_figure()\n",
    "display(figure1)\n",
    "figure1.clf()\n",
    "\n",
    "mean = np.mean(full_lengths)\n",
    "std = np.std(full_lengths)\n",
    "\n",
    "print(mean)\n",
    "print(std)\n",
    "    \n",
    "print(full_lengths_len)\n",
    "    \n",
    "length_filt_dict = full_file_dict.copy()\n",
    "for key in full_file_dict:\n",
    "    if len(full_file_dict[key].seq) < (mean-1.645*std) or len(full_file_dict[key].seq) > (mean+1.645*std):\n",
    "        del length_filt_dict[key]\n",
    "print(len(length_filt_dict))\n",
    "\n",
    "        \n",
    "        \n",
    "SeqIO.write(length_filt_dict.values(), '../../analysis/Length_Filtered/20171207_FAH18654/barcode03/length_restricted_extended.fasta', \"fasta\")\n",
    "# if args.verbose:\n",
    "#     print('\\033[1;36m' + 'Saved %s' % ('/'.join([output_folder, 'length_restricted_reads.fasta'])) + '\\033[1;37m')     \n",
    "    \n",
    "    \n",
    "    \n",
    "# length_filt_lens = []\n",
    "# len_filt_keys = []\n",
    "# for key in length_filt_dict:\n",
    "#     length_filt_lens.append(len(length_filt_dict[key].seq))\n",
    "#     len_filt_keys.append(key)\n",
    "\n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "    \n",
    "# Extract the qscores\n",
    "# if args.verbose:\n",
    "#     print('\\033[0;34m' + \"Loading \" + 'Basecalled/'+'/'.join(args.full_file.rsplit('/')[-3:-1])+'/sequencing_summary.txt' + '\\033[1;37m')\n",
    "# summ_stats_csv = pd.read_csv('Basecalled/'+'/'.join(args.full_file.rsplit('/')[-3:-1])+'/sequencing_summary.txt', sep='\\t', header=None, names=['filename', 'read_id', 'run_id', 'batch_id', 'channel', 'mux', 'start_time', 'duration', 'num_events', 'passes_filtering', 'template_start', 'num_events_template', 'template_duration', 'sequence_length_template', 'mean_qscore_template', 'strand_score_template', 'median_template', 'mad_template'], engine='python')\n",
    "# summ_stats_csv = pd.DataFrame(summ_stats_csv[1:])\n",
    "# summary_list = []\n",
    "# for column, row in summ_stats_csv.iterrows():\n",
    "#     if row['read_id'] in full_keys:\n",
    "#         summary_list.append([row['read_id'], row['mean_qscore_template']])\n",
    "# summary_frame = pd.DataFrame(summary_list)\n",
    "# if args.verbose:\n",
    "#     print('\\033[0;34m' + \"Finished with \" + 'Basecalled/'+'/'.join(args.full_file.rsplit('/')[-3:-1])+'/sequencing_summary.txt' + '\\033[1;37m')\n",
    "    \n",
    "# Create a dictionary containing the statistics for the filtered dataset\n",
    "#     Total no. frDNA reads, Min. read length, Max. read length, Mean read length, Median read length, Quality score\n",
    "\n",
    "# stats_dict = {'number of frDNA reads':len(length_filt_lens),'minimum read length':min(length_filt_lens),'maximum read length':max(length_filt_lens),'mean read length':\"{:.0f}\".format(np.mean(length_filt_lens)),'std dev':\"{:.0f}\".format(np.std(length_filt_lens)),'median read length':\"{:.0f}\".format(np.median(length_filt_lens))\n",
    "#               ,'min_qscore':\"{:.2f}\".format(min(summary_frame[1].astype(float))), 'max_qscore':\"{:.2f}\".format(max(summary_frame[1].astype(float))), 'mean_qscore':\"{:.2f}\".format(np.mean(summary_frame[1].astype(float))), 'median_qscore':\"{:.2f}\".format(np.median(summary_frame[1].astype(float)))\n",
    "#              }\n",
    "# stats = pd.DataFrame(stats_dict, index=['%s' % '/'.join(args.full_file.rsplit('/')[-3:-1])])    \n",
    "              \n",
    "# bx = sns.distplot(length_filt_lens, color=\"k\", kde=False)\n",
    "# bx.set(xlim=(250, 3500))\n",
    "# bx.set_title(\"frDNA reads for %s\" % '/'.join(args.full_file.rsplit('/')[-3:-1]), fontsize=15)\n",
    "# bx.set_xlabel(\"Length of read\", fontsize=13)\n",
    "# bx.set_ylabel(\"Number of reads\", fontsize=13)\n",
    "# figure2 = bx.get_figure()\n",
    "# figure2.savefig('/'.join([output_folder, 'frDNA_len_filt_full.png']))\n",
    "# figure2.clf()\n",
    "# if args.verbose:\n",
    "#     print('\\033[0;32m' + \"frDNA spread image file saved to \" + '/'.join([output_folder, 'frDNA_len_filt_full.png']) + '\\033[1;37m')\n",
    "\n",
    "# cx = sns.distplot(length_filt_lens, color=\"k\", kde=False)\n",
    "# cx.set(xlim=((mean-1.645*std)-100, (mean+1.645*std)+100))\n",
    "# cx.set_title(\"frDNA reads for %s\" % '/'.join(args.full_file.rsplit('/')[-3:-1]), fontsize=15)\n",
    "# cx.set_xlabel(\"Length of read\", fontsize=13)\n",
    "# cx.set_ylabel(\"Number of reads\", fontsize=13)\n",
    "# figure3 = cx.get_figure()\n",
    "# figure3.savefig('/'.join([output_folder, 'frDNA_len_filt_limited.png']))\n",
    "# figure3.clf()\n",
    "# if args.verbose:\n",
    "#     print('\\033[0;32m' + \"Zoomed-in frDNA spread image file saved to \" + '/'.join([output_folder, 'frDNA_len_filt_limited.png']) + '\\033[1;37m')\n",
    "\n",
    "# stats.to_csv('/'.join([output_folder, 'frDNA_len_filt_statistics.csv']), index=False)\n",
    "# if args.verbose:\n",
    "#     print('\\033[0;32m' + \"Summary statistics file saved to \" + '/'.join([output_folder, 'frDNA_len_filt_statistics.csv']) + '\\033[1;37m')\n",
    "    \n",
    "# print('\\033[0;35m'+'END'+'\\033[1;37m')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_dict = {\n",
    "    1: 'kingdom fungi',\n",
    "              2: 'phylum ascomycota',\n",
    "              3: 'class eurotiomycetes',\n",
    "              4: 'family aspergillaceae',\n",
    "              5: 'genus aspergillus',\n",
    "              6: 'class dothideomycetes',\n",
    "              7: 'genus oculimacula',\n",
    "              8: 'order saccharomycetales',\n",
    "              9: 'family debaryomycetaceae',\n",
    "              10: 'genus yamadazyma',\n",
    "              11: 'family dipodascaceae',\n",
    "              12: 'family metschnikowiaceae',\n",
    "              13: 'genus pichia',\n",
    "              14: 'family saccharomycetaceae',\n",
    "              15: 'genus candida',\n",
    "              16: 'genus kluyveromyces',\n",
    "              17: 'family trichomonascaceae',\n",
    "              18: 'class sordariomycetes',\n",
    "              19: 'order diaporthales',\n",
    "              20: 'genus diaporthe',\n",
    "              21: 'genus asteroma',\n",
    "              22: 'phylum basidiomycota'\n",
    "             }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "kingdom fungi\n",
      "phylum ascomycota\n",
      "class eurotiomycetes\n",
      "family aspergillaceae\n",
      "genus aspergillus\n",
      "class dothideomycetes\n",
      "genus oculimacula\n",
      "order saccharomycetales\n",
      "family debaryomycetaceae\n",
      "genus yamadazyma\n",
      "family dipodascaceae\n",
      "family metschnikowiaceae\n",
      "genus pichia\n",
      "family saccharomycetaceae\n",
      "genus candida\n",
      "genus kluyveromyces\n",
      "family trichomonascaceae\n",
      "class sordariomycetes\n",
      "order diaporthales\n",
      "genus diaporthe\n",
      "genus asteroma\n",
      "phylum basidiomycota\n"
     ]
    }
   ],
   "source": [
    "for key in nodes_dict:\n",
    "    print(nodes_dict[key].split(' ')[0], nodes_dict[key].split(' ')[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "kingdom fungi\n",
      "phylum ascomycota\n",
      "class eurotiomycetes\n",
      "family aspergillaceae\n",
      "genus aspergillus\n",
      "class dothideomycetes\n",
      "genus oculimacula\n",
      "order saccharomycetales\n",
      "family debaryomycetaceae\n",
      "genus yamadazyma\n",
      "family dipodascaceae\n",
      "family metschnikowiaceae\n",
      "genus pichia\n",
      "family saccharomycetaceae\n",
      "genus candida\n",
      "genus kluyveromyces\n",
      "family trichomonascaceae\n",
      "class sordariomycetes\n",
      "order diaporthales\n",
      "genus diaporthe\n",
      "genus asteroma\n",
      "phylum basidiomycota\n"
     ]
    }
   ],
   "source": [
    "with open('../../analysis/Stats/nodes.csv', 'w+') as f:\n",
    "    for key in nodes_dict.keys():\n",
    "        print(nodes_dict[key])\n",
    "        f.write(\"%s\\n\"%(nodes_dict[key]))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "ref_df = pd.read_csv('../../analysis/Stats/reference_dataframe.csv', index_col=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "analysis/Length_Filtered/20171103_FAH15473/barcode01/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171103_FAH15473/barcode02/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171103_FAH15473/barcode03/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171103_FAH15473/barcode04/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171103_FAH15473/barcode05/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171103_FAH15473/barcode06/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171103_FAH15473/barcode07/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171103_FAH15473/barcode08/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171103_FAH15473/barcode09/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171103_FAH15473/barcode10/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171103_FAH15473/barcode11/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171103_FAH15473/barcode12/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171207_FAH18654/barcode01/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171207_FAH18654/barcode02/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171207_FAH18654/barcode03/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171207_FAH18654/barcode04/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171207_FAH18654/barcode05/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171207_FAH18654/barcode06/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171207_FAH18654/barcode07/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171207_FAH18654/barcode08/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171207_FAH18654/barcode09/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171207_FAH18654/barcode11/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171207_FAH18654/barcode12/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171212_FAH18688/barcode01/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171212_FAH18688/barcode02/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171212_FAH18688/barcode03/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171212_FAH18688/barcode04/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171212_FAH18688/barcode05/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171212_FAH18688/barcode06/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171212_FAH18688/barcode07/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171212_FAH18688/barcode08/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171212_FAH18688/barcode09/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171212_FAH18688/barcode11/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20171212_FAH18688/barcode12/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20180108_FAH18647/barcode01/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20180108_FAH18647/barcode02/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20180108_FAH18647/barcode03/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20180108_FAH18647/barcode04/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20180108_FAH18647/barcode05/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20180108_FAH18647/barcode06/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20180108_FAH18647/barcode09/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20180108_FAH18647/barcode10/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20180108_FAH18647/barcode11/length_restricted_for_use.fasta\n",
      "\n",
      "\n",
      "analysis/Length_Filtered/20180108_FAH18647/barcode12/length_restricted_for_use.fasta\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for index, row in ref_df.iterrows():\n",
    "    print(row['path for use'])\n",
    "    print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "dicta = {'a': 'a', 'b': 'b', 'c': 'c'}\n",
    "\n",
    "with open('../../test.txt', 'w+') as handle:\n",
    "    handle.writelines(\"%s\\n\" % entry for entry in dicta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9420492082253151"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean([0.916452442159383,0.9244889445139758,0.9760975609756097,0.9333333333333333,0.9598737601442742])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9408390267102386"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean([0.943954104148279,0.9613882863340564,0.9083068542896051,0.957592339261286,0.9329535495179667])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9410179818515244"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean([0.93,0.942577626541897,0.9409828356454268,0.9453072248480757,0.9462222222222223])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2135\n",
      "3560\n",
      "2982.837424733673\n",
      "3008.0\n",
      "125.65695734487451\n",
      "21590\n"
     ]
    }
   ],
   "source": [
    "fasta_dict = SeqIO.to_dict(SeqIO.parse(\"../../analysis/Length_Filtered/wheat/barcode01/length_restricted_reads.fasta\", \"fasta\"))\n",
    "lengthe = []\n",
    "for key in fasta_dict:\n",
    "    lengthe.append(len(fasta_dict[key].seq))\n",
    "print(min(lengthe))\n",
    "print(max(lengthe))\n",
    "print(np.mean(lengthe))\n",
    "print(np.median(lengthe))\n",
    "print(np.std(lengthe))\n",
    "print(len(lengthe))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### What I want to do\n",
    "\n",
    "For barcode in the wheat samples, save into a dataframe: <br> <br>\n",
    "    - suspected species 1, suspected species 2, # reads in merged.fastq, # reads after homology filtering, # reads after length filtering, # reads for use, path to raw, path to homology, path to length, path for use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting ../generate_wheat_summary_dataframe.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile ../generate_wheat_summary_dataframe.py\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord as SR\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas import DataFrame as df\n",
    "# import seaborn as sns\n",
    "import csv\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import os\n",
    "from shutil import copy\n",
    "import random\n",
    "import warnings\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "# from ipysankeywidget import SankeyWidget\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import argparse\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"\"\"\n",
    "Enter the analysis directory to generate a dataframe that extracts the details\n",
    "for each sample\n",
    "\"\"\")\n",
    "group = parser.add_mutually_exclusive_group()\n",
    "group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "parser.add_argument(\"input_directory\", help=\"Run from honours folder and input analysis/\")\n",
    "args = parser.parse_args()\n",
    "\n",
    "suspected = {'barcode01': 's__Puccinia_striiformis',\n",
    "             'barcode02': 's__Zymoseptoria_tritici',\n",
    "             'barcode03': 's__unknown', # healthy wheat\n",
    "             'barcode04': 's__Puccinia_striiformis,s__Zymoseptoria_tritici', # both Pst + Zymo\n",
    "             'barcode05': 's__Pyrenophora_tritici-repentis',\n",
    "             'barcode06': 's__unknown', # healthy, resistant wheat, 2nd year,\n",
    "             'barcode07': 's__unknown', # healthy, susceptible wheat, 2nd year\n",
    "             'barcode08': 's__Puccinia_striiformis',\n",
    "             'barcode09': 's__Puccinia_striiformis,s__Pyrenophora_tritici-repentis', # both Pst + Pyr\n",
    "             'barcode10': 's__Puccinia_striiformis,s__Zymoseptoria_tritici', # both Pst + Zymo\n",
    "             'barcode11': 's__Zymoseptoria_tritici,s__Puccinia_striiformis', # Zymo + Pst?\n",
    "             'barcode12': 'none'# is just water\n",
    "            }\n",
    "\n",
    "\n",
    "summary = pd.DataFrame(data=None, columns = ['species 1', 'species 2','# raw reads','# reads after homology filtering','# reads after length filtering','path to raw reads','path to homology filtering','path to length filtering'])\n",
    "\n",
    "import glob\n",
    "path = \"/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/*/\"\n",
    "path_names = glob.glob(path)\n",
    "for path in path_names:\n",
    "    if path[-13:-1] != 'unclassified' and path[-10:-1] != 'barcode12':\n",
    "        key = path[-10:-1]\n",
    "        if args.verbose:\n",
    "            print('\\033[0;34m' + \"Opened barcode \" + '\\033[0;35m' + key + '\\033[1;37m')\n",
    "        if len(suspected[key].split(',')) == 1:\n",
    "            species1 = suspected[key].split(',')[0].split('__')[1]\n",
    "            species2 = None\n",
    "        elif len(suspected[key].split(',')) == 2:\n",
    "            species1 = suspected[key].split(',')[0].split('__')[1]\n",
    "            species2 = suspected[key].split(',')[1].split('__')[1]\n",
    "        \n",
    "        if args.verbose:\n",
    "            print('\\033[1;36m' + \"BEGIN RAW\"'\\033[1;37m')\n",
    "        raw_path = \"analysis/Concatenated/wheat/\"+key+\"/merged.fastq\"\n",
    "        raw_reads = SeqIO.to_dict(SeqIO.parse(raw_path, \"fastq\"))\n",
    "        raw_count = 0\n",
    "        for entry in raw_reads:\n",
    "            raw_count += 1\n",
    "            \n",
    "        if args.verbose:\n",
    "            print('\\033[1;36m' + \"BEGIN HOMOLOGY\"+ '\\033[1;37m')\n",
    "        homology_path = \"analysis/Python_Processing/wheat/\"+key+\"/combined_test.paf\"\n",
    "        homology_paf = pd.read_csv(homology_path, sep='\\t', header=None, engine='python')\n",
    "        homology_reads = {}\n",
    "        homology_count = 0\n",
    "        for entry in homology_paf[0].unique():\n",
    "            homology_count += 1\n",
    "            \n",
    "        if args.verbose:\n",
    "            print('\\033[1;36m' + \"BEGIN LENGTH\" + '\\033[1;37m')\n",
    "        length_path = \"analysis/Length_Filtered/wheat/\"+key+\"/length_restricted_reads.fasta\"\n",
    "        length_reads = SeqIO.to_dict(SeqIO.parse(length_path, \"fasta\"))\n",
    "        length_count = 0\n",
    "        for entry in length_reads:\n",
    "            length_count += 1\n",
    "            \n",
    "        if args.verbose:\n",
    "            print('\\033[1;36m' + \"BEGIN USE\" + '\\033[1;37m')\n",
    "        use_path = length_path\n",
    "        use_count = length_count\n",
    "\n",
    "\n",
    "        add = pd.DataFrame([[species1, species2, raw_count, homology_count, length_count, use_count, raw_path, homology_path, length_path, use_path]], columns = ['species1', 'species2','# raw reads','# reads after homology filtering','# reads after length filtering','# for use', 'path to raw reads','path to homology filtering','path to length filtering', 'path for use'], index=[key])\n",
    "        summary = summary.append([add])\n",
    "summary = summary.sort_index(axis=0)\n",
    "summary = summary[['species1', 'species2','# raw reads','# reads after homology filtering','# reads after length filtering','# for use', 'path to raw reads','path to homology filtering','path to length filtering', 'path for use']]\n",
    "summary.to_csv(\"analysis/Stats/wheat_reference_dataframe.csv\")\n",
    "if args.verbose:\n",
    "            print('\\033[0;34m' + \"Reference Dataframe saved to \" + '\\033[0;35m' + \"analysis/Stats/wheat_reference_dataframe.csv\" + '\\033[1;37m')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
